{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial: Rare Variant Association Testing with `cellink`\n",
    "\n",
    "This tutorial demonstrates how to perform rare variant association testing (RVAT) in single-cell data using the `cellink` package. We focus on two major types of RVAT: burden tests and the Sequence Kernel Association Test (SKAT). Rare variants can be especially informative for understanding disease biology, but they require careful aggregation and testing approaches due to their low frequency. \n",
    "\n",
    "We’ll walk through how to:\n",
    "- Prepare single-cell and genotype data using `DonorData` objects.\n",
    "- Apply burden tests with multiple variant annotation-based weights.\n",
    "- Combine p-values from multiple tests using ACAT.\n",
    "- Perform SKAT for gene-level rare variant analysis.\n",
    "\n",
    "This tutorial uses a subset of data from the 1k1k project, filtered to the **CD8 Naive** T cell type, and focuses on **chromosome 22** for runtime feasibility. It builds on earlier tutorials that cover pseudobulk expression and variant annotation, which should be reviewed for full context.\n",
    "\n",
    "> Note: This notebook assumes that you have already run the variant annotation step as described in the [Variant Annotation Tutorial](./explore_annotations.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment Setup\n",
    "\n",
    "We begin by importing necessary libraries and defining key parameters for our analysis. `cellink` provides utilities that extend `AnnData` to handle both donor-level genotype data and single-cell RNA-seq data efficiently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.5 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/asyncio/base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/asyncio/base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/asyncio/events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3098, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3153, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3365, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3610, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3670, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/wp/gcjyd9hj0dq8y9__h1sltq000000gn/T/ipykernel_46128/2584568713.py\", line 5, in <module>\n",
      "    import anndata as ad\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/__init__.py\", line 29, in <module>\n",
      "    from .io import read_h5ad, read_zarr\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/io.py\", line 7, in <module>\n",
      "    from ._io.h5ad import read_h5ad, write_h5ad\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/_io/h5ad.py\", line 25, in <module>\n",
      "    from ..experimental import read_dispatched\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/experimental/__init__.py\", line 12, in <module>\n",
      "    from .pytorch import AnnLoader\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/experimental/pytorch/__init__.py\", line 3, in <module>\n",
      "    from ._annloader import AnnLoader\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/anndata/experimental/pytorch/_annloader.py\", line 19, in <module>\n",
      "    import torch\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/__init__.py\", line 1477, in <module>\n",
      "    from .functional import *  # noqa: F403\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/functional.py\", line 9, in <module>\n",
      "    import torch.nn.functional as F\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/nn/__init__.py\", line 1, in <module>\n",
      "    from .modules import *  # noqa: F403\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/nn/modules/__init__.py\", line 35, in <module>\n",
      "    from .transformer import TransformerEncoder, TransformerDecoder, \\\n",
      "  File \"/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/nn/modules/transformer.py\", line 20, in <module>\n",
      "    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n",
      "/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: _ARRAY_API not found (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:84.)\n",
      "  device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n",
      "/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import gc\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import dask.array as da\n",
    "import numpy as np\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "import cellink as cl\n",
    "from cellink._core import DAnn, GAnn\n",
    "from cellink.tl._rvat import run_burden_test, run_skat_test, beta_weighting\n",
    "from cellink.utils import column_normalize, gaussianize\n",
    "\n",
    "from cellink.at.acat import acat_test\n",
    "from cellink.datasets import get_onek1k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA = Path(cl.__file__).parent.parent.parent / \"docs/tutorials/data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_gpcs = 20\n",
    "n_epcs = 15\n",
    "batch_e_pcs_n_top_genes = 2000\n",
    "chrom = 22\n",
    "cis_window = 500_000\n",
    "cell_type = \"CD8 Naive\"\n",
    "celltype_key = \"predicted.celltype.l2\"\n",
    "pb_gex_key = f\"PB_{cell_type}\"  # pseudobulk expression in dd.G.obsm[key_added]\n",
    "original_donor_col = \"donor_id\"\n",
    "min_percent_donors_expressed = 0.1\n",
    "do_debug = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and Prepare Data\n",
    "\n",
    "Here, we load a prepared dataset (`onek1k`) that includes genotype and expression information from human donors. We also extract gene annotations using Ensembl via `pybiomart`, which are essential for defining cis-windows during eQTL analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/larnoldt/cellink_sample_data/onek1k/onek1k_cellxgene.h5ad already exists. Verifying checksum.\n",
      "/Users/larnoldt/cellink_sample_data/onek1k/OneK1K.noGP.vcf.gz already exists. Verifying checksum.\n",
      "/Users/larnoldt/cellink_sample_data/onek1k/OneK1K.noGP.vcf.gz.csi already exists. Verifying checksum.\n",
      "/Users/larnoldt/cellink_sample_data/onek1k/gene_counts_Ensembl_105_phenotype_metadata.tsv.gz already exists. Verifying checksum.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╔═<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> DonorData(n_donors=981, n_cells_per_donor=[333-3,511], donor_id='donor_id') </span>═══════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> G (donors)                                         </span>┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> C (cells)                                          </span>┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 981 ×         │ AnnData object with n_obs × n_vars = 1,248,980 ×   │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'nFeature_RNA', 'percent.mt', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'donor_id', </span>         │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │ 'length', 'quality'                                │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     uns: 'kinship'                                 │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │                                                    │ 'vst.variance.expected',                           │ ║\n",
       "║ │                                                    │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │                                                    │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │                                                    │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type'                   │ ║\n",
       "║ │     obsm: 'gPCs'                                   │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title'      │ ║\n",
       "║ │     varm: 'filter'                                 │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │                                                    │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╔═\u001b[1;38;5;197m DonorData(n_donors=981, n_cells_per_donor=[333-3,511], donor_id='donor_id') \u001b[0m═══════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mG (donors)                                        \u001b[0m\u001b[1;38;5;197m \u001b[0m┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mC (cells)                                         \u001b[0m\u001b[1;38;5;197m \u001b[0m┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 981 ×         │ AnnData object with n_obs × n_vars = 1,248,980 ×   │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'nFeature_RNA', 'percent.mt', \u001b[1;38;5;197m'donor_id', \u001b[0m         │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │ 'length', 'quality'                                │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     uns: 'kinship'                                 │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │                                                    │ 'vst.variance.expected',                           │ ║\n",
       "║ │                                                    │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │                                                    │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │                                                    │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type'                   │ ║\n",
       "║ │     obsm: 'gPCs'                                   │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title'      │ ║\n",
       "║ │     varm: 'filter'                                 │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │                                                    │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = get_onek1k(config_path='../../src/cellink/datasets/config/onek1k.yaml')\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_ensembl_gene_id_start_end_chr():\n",
    "    from pybiomart import Server\n",
    "    server = Server(host='http://www.ensembl.org')\n",
    "    dataset = (server.marts['ENSEMBL_MART_ENSEMBL'].datasets['hsapiens_gene_ensembl'])\n",
    "    ensembl_gene_id_start_end_chr = dataset.query(attributes=['ensembl_gene_id', 'start_position', 'end_position', 'chromosome_name'])\n",
    "    ensembl_gene_id_start_end_chr = ensembl_gene_id_start_end_chr.set_index(\"Gene stable ID\")\n",
    "    ensembl_gene_id_start_end_chr = ensembl_gene_id_start_end_chr.rename(columns={\n",
    "        \"Gene start (bp)\": GAnn.start,\n",
    "        \"Gene end (bp)\": GAnn.end,\n",
    "        \"Chromosome/scaffold name\": GAnn.chrom,\n",
    "    })\n",
    "    return ensembl_gene_id_start_end_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210082</th>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000209082</th>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197989</th>\n",
       "      <td>28578536</td>\n",
       "      <td>28583132</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229388</th>\n",
       "      <td>28643159</td>\n",
       "      <td>28648829</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000289291</th>\n",
       "      <td>28736044</td>\n",
       "      <td>28737670</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274978</th>\n",
       "      <td>28648600</td>\n",
       "      <td>28648734</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000296488</th>\n",
       "      <td>144460887</td>\n",
       "      <td>144462283</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86364 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     start        end chrom\n",
       "Gene stable ID                             \n",
       "ENSG00000210049        577        647    MT\n",
       "ENSG00000211459        648       1601    MT\n",
       "ENSG00000210077       1602       1670    MT\n",
       "ENSG00000210082       1671       3229    MT\n",
       "ENSG00000209082       3230       3304    MT\n",
       "...                    ...        ...   ...\n",
       "ENSG00000197989   28578536   28583132     1\n",
       "ENSG00000229388   28643159   28648829     1\n",
       "ENSG00000289291   28736044   28737670     1\n",
       "ENSG00000274978   28648600   28648734     1\n",
       "ENSG00000296488  144460887  144462283     1\n",
       "\n",
       "[86364 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_gene_id_start_end_chr = _get_ensembl_gene_id_start_end_chr()\n",
    "ensembl_gene_id_start_end_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "dd.C.var = dd.C.var.join(ensembl_gene_id_start_end_chr)\n",
    "dd.C.obs[DAnn.donor] = dd.C.obs[original_donor_col]\n",
    "dd.G.obsm[\"gPCs\"] = dd.G.obsm[\"gPCs\"][dd.G.obsm[\"gPCs\"].columns[:n_gpcs]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalize and Filter Expression Data\n",
    "\n",
    "We normalize single-cell expression using total-count normalization and log transformation. We then filter the dataset to only include cells of a specific type (`CD8 Naive`). This step ensures we're analyzing homogeneous populations, which increases power in eQTL detection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(dd.C)\n",
    "sc.pp.log1p(dd.C)\n",
    "sc.pp.normalize_total(dd.C)\n",
    "\n",
    "# are the expression pcs computed by pseudobulking across all cell types?\n",
    "mdata = sc.get.aggregate(dd.C, by=DAnn.donor, func=\"mean\")\n",
    "mdata.X = mdata.layers.pop(\"mean\")\n",
    "\n",
    "sc.pp.highly_variable_genes(mdata, n_top_genes=batch_e_pcs_n_top_genes)\n",
    "sc.tl.pca(mdata, n_comps=n_epcs)\n",
    "\n",
    "dd.G.obsm[\"ePCs\"] = mdata[dd.G.obs_names].obsm[\"X_pca\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╔═<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') </span>═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> G (donors)                                         </span>┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> C (cells)                                          </span>┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 ×         │ AnnData object with n_obs × n_vars = 52,538 ×      │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'nFeature_RNA', 'percent.mt', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'donor_id', </span>         │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │ 'length', 'quality'                                │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     uns: 'kinship'                                 │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │                                                    │ 'vst.variance.expected',                           │ ║\n",
       "║ │                                                    │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │                                                    │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │                                                    │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs'                           │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     varm: 'filter'                                 │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │                                                    │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╔═\u001b[1;38;5;197m DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') \u001b[0m═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mG (donors)                                        \u001b[0m\u001b[1;38;5;197m \u001b[0m┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mC (cells)                                         \u001b[0m\u001b[1;38;5;197m \u001b[0m┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 ×         │ AnnData object with n_obs × n_vars = 52,538 ×      │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'nFeature_RNA', 'percent.mt', \u001b[1;38;5;197m'donor_id', \u001b[0m         │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │ 'length', 'quality'                                │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     uns: 'kinship'                                 │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │                                                    │ 'vst.variance.expected',                           │ ║\n",
       "║ │                                                    │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │                                                    │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │                                                    │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs'                           │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     varm: 'filter'                                 │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │                                                    │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = dd[..., dd.C.obs[celltype_key] == cell_type, :].copy()\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2255"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:cellink._core.donordata:Aggregated X to PB_CD8 Naive\n",
      "INFO:cellink._core.donordata:Observation found for 980 donors.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╔═<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') </span>═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> G (donors)                                         </span>┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> C (cells)                                          </span>┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 ×         │ AnnData object with n_obs × n_vars = 52,538 ×      │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'donor_id', </span>         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship'                                 │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'PB_CD8 Naive'</span>           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter'                                 │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╔═\u001b[1;38;5;197m DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') \u001b[0m═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mG (donors)                                        \u001b[0m\u001b[1;38;5;197m \u001b[0m┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mC (cells)                                         \u001b[0m\u001b[1;38;5;197m \u001b[0m┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 ×         │ AnnData object with n_obs × n_vars = 52,538 ×      │ ║\n",
       "║ │ 10,595,884                                         │ 36,469                                             │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', \u001b[1;38;5;197m'donor_id', \u001b[0m         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship'                                 │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', \u001b[1;38;5;197m'PB_CD8 Naive'\u001b[0m           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter'                                 │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd.aggregate(key_added=pb_gex_key, sync_var=True, verbose=True)\n",
    "dd.aggregate(obs=[\"sex\", \"age\"], func=\"first\", add_to_obs=True)\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PB_CD8 Naive shape: (980, 36469)\n",
      "dd.shape: (980, 10595884, 52538, 36469)\n",
      "after filtering\n",
      "PB_CD8 Naive shape: (980, 12642)\n",
      "dd.shape: (980, 10595884, 52538, 12642)\n"
     ]
    }
   ],
   "source": [
    "print(f\"{pb_gex_key} shape:\", dd.G.obsm[pb_gex_key].shape)\n",
    "print(\"dd.shape:\", dd.shape)\n",
    "\n",
    "keep_genes = ((dd.G.obsm[pb_gex_key] > 0).mean(axis=0) >= min_percent_donors_expressed).values\n",
    "dd = dd[..., keep_genes]\n",
    "print(\"after filtering\")\n",
    "print(f\"{pb_gex_key} shape:\", dd.G.obsm[pb_gex_key].shape)\n",
    "print(\"dd.shape:\", dd.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╔═<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') </span>═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> G (donors)                                         </span>┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> C (cells)                                          </span>┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 × 136,776 │ AnnData object with n_obs × n_vars = 52,538 × 317  │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'donor_id', </span>         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship'                                 │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'PB_CD8 Naive'</span>           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter'                                 │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╔═\u001b[1;38;5;197m DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') \u001b[0m═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mG (donors)                                        \u001b[0m\u001b[1;38;5;197m \u001b[0m┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mC (cells)                                         \u001b[0m\u001b[1;38;5;197m \u001b[0m┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 × 136,776 │ AnnData object with n_obs × n_vars = 52,538 × 317  │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', \u001b[1;38;5;197m'donor_id', \u001b[0m         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship'                                 │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', \u001b[1;38;5;197m'PB_CD8 Naive'\u001b[0m           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter'                                 │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alternative to dd[:, dd.G.var.chrom == str(chrom), :, dd.C.var.chrom == str(chrom)]\n",
    "dd = dd.sel(G_var=dd.G.var.chrom == str(chrom), C_var=dd.C.var.chrom == str(chrom)).copy()\n",
    "dd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Annotate Variants with Functional Information\n",
    "\n",
    "To inform the weighting in burden tests, we annotate variants using the VEP tool. These annotations include predicted functional consequences (e.g., missense, stop gained), CADD scores, and distance to gene transcription start sites (TSS).\n",
    "\n",
    "This information allows biologically meaningful prioritization of variants when aggregating their effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "vep_annotation_file = DATA / \"variant_annotation/variants_vep_annotated_TEST.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:cellink.tl._annotate_snps_genotype_data:Preparing VEP annotations for addition to gdata\n",
      "INFO:cellink.tl._annotate_snps_genotype_data:Reading annotation file /Users/larnoldt/sc-genetics/docs/tutorials/data/variant_annotation/variants_vep_annotated_TEST.txt\n",
      "/Users/larnoldt/sc-genetics/src/cellink/tl/_annotate_snps_genotype_data.py:418: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annos = pd.read_csv(vep_anno_file, sep=\"\\t\", skiprows=_get_vep_start_row(vep_anno_file))\n",
      "INFO:cellink.tl._annotate_snps_genotype_data:Annotation file loaded\n",
      "INFO:cellink.tl._annotate_snps_genotype_data:Annotation columns: ['snp_id', 'Location', 'Allele', 'gene_id', 'transcript_id', 'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen', 'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_MID_AF', 'gnomADe_NFE_AF', 'gnomADe_REMAINING_AF', 'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO']\n",
      "INFO:cellink.tl._annotate_snps_genotype_data:Changing dtype of categorical columns ['snp_id', 'gene_id', 'transcript_id', 'gnomADe_FIN_AF', 'IMPACT', 'gnomADe_SAS_AF', 'STRAND', 'gnomADe_AF', 'Existing_variation', 'gnomADe_ASJ_AF', 'cDNA_position', 'SIFT', 'CANONICAL', 'FLAGS', 'gnomADe_AFR_AF', 'Codons', 'Protein_position', 'gnomADe_REMAINING_AF', 'gnomADe_MID_AF', 'ENSP', 'gnomADe_AMR_AF', 'gnomADe_EAS_AF', 'SOMATIC', 'PolyPhen', 'CDS_position', 'DISTANCE', 'CLIN_SIG', 'gnomADe_NFE_AF', 'PHENO', 'Amino_acids', 'BIOTYPE']\n",
      "INFO:cellink.tl._annotate_snps_genotype_data:VEP annotations added to gdata\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Consequence_3_prime_UTR_variant</th>\n",
       "      <th>Consequence_5_prime_UTR_variant</th>\n",
       "      <th>Consequence_NMD_transcript_variant</th>\n",
       "      <th>Consequence_coding_sequence_variant</th>\n",
       "      <th>Consequence_downstream_gene_variant</th>\n",
       "      <th>Consequence_frameshift_variant</th>\n",
       "      <th>Consequence_inframe_deletion</th>\n",
       "      <th>Consequence_inframe_insertion</th>\n",
       "      <th>Consequence_intergenic_variant</th>\n",
       "      <th>Consequence_intron_variant</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomADe_EAS_AF</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>gnomADe_NFE_AF</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>BIOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_16388891_G_A</th>\n",
       "      <th>ENSG00000230643</th>\n",
       "      <th>ENST00000447704</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>594.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16388968_C_T</th>\n",
       "      <th>ENSG00000230643</th>\n",
       "      <th>ENST00000447704</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>517.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16389525_A_G</th>\n",
       "      <th>ENSG00000230643</th>\n",
       "      <th>ENST00000447704</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16390411_G_A</th>\n",
       "      <th>ENSG00000230643</th>\n",
       "      <th>ENST00000447704</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>809.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16391555_G_C</th>\n",
       "      <th>ENSG00000230643</th>\n",
       "      <th>ENST00000447704</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50796327_CCA_C</th>\n",
       "      <th>ENSG00000100239</th>\n",
       "      <th>ENST00000395741</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50798021_A_G</th>\n",
       "      <th>ENSG00000100239</th>\n",
       "      <th>ENST00000395741</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50798635_T_C</th>\n",
       "      <th>ENSG00000100239</th>\n",
       "      <th>ENST00000395741</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50799821_A_C</th>\n",
       "      <th>ENSG00000100239</th>\n",
       "      <th>ENST00000395741</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50802428_C_G</th>\n",
       "      <th>ENSG00000100239</th>\n",
       "      <th>ENST00000395741</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169322 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Consequence_3_prime_UTR_variant  \\\n",
       "snp_id            gene_id         transcript_id                                      \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                                0   \n",
       "...                                                                            ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                                0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                                0   \n",
       "\n",
       "                                                   Consequence_5_prime_UTR_variant  \\\n",
       "snp_id            gene_id         transcript_id                                      \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                                0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                                0   \n",
       "...                                                                            ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                                0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                                0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                                0   \n",
       "\n",
       "                                                   Consequence_NMD_transcript_variant  \\\n",
       "snp_id            gene_id         transcript_id                                         \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                                   0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                                   0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                                   0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                                   0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                                   0   \n",
       "...                                                                               ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                                   0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                                   0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                                   0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                                   0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                                   0   \n",
       "\n",
       "                                                   Consequence_coding_sequence_variant  \\\n",
       "snp_id            gene_id         transcript_id                                          \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                                    0   \n",
       "...                                                                                ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                                    0   \n",
       "\n",
       "                                                   Consequence_downstream_gene_variant  \\\n",
       "snp_id            gene_id         transcript_id                                          \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                                    1   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                                    1   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                                    0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                                    0   \n",
       "...                                                                                ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                                    0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                                    0   \n",
       "\n",
       "                                                   Consequence_frameshift_variant  \\\n",
       "snp_id            gene_id         transcript_id                                     \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                               0   \n",
       "...                                                                           ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                               0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                               0   \n",
       "\n",
       "                                                   Consequence_inframe_deletion  \\\n",
       "snp_id            gene_id         transcript_id                                   \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                             0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                             0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                             0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                             0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                             0   \n",
       "...                                                                         ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                             0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                             0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                             0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                             0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                             0   \n",
       "\n",
       "                                                   Consequence_inframe_insertion  \\\n",
       "snp_id            gene_id         transcript_id                                    \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                              0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                              0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                              0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                              0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                              0   \n",
       "...                                                                          ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                              0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                              0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                              0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                              0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                              0   \n",
       "\n",
       "                                                   Consequence_intergenic_variant  \\\n",
       "snp_id            gene_id         transcript_id                                     \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                               0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                               0   \n",
       "...                                                                           ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                               0   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                               0   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                               0   \n",
       "\n",
       "                                                   Consequence_intron_variant  \\\n",
       "snp_id            gene_id         transcript_id                                 \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704                           0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704                           0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704                           0   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704                           0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704                           0   \n",
       "...                                                                       ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741                           1   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741                           1   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741                           1   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741                           1   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741                           1   \n",
       "\n",
       "                                                   ...  gnomADe_EAS_AF  \\\n",
       "snp_id            gene_id         transcript_id    ...                   \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704  ...             NaN   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704  ...             NaN   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704  ...             NaN   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704  ...             NaN   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704  ...             NaN   \n",
       "...                                                ...             ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741  ...             NaN   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741  ...             NaN   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741  ...             NaN   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741  ...             NaN   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741  ...             NaN   \n",
       "\n",
       "                                                   SOMATIC  PolyPhen  \\\n",
       "snp_id            gene_id         transcript_id                        \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704        -       NaN   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704        -       NaN   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704        -       NaN   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704        -       NaN   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704        -       NaN   \n",
       "...                                                    ...       ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741        -       NaN   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741        -       NaN   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741        -       NaN   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741        -       NaN   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741        -       NaN   \n",
       "\n",
       "                                                   CDS_position  DISTANCE  \\\n",
       "snp_id            gene_id         transcript_id                             \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704             -     594.0   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704             -     517.0   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704             -       NaN   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704             -     809.0   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704             -    1953.0   \n",
       "...                                                         ...       ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741             -       NaN   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741             -       NaN   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741             -       NaN   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741             -       NaN   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741             -       NaN   \n",
       "\n",
       "                                                   CLIN_SIG  gnomADe_NFE_AF  \\\n",
       "snp_id            gene_id         transcript_id                               \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704         -             NaN   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704         -             NaN   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704         -             NaN   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704         -             NaN   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704         -             NaN   \n",
       "...                                                     ...             ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741         -             NaN   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741         -             NaN   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741         -             NaN   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741         -             NaN   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741         -             NaN   \n",
       "\n",
       "                                                   PHENO  Amino_acids  \\\n",
       "snp_id            gene_id         transcript_id                         \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704      -            -   \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704      -            -   \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704      -            -   \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704      -            -   \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704      -            -   \n",
       "...                                                  ...          ...   \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741      -            -   \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741      -            -   \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741      -            -   \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741      -            -   \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741      -            -   \n",
       "\n",
       "                                                                  BIOTYPE  \n",
       "snp_id            gene_id         transcript_id                            \n",
       "22_16388891_G_A   ENSG00000230643 ENST00000447704  unprocessed_pseudogene  \n",
       "22_16388968_C_T   ENSG00000230643 ENST00000447704  unprocessed_pseudogene  \n",
       "22_16389525_A_G   ENSG00000230643 ENST00000447704  unprocessed_pseudogene  \n",
       "22_16390411_G_A   ENSG00000230643 ENST00000447704  unprocessed_pseudogene  \n",
       "22_16391555_G_C   ENSG00000230643 ENST00000447704  unprocessed_pseudogene  \n",
       "...                                                                   ...  \n",
       "22_50796327_CCA_C ENSG00000100239 ENST00000395741          protein_coding  \n",
       "22_50798021_A_G   ENSG00000100239 ENST00000395741          protein_coding  \n",
       "22_50798635_T_C   ENSG00000100239 ENST00000395741          protein_coding  \n",
       "22_50799821_A_C   ENSG00000100239 ENST00000395741          protein_coding  \n",
       "22_50802428_C_G   ENSG00000100239 ENST00000395741          protein_coding  \n",
       "\n",
       "[169322 rows x 55 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.tl.add_vep_annos_to_gdata(vep_anno_file=vep_annotation_file, gdata=dd.G, dummy_consequence=True)\n",
    "dd.G.uns[\"variant_annotation_vep\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:cellink.tl._annotate_snps_genotype_data:Aggregating using method: first\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Consequence_3_prime_UTR_variant</th>\n",
       "      <th>Consequence_5_prime_UTR_variant</th>\n",
       "      <th>Consequence_NMD_transcript_variant</th>\n",
       "      <th>Consequence_coding_sequence_variant</th>\n",
       "      <th>Consequence_downstream_gene_variant</th>\n",
       "      <th>Consequence_frameshift_variant</th>\n",
       "      <th>Consequence_inframe_deletion</th>\n",
       "      <th>Consequence_inframe_insertion</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomADe_EAS_AF</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>gnomADe_NFE_AF</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>BIOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_16388891_G_A</th>\n",
       "      <td>ENSG00000230643</td>\n",
       "      <td>ENST00000447704</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>594.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16388968_C_T</th>\n",
       "      <td>ENSG00000230643</td>\n",
       "      <td>ENST00000447704</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>517.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16389525_A_G</th>\n",
       "      <td>ENSG00000230643</td>\n",
       "      <td>ENST00000447704</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16390411_G_A</th>\n",
       "      <td>ENSG00000230643</td>\n",
       "      <td>ENST00000447704</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>809.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16391555_G_C</th>\n",
       "      <td>ENSG00000230643</td>\n",
       "      <td>ENST00000447704</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50796327_CCA_C</th>\n",
       "      <td>ENSG00000100239</td>\n",
       "      <td>ENST00000395741</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50798021_A_G</th>\n",
       "      <td>ENSG00000100239</td>\n",
       "      <td>ENST00000395741</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50798635_T_C</th>\n",
       "      <td>ENSG00000100239</td>\n",
       "      <td>ENST00000395741</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50799821_A_C</th>\n",
       "      <td>ENSG00000100239</td>\n",
       "      <td>ENST00000395741</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50802428_C_G</th>\n",
       "      <td>ENSG00000100239</td>\n",
       "      <td>ENST00000395741</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136776 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    transcript_id  \\\n",
       "snp_id                                                \n",
       "22_16388891_G_A    ENSG00000230643  ENST00000447704   \n",
       "22_16388968_C_T    ENSG00000230643  ENST00000447704   \n",
       "22_16389525_A_G    ENSG00000230643  ENST00000447704   \n",
       "22_16390411_G_A    ENSG00000230643  ENST00000447704   \n",
       "22_16391555_G_C    ENSG00000230643  ENST00000447704   \n",
       "...                            ...              ...   \n",
       "22_50796327_CCA_C  ENSG00000100239  ENST00000395741   \n",
       "22_50798021_A_G    ENSG00000100239  ENST00000395741   \n",
       "22_50798635_T_C    ENSG00000100239  ENST00000395741   \n",
       "22_50799821_A_C    ENSG00000100239  ENST00000395741   \n",
       "22_50802428_C_G    ENSG00000100239  ENST00000395741   \n",
       "\n",
       "                   Consequence_3_prime_UTR_variant  \\\n",
       "snp_id                                               \n",
       "22_16388891_G_A                                  0   \n",
       "22_16388968_C_T                                  0   \n",
       "22_16389525_A_G                                  0   \n",
       "22_16390411_G_A                                  0   \n",
       "22_16391555_G_C                                  0   \n",
       "...                                            ...   \n",
       "22_50796327_CCA_C                                0   \n",
       "22_50798021_A_G                                  0   \n",
       "22_50798635_T_C                                  0   \n",
       "22_50799821_A_C                                  0   \n",
       "22_50802428_C_G                                  0   \n",
       "\n",
       "                   Consequence_5_prime_UTR_variant  \\\n",
       "snp_id                                               \n",
       "22_16388891_G_A                                  0   \n",
       "22_16388968_C_T                                  0   \n",
       "22_16389525_A_G                                  0   \n",
       "22_16390411_G_A                                  0   \n",
       "22_16391555_G_C                                  0   \n",
       "...                                            ...   \n",
       "22_50796327_CCA_C                                0   \n",
       "22_50798021_A_G                                  0   \n",
       "22_50798635_T_C                                  0   \n",
       "22_50799821_A_C                                  0   \n",
       "22_50802428_C_G                                  0   \n",
       "\n",
       "                   Consequence_NMD_transcript_variant  \\\n",
       "snp_id                                                  \n",
       "22_16388891_G_A                                     0   \n",
       "22_16388968_C_T                                     0   \n",
       "22_16389525_A_G                                     0   \n",
       "22_16390411_G_A                                     0   \n",
       "22_16391555_G_C                                     0   \n",
       "...                                               ...   \n",
       "22_50796327_CCA_C                                   0   \n",
       "22_50798021_A_G                                     0   \n",
       "22_50798635_T_C                                     0   \n",
       "22_50799821_A_C                                     0   \n",
       "22_50802428_C_G                                     0   \n",
       "\n",
       "                   Consequence_coding_sequence_variant  \\\n",
       "snp_id                                                   \n",
       "22_16388891_G_A                                      0   \n",
       "22_16388968_C_T                                      0   \n",
       "22_16389525_A_G                                      0   \n",
       "22_16390411_G_A                                      0   \n",
       "22_16391555_G_C                                      0   \n",
       "...                                                ...   \n",
       "22_50796327_CCA_C                                    0   \n",
       "22_50798021_A_G                                      0   \n",
       "22_50798635_T_C                                      0   \n",
       "22_50799821_A_C                                      0   \n",
       "22_50802428_C_G                                      0   \n",
       "\n",
       "                   Consequence_downstream_gene_variant  \\\n",
       "snp_id                                                   \n",
       "22_16388891_G_A                                      1   \n",
       "22_16388968_C_T                                      1   \n",
       "22_16389525_A_G                                      0   \n",
       "22_16390411_G_A                                      0   \n",
       "22_16391555_G_C                                      0   \n",
       "...                                                ...   \n",
       "22_50796327_CCA_C                                    0   \n",
       "22_50798021_A_G                                      0   \n",
       "22_50798635_T_C                                      0   \n",
       "22_50799821_A_C                                      0   \n",
       "22_50802428_C_G                                      0   \n",
       "\n",
       "                   Consequence_frameshift_variant  \\\n",
       "snp_id                                              \n",
       "22_16388891_G_A                                 0   \n",
       "22_16388968_C_T                                 0   \n",
       "22_16389525_A_G                                 0   \n",
       "22_16390411_G_A                                 0   \n",
       "22_16391555_G_C                                 0   \n",
       "...                                           ...   \n",
       "22_50796327_CCA_C                               0   \n",
       "22_50798021_A_G                                 0   \n",
       "22_50798635_T_C                                 0   \n",
       "22_50799821_A_C                                 0   \n",
       "22_50802428_C_G                                 0   \n",
       "\n",
       "                   Consequence_inframe_deletion  \\\n",
       "snp_id                                            \n",
       "22_16388891_G_A                               0   \n",
       "22_16388968_C_T                               0   \n",
       "22_16389525_A_G                               0   \n",
       "22_16390411_G_A                               0   \n",
       "22_16391555_G_C                               0   \n",
       "...                                         ...   \n",
       "22_50796327_CCA_C                             0   \n",
       "22_50798021_A_G                               0   \n",
       "22_50798635_T_C                               0   \n",
       "22_50799821_A_C                               0   \n",
       "22_50802428_C_G                               0   \n",
       "\n",
       "                   Consequence_inframe_insertion  ...  gnomADe_EAS_AF  \\\n",
       "snp_id                                            ...                   \n",
       "22_16388891_G_A                                0  ...             NaN   \n",
       "22_16388968_C_T                                0  ...             NaN   \n",
       "22_16389525_A_G                                0  ...             NaN   \n",
       "22_16390411_G_A                                0  ...             NaN   \n",
       "22_16391555_G_C                                0  ...             NaN   \n",
       "...                                          ...  ...             ...   \n",
       "22_50796327_CCA_C                              0  ...             NaN   \n",
       "22_50798021_A_G                                0  ...             NaN   \n",
       "22_50798635_T_C                                0  ...             NaN   \n",
       "22_50799821_A_C                                0  ...             NaN   \n",
       "22_50802428_C_G                                0  ...             NaN   \n",
       "\n",
       "                   SOMATIC  PolyPhen  CDS_position  DISTANCE  CLIN_SIG  \\\n",
       "snp_id                                                                   \n",
       "22_16388891_G_A          -       NaN             -     594.0         -   \n",
       "22_16388968_C_T          -       NaN             -     517.0         -   \n",
       "22_16389525_A_G          -       NaN             -       NaN         -   \n",
       "22_16390411_G_A          -       NaN             -     809.0         -   \n",
       "22_16391555_G_C          -       NaN             -    1953.0         -   \n",
       "...                    ...       ...           ...       ...       ...   \n",
       "22_50796327_CCA_C        -       NaN             -       NaN         -   \n",
       "22_50798021_A_G          -       NaN             -       NaN         -   \n",
       "22_50798635_T_C          -       NaN             -       NaN         -   \n",
       "22_50799821_A_C          -       NaN             -       NaN         -   \n",
       "22_50802428_C_G          -       NaN             -       NaN         -   \n",
       "\n",
       "                   gnomADe_NFE_AF  PHENO  Amino_acids                 BIOTYPE  \n",
       "snp_id                                                                         \n",
       "22_16388891_G_A               NaN      -            -  unprocessed_pseudogene  \n",
       "22_16388968_C_T               NaN      -            -  unprocessed_pseudogene  \n",
       "22_16389525_A_G               NaN      -            -  unprocessed_pseudogene  \n",
       "22_16390411_G_A               NaN      -            -  unprocessed_pseudogene  \n",
       "22_16391555_G_C               NaN      -            -  unprocessed_pseudogene  \n",
       "...                           ...    ...          ...                     ...  \n",
       "22_50796327_CCA_C             NaN      -            -          protein_coding  \n",
       "22_50798021_A_G               NaN      -            -          protein_coding  \n",
       "22_50798635_T_C               NaN      -            -          protein_coding  \n",
       "22_50799821_A_C               NaN      -            -          protein_coding  \n",
       "22_50802428_C_G               NaN      -            -          protein_coding  \n",
       "\n",
       "[136776 rows x 57 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.tl.aggregate_annotations_for_varm(\n",
    "    dd.G, \"variant_annotation_vep\", agg_type=\"first\", return_data=True\n",
    ")  # TODO change agg type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_id', 'transcript_id', 'Consequence_3_prime_UTR_variant',\n",
       "       'Consequence_5_prime_UTR_variant', 'Consequence_NMD_transcript_variant',\n",
       "       'Consequence_coding_sequence_variant',\n",
       "       'Consequence_downstream_gene_variant', 'Consequence_frameshift_variant',\n",
       "       'Consequence_inframe_deletion', 'Consequence_inframe_insertion',\n",
       "       'Consequence_intergenic_variant', 'Consequence_intron_variant',\n",
       "       'Consequence_mature_miRNA_variant', 'Consequence_missense_variant',\n",
       "       'Consequence_non_coding_transcript_exon_variant',\n",
       "       'Consequence_non_coding_transcript_variant',\n",
       "       'Consequence_protein_altering_variant',\n",
       "       'Consequence_splice_acceptor_variant',\n",
       "       'Consequence_splice_donor_5th_base_variant',\n",
       "       'Consequence_splice_donor_region_variant',\n",
       "       'Consequence_splice_donor_variant',\n",
       "       'Consequence_splice_polypyrimidine_tract_variant',\n",
       "       'Consequence_splice_region_variant', 'Consequence_start_lost',\n",
       "       'Consequence_stop_gained', 'Consequence_stop_lost',\n",
       "       'Consequence_stop_retained_variant', 'Consequence_synonymous_variant',\n",
       "       'Consequence_upstream_gene_variant', 'gnomADe_FIN_AF', 'IMPACT',\n",
       "       'gnomADe_SAS_AF', 'STRAND', 'gnomADe_AF', 'Existing_variation',\n",
       "       'gnomADe_ASJ_AF', 'cDNA_position', 'SIFT', 'CANONICAL', 'FLAGS',\n",
       "       'gnomADe_AFR_AF', 'Codons', 'Protein_position', 'gnomADe_REMAINING_AF',\n",
       "       'gnomADe_MID_AF', 'ENSP', 'gnomADe_AMR_AF', 'gnomADe_EAS_AF', 'SOMATIC',\n",
       "       'PolyPhen', 'CDS_position', 'DISTANCE', 'CLIN_SIG', 'gnomADe_NFE_AF',\n",
       "       'PHENO', 'Amino_acids', 'BIOTYPE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd.G.varm[\"variant_annotation\"].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'CADD_RAW'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/pandas/core/indexes/base.py:3805\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3804\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m3805\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3806\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:167\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:196\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: 'CADD_RAW'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mdd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mG\u001b[49m\u001b[43m.\u001b[49m\u001b[43muns\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mvariant_annotation_vep\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mCADD_RAW\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m.describe()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/pandas/core/frame.py:4102\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4100\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.nlevels > \u001b[32m1\u001b[39m:\n\u001b[32m   4101\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._getitem_multilevel(key)\n\u001b[32m-> \u001b[39m\u001b[32m4102\u001b[39m indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4103\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[32m   4104\u001b[39m     indexer = [indexer]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/pandas/core/indexes/base.py:3812\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3807\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   3808\u001b[39m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc.Iterable)\n\u001b[32m   3809\u001b[39m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[32m   3810\u001b[39m     ):\n\u001b[32m   3811\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[32m-> \u001b[39m\u001b[32m3812\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   3813\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[32m   3814\u001b[39m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[32m   3815\u001b[39m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[32m   3816\u001b[39m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[32m   3817\u001b[39m     \u001b[38;5;28mself\u001b[39m._check_indexing_error(key)\n",
      "\u001b[31mKeyError\u001b[39m: 'CADD_RAW'"
     ]
    }
   ],
   "source": [
    "dd.G.uns[\"variant_annotation_vep\"][\"CADD_RAW\"].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Burden Testing\n",
    "\n",
    "We now perform burden tests, which evaluate whether the **aggregate effect** of rare variants within a gene is associated with a phenotype — in this case, gene expression in CD8 Naive cells. We use different weighting schemes allow us to prioritize variants differently based on functional scores or allele frequency, including:\n",
    "- **CADD_RAW**: Raw CADD scores.\n",
    "- **maf_beta**: Beta weighting based on minor allele frequency (MAF).\n",
    "- **tss_distance**: Distance to the transcription start site (TSS).\n",
    "- **tss_distance_exp**: Exponential weighting based on TSS distance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "burden_agg_fct = \"sum\"\n",
    "run_lrt = True\n",
    "annotation_cols = [\"maf_beta\"] #[\"CADD_RAW\", \"maf_beta\", \"tss_distance\", \"tss_distance_exp\"]\n",
    "\n",
    "rare_maf_threshold = 0.05"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering for Rare Variants\n",
    "Burden tests focus on rare variants—typically those with minor allele frequency (MAF) below 0.01 or 0.05. We filter our genotype data accordingly to isolate variants in this frequency range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╔═<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') </span>═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> G (donors)                                         </span>┃<span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\"> C (cells)                                          </span>┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 × 39,454  │ AnnData object with n_obs × n_vars = 52,538 × 317  │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'donor_id', </span>         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship', 'variant_annotation_vep'       │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', <span style=\"color: #ff005f; text-decoration-color: #ff005f; font-weight: bold\">'PB_CD8 Naive'</span>           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter', 'variant_annotation'           │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╔═\u001b[1;38;5;197m DonorData(n_donors=980, n_cells_per_donor=[1-583], donor_id='donor_id') \u001b[0m═══════════════════════════════════╗\n",
       "║ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ║\n",
       "║ ┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mG (donors)                                        \u001b[0m\u001b[1;38;5;197m \u001b[0m┃\u001b[1;38;5;197m \u001b[0m\u001b[1;38;5;197mC (cells)                                         \u001b[0m\u001b[1;38;5;197m \u001b[0m┃ ║\n",
       "║ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩ ║\n",
       "║ │ AnnData object with n_obs × n_vars = 980 × 39,454  │ AnnData object with n_obs × n_vars = 52,538 × 317  │ ║\n",
       "║ │     obs: 'sex', 'age'                              │     obs: 'orig.ident', 'nCount_RNA',               │ ║\n",
       "║ │                                                    │ 'nFeature_RNA', 'percent.mt', \u001b[1;38;5;197m'donor_id', \u001b[0m         │ ║\n",
       "║ │                                                    │ 'pool_number', 'predicted.celltype.l2',            │ ║\n",
       "║ │                                                    │ 'predicted.celltype.l2.score', 'age',              │ ║\n",
       "║ │                                                    │ 'organism_ontology_term_id',                       │ ║\n",
       "║ │                                                    │ 'tissue_ontology_term_id',                         │ ║\n",
       "║ │                                                    │ 'assay_ontology_term_id',                          │ ║\n",
       "║ │                                                    │ 'disease_ontology_term_id',                        │ ║\n",
       "║ │                                                    │ 'cell_type_ontology_term_id',                      │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity_ontology_term_id',        │ ║\n",
       "║ │                                                    │ 'development_stage_ontology_term_id',              │ ║\n",
       "║ │                                                    │ 'sex_ontology_term_id', 'is_primary_data',         │ ║\n",
       "║ │                                                    │ 'suspension_type', 'tissue_type', 'cell_type',     │ ║\n",
       "║ │                                                    │ 'assay', 'disease', 'organism', 'sex', 'tissue',   │ ║\n",
       "║ │                                                    │ 'self_reported_ethnicity', 'development_stage',    │ ║\n",
       "║ │                                                    │ 'observation_joinid'                               │ ║\n",
       "║ │     var: 'chrom', 'pos', 'a0', 'a1', 'AC',         │     var: 'vst.mean', 'vst.variance',               │ ║\n",
       "║ │ 'AC_Hemi', 'AC_Het', 'AC_Hom', 'AF', 'AN', 'ER2',  │ 'vst.variance.expected',                           │ ║\n",
       "║ │ 'ExcHet', 'HWE', 'IMPUTED', 'maf', 'NS', 'R2',     │ 'vst.variance.standardized', 'vst.variable',       │ ║\n",
       "║ │ 'TYPED', 'TYPED_ONLY', 'contig', 'id', 'id_mask',  │ 'feature_is_filtered', 'feature_name',             │ ║\n",
       "║ │ 'length', 'quality'                                │ 'feature_reference', 'feature_biotype',            │ ║\n",
       "║ │                                                    │ 'feature_length', 'feature_type', 'start', 'end',  │ ║\n",
       "║ │                                                    │ 'chrom'                                            │ ║\n",
       "║ │     uns: 'kinship', 'variant_annotation_vep'       │     uns: 'cell_type_ontology_term_id_colors',      │ ║\n",
       "║ │                                                    │ 'citation', 'default_embedding',                   │ ║\n",
       "║ │                                                    │ 'schema_reference', 'schema_version', 'title',     │ ║\n",
       "║ │                                                    │ 'log1p'                                            │ ║\n",
       "║ │     obsm: 'gPCs', 'ePCs', \u001b[1;38;5;197m'PB_CD8 Naive'\u001b[0m           │     obsm: 'X_azimuth_spca', 'X_azimuth_umap',      │ ║\n",
       "║ │                                                    │ 'X_harmony', 'X_pca', 'X_umap'                     │ ║\n",
       "║ │     varm: 'filter', 'variant_annotation'           │     varm: 'PCs'                                    │ ║\n",
       "║ └────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘ ║\n",
       "╚═════════════════════════════════════════════════════════════════════════════════════════════════════════════╝\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = dd.sel(G_var=dd.G.var.maf < rare_maf_threshold).copy()\n",
    "dd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Custom MAF Weights\n",
    "\n",
    "We add custom MAF weights commonly used in burden tests, such as `Beta(MAF, 1, 25)`. These weights prioritize rarer variants in the analysis. TSS distance weight as used in the SAIGE-QTL paper are added manually for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "dd.G.varm[\"variant_annotation\"][\"maf_beta\"] = beta_weighting(dd.G.var[\"maf\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run burden tests using each annotation individually for weighting\n",
    "\n",
    "Next, we specify the covariate matrix `F`, which includes sex, age, genetic PCs, and expression PCs. This controls for potential confounders in the association analysis.\n",
    "We then iterate over all genes on chromosome 22, define a cis-window of ±500kb, and run a burden test using each of the annotation-based weighting schemes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This specifies covariates/fixed effects\n",
    "F = np.concatenate(\n",
    "    [\n",
    "        np.ones((dd.shape[0], 1)),\n",
    "        dd.G.obs[[\"sex\"]].values - 1,\n",
    "        dd.G.obs[[\"age\"]].values,\n",
    "        dd.G.obsm[\"gPCs\"].values,\n",
    "        dd.G.obsm[\"ePCs\"],\n",
    "    ],\n",
    "    axis=1,\n",
    ")\n",
    "F[:, 2:] = column_normalize(F[:, 2:].astype(\"float\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "F = F.astype(\"float\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "if isinstance(dd.G.X, da.Array | ad._core.views.DaskArrayView):\n",
    "    if dd.G.is_view:\n",
    "        dd._G = dd._G.copy()\n",
    "    dd.G.X = dd.G.X.compute()\n",
    "\n",
    "if do_debug:\n",
    "    warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/317 [00:00<?, ?it/s]/Users/larnoldt/sc-genetics/src/cellink/at/gwas.py:57: RuntimeWarning: divide by zero encountered in divide\n",
      "  n = 1.0 / (GG - np.einsum(\"ij,ij->j\", FG, A0iFG))\n",
      "/Users/larnoldt/sc-genetics/src/cellink/at/gwas.py:58: RuntimeWarning: invalid value encountered in multiply\n",
      "  M = -n * A0iFG\n",
      "/Users/larnoldt/sc-genetics/src/cellink/at/gwas.py:62: RuntimeWarning: invalid value encountered in multiply\n",
      "  self.beta_g += n[:, None] * GY\n",
      "100%|██████████| 317/317 [01:12<00:00,  4.40it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_gene</th>\n",
       "      <th>egene</th>\n",
       "      <th>weight_col</th>\n",
       "      <th>burden_agg_fct</th>\n",
       "      <th>pv</th>\n",
       "      <th>beta</th>\n",
       "      <th>betaste</th>\n",
       "      <th>lrt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000206195</td>\n",
       "      <td>ENSG00000206195</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.113135</td>\n",
       "      <td>-1.056926e-04</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>2.509878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.724483</td>\n",
       "      <td>-2.297332e-05</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.124237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.075105</td>\n",
       "      <td>-1.045611e-04</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>3.167758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000131100</td>\n",
       "      <td>ENSG00000131100</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.692283</td>\n",
       "      <td>2.165998e-05</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.156625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000205560</td>\n",
       "      <td>ENSG00000205560</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.631133</td>\n",
       "      <td>-2.215657e-05</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.230527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.995639</td>\n",
       "      <td>2.560052e-07</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000205559</td>\n",
       "      <td>ENSG00000205559</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.534873</td>\n",
       "      <td>-2.858532e-05</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.385125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.968449</td>\n",
       "      <td>-1.906906e-06</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.001565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>sum</td>\n",
       "      <td>0.807705</td>\n",
       "      <td>-1.433886e-05</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.059237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        burden_gene            egene weight_col burden_agg_fct        pv  \\\n",
       "0   ENSG00000206195  ENSG00000206195   maf_beta            sum       NaN   \n",
       "0   ENSG00000177663  ENSG00000177663   maf_beta            sum  0.113135   \n",
       "0   ENSG00000069998  ENSG00000069998   maf_beta            sum  0.724483   \n",
       "0   ENSG00000093072  ENSG00000093072   maf_beta            sum  0.075105   \n",
       "0   ENSG00000131100  ENSG00000131100   maf_beta            sum  0.692283   \n",
       "..              ...              ...        ...            ...       ...   \n",
       "0   ENSG00000205560  ENSG00000205560   maf_beta            sum  0.631133   \n",
       "0   ENSG00000100288  ENSG00000100288   maf_beta            sum  0.995639   \n",
       "0   ENSG00000205559  ENSG00000205559   maf_beta            sum  0.534873   \n",
       "0   ENSG00000100299  ENSG00000100299   maf_beta            sum  0.968449   \n",
       "0   ENSG00000079974  ENSG00000079974   maf_beta            sum  0.807705   \n",
       "\n",
       "            beta   betaste       lrt  \n",
       "0            NaN       NaN       NaN  \n",
       "0  -1.056926e-04  0.000067  2.509878  \n",
       "0  -2.297332e-05  0.000065  0.124237  \n",
       "0  -1.045611e-04  0.000059  3.167758  \n",
       "0   2.165998e-05  0.000055  0.156625  \n",
       "..           ...       ...       ...  \n",
       "0  -2.215657e-05  0.000046  0.230527  \n",
       "0   2.560052e-07  0.000047  0.000030  \n",
       "0  -2.858532e-05  0.000046  0.385125  \n",
       "0  -1.906906e-06  0.000048  0.001565  \n",
       "0  -1.433886e-05  0.000059  0.059237  \n",
       "\n",
       "[317 rows x 8 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "for gene, row in tqdm(dd.C.var.iterrows(), total=dd.shape[3]):\n",
    "    Y = gaussianize(dd.G.obsm[pb_gex_key][[gene]].values.astype(\"float\") + 1e-5 * np.random.randn(dd.shape[0], 1))\n",
    "\n",
    "    start = max(0, row.start - cis_window)\n",
    "    end = row.end + cis_window\n",
    "    _G = dd.G[:, (dd.G.var.pos < end)]\n",
    "    _G = _G[:, (_G.var.pos > start)]\n",
    "    _G = _G[:, (_G.X.std(0) != 0)]\n",
    "    _G = _G.copy()\n",
    "\n",
    "    # TODO make strand aware\n",
    "    #_G.varm[\"variant_annotation\"][\"tss_distance\"] = np.abs(row.start - _G.var[\"pos\"])\n",
    "    #_G.varm[\"variant_annotation\"][\"tss_distance_exp\"] = np.exp(-1e-5 * _G.varm[\"variant_annotation\"][\"tss_distance\"])\n",
    "\n",
    "    rdf = run_burden_test(\n",
    "        _G, Y, F, gene, annotation_cols=annotation_cols, burden_agg_fct=burden_agg_fct, run_lrt=run_lrt\n",
    "    )\n",
    "    results.append(rdf)\n",
    "\n",
    "rdf = pd.concat(results)\n",
    "rdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Combine Multiple Tests with ACAT\n",
    "\n",
    "To summarize evidence from multiple annotations, we use the **ACAT (Aggregated Cauchy Association Test)**. This meta-analysis method combines p-values from the burden tests per gene into a single statistic, improving power and interpretability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>egene</th>\n",
       "      <th>pv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>ENSG00000212939</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>ENSG00000244752</td>\n",
       "      <td>0.000429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>ENSG00000100321</td>\n",
       "      <td>0.002216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ENSG00000183530</td>\n",
       "      <td>0.007274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>ENSG00000100353</td>\n",
       "      <td>0.007837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>ENSG00000100162</td>\n",
       "      <td>0.990140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>ENSG00000183628</td>\n",
       "      <td>0.992097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>0.995639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>ENSG00000221963</td>\n",
       "      <td>0.996005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>ENSG00000186654</td>\n",
       "      <td>0.996168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>316 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               egene        pv\n",
       "243  ENSG00000212939  0.000084\n",
       "279  ENSG00000244752  0.000429\n",
       "104  ENSG00000100321  0.002216\n",
       "188  ENSG00000183530  0.007274\n",
       "116  ENSG00000100353  0.007837\n",
       "..               ...       ...\n",
       "68   ENSG00000100162  0.990140\n",
       "190  ENSG00000183628  0.992097\n",
       "92   ENSG00000100288  0.995639\n",
       "250  ENSG00000221963  0.996005\n",
       "213  ENSG00000186654  0.996168\n",
       "\n",
       "[316 rows x 2 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = rdf.dropna(subset=[\"pv\"]).groupby(\"egene\")[\"pv\"].agg(lambda x: acat_test(x.values)).reset_index()\n",
    "combined.sort_values(\"pv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: SKAT\n",
    "\n",
    "As an alternative to burden testing, we apply the Sequence Kernel Association Test (SKAT), which models the variance component of aggregated rare variants rather than assuming a unidirectional effect. SKAT is more robust when variant effects differ in direction or magnitude.\n",
    "\n",
    "Currently, `cellink` supports SKAT with the standard Beta(MAF, 1, 25) weighting scheme."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/317 [00:00<?, ?it/s]INFO:root:Starting optimization ...\n",
      "/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/chiscore/_davies.py:134: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  s1 = c1[2] / c1[1] ** (3 / 2)\n",
      "/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/chiscore/_davies.py:135: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  s2 = c1[3] / c1[1] ** 2\n",
      "/opt/miniconda3/envs/single_cell_base6/lib/python3.12/site-packages/chiscore/_davies.py:113: RuntimeWarning: invalid value encountered in divide\n",
      "  Q_Norm = (Q_all - param[\"muQ\"]) / param[\"sigmaQ\"]\n",
      "  0%|          | 1/317 [00:00<01:42,  3.09it/s]INFO:root:Starting optimization ...\n",
      "  1%|          | 2/317 [00:02<06:35,  1.26s/it]INFO:root:Starting optimization ...\n",
      "  1%|          | 3/317 [00:03<07:25,  1.42s/it]INFO:root:Starting optimization ...\n",
      "  1%|▏         | 4/317 [00:06<08:58,  1.72s/it]INFO:root:Starting optimization ...\n",
      "  2%|▏         | 5/317 [00:09<12:20,  2.37s/it]INFO:root:Starting optimization ...\n",
      "  2%|▏         | 6/317 [00:12<13:32,  2.61s/it]INFO:root:Starting optimization ...\n",
      "  2%|▏         | 7/317 [00:14<12:47,  2.48s/it]INFO:root:Starting optimization ...\n",
      "  3%|▎         | 8/317 [00:16<12:12,  2.37s/it]INFO:root:Starting optimization ...\n",
      "  3%|▎         | 9/317 [00:19<12:01,  2.34s/it]INFO:root:Starting optimization ...\n",
      "  3%|▎         | 10/317 [00:20<09:50,  1.92s/it]INFO:root:Starting optimization ...\n",
      "  3%|▎         | 11/317 [00:21<08:04,  1.58s/it]INFO:root:Starting optimization ...\n",
      "  4%|▍         | 12/317 [00:21<07:03,  1.39s/it]INFO:root:Starting optimization ...\n",
      "  4%|▍         | 13/317 [00:22<05:37,  1.11s/it]INFO:root:Starting optimization ...\n",
      "  4%|▍         | 14/317 [00:23<04:50,  1.04it/s]INFO:root:Starting optimization ...\n",
      "  5%|▍         | 15/317 [00:23<04:27,  1.13it/s]INFO:root:Starting optimization ...\n",
      "  5%|▌         | 16/317 [00:24<04:32,  1.11it/s]INFO:root:Starting optimization ...\n",
      "  5%|▌         | 17/317 [00:25<04:19,  1.15it/s]INFO:root:Starting optimization ...\n",
      "  6%|▌         | 18/317 [00:26<04:23,  1.14it/s]INFO:root:Starting optimization ...\n",
      "  6%|▌         | 19/317 [00:27<04:51,  1.02it/s]INFO:root:Starting optimization ...\n",
      "  6%|▋         | 20/317 [00:29<06:01,  1.22s/it]INFO:root:Starting optimization ...\n",
      "  7%|▋         | 21/317 [00:30<06:25,  1.30s/it]INFO:root:Starting optimization ...\n",
      "  7%|▋         | 22/317 [00:32<06:31,  1.33s/it]INFO:root:Starting optimization ...\n",
      "  7%|▋         | 23/317 [00:34<07:11,  1.47s/it]INFO:root:Starting optimization ...\n",
      "  8%|▊         | 24/317 [00:35<07:23,  1.51s/it]INFO:root:Starting optimization ...\n",
      "  8%|▊         | 25/317 [00:37<07:22,  1.52s/it]INFO:root:Starting optimization ...\n",
      "  8%|▊         | 26/317 [00:39<08:17,  1.71s/it]INFO:root:Starting optimization ...\n",
      "  9%|▊         | 27/317 [00:41<08:24,  1.74s/it]INFO:root:Starting optimization ...\n",
      "  9%|▉         | 28/317 [00:42<07:45,  1.61s/it]INFO:root:Starting optimization ...\n",
      "  9%|▉         | 29/317 [00:43<07:13,  1.51s/it]INFO:root:Starting optimization ...\n",
      "  9%|▉         | 30/317 [00:45<06:50,  1.43s/it]INFO:root:Starting optimization ...\n",
      " 10%|▉         | 31/317 [00:46<06:58,  1.46s/it]INFO:root:Starting optimization ...\n",
      " 10%|█         | 32/317 [00:47<06:16,  1.32s/it]INFO:root:Starting optimization ...\n",
      " 10%|█         | 33/317 [00:49<06:36,  1.40s/it]INFO:root:Starting optimization ...\n",
      " 11%|█         | 34/317 [00:50<06:08,  1.30s/it]INFO:root:Starting optimization ...\n",
      " 11%|█         | 35/317 [00:51<05:34,  1.19s/it]INFO:root:Starting optimization ...\n",
      " 11%|█▏        | 36/317 [00:52<05:25,  1.16s/it]INFO:root:Starting optimization ...\n",
      " 12%|█▏        | 37/317 [00:53<05:08,  1.10s/it]INFO:root:Starting optimization ...\n",
      " 12%|█▏        | 38/317 [00:54<04:58,  1.07s/it]INFO:root:Starting optimization ...\n",
      " 12%|█▏        | 39/317 [00:55<04:45,  1.03s/it]INFO:root:Starting optimization ...\n",
      " 13%|█▎        | 40/317 [00:55<04:30,  1.02it/s]INFO:root:Starting optimization ...\n",
      " 13%|█▎        | 41/317 [00:56<03:56,  1.17it/s]INFO:root:Starting optimization ...\n",
      " 13%|█▎        | 42/317 [00:57<03:54,  1.17it/s]INFO:root:Starting optimization ...\n",
      " 14%|█▎        | 43/317 [00:58<03:35,  1.27it/s]INFO:root:Starting optimization ...\n",
      " 14%|█▍        | 44/317 [00:58<03:32,  1.28it/s]INFO:root:Starting optimization ...\n",
      " 14%|█▍        | 45/317 [00:59<03:12,  1.41it/s]INFO:root:Starting optimization ...\n",
      " 15%|█▍        | 46/317 [00:59<03:04,  1.47it/s]INFO:root:Starting optimization ...\n",
      " 15%|█▍        | 47/317 [01:00<02:55,  1.54it/s]INFO:root:Starting optimization ...\n",
      " 15%|█▌        | 48/317 [01:01<02:59,  1.50it/s]INFO:root:Starting optimization ...\n",
      " 15%|█▌        | 49/317 [01:02<03:11,  1.40it/s]INFO:root:Starting optimization ...\n",
      " 16%|█▌        | 50/317 [01:02<03:14,  1.37it/s]INFO:root:Starting optimization ...\n",
      " 16%|█▌        | 51/317 [01:03<03:21,  1.32it/s]INFO:root:Starting optimization ...\n",
      " 16%|█▋        | 52/317 [01:04<03:28,  1.27it/s]INFO:root:Starting optimization ...\n",
      " 17%|█▋        | 53/317 [01:05<03:40,  1.20it/s]INFO:root:Starting optimization ...\n",
      " 17%|█▋        | 54/317 [01:06<03:44,  1.17it/s]INFO:root:Starting optimization ...\n",
      " 17%|█▋        | 55/317 [01:07<03:49,  1.14it/s]INFO:root:Starting optimization ...\n",
      " 18%|█▊        | 56/317 [01:08<04:03,  1.07it/s]INFO:root:Starting optimization ...\n",
      " 18%|█▊        | 57/317 [01:09<04:31,  1.05s/it]INFO:root:Starting optimization ...\n",
      " 18%|█▊        | 58/317 [01:11<04:56,  1.14s/it]INFO:root:Starting optimization ...\n",
      " 19%|█▊        | 59/317 [01:12<04:56,  1.15s/it]INFO:root:Starting optimization ...\n",
      " 19%|█▉        | 60/317 [01:13<04:50,  1.13s/it]INFO:root:Starting optimization ...\n",
      " 19%|█▉        | 61/317 [01:14<04:29,  1.05s/it]INFO:root:Starting optimization ...\n",
      " 20%|█▉        | 62/317 [01:14<04:06,  1.03it/s]INFO:root:Starting optimization ...\n",
      " 20%|█▉        | 63/317 [01:15<04:08,  1.02it/s]INFO:root:Starting optimization ...\n",
      " 20%|██        | 64/317 [01:16<03:51,  1.09it/s]INFO:root:Starting optimization ...\n",
      " 21%|██        | 65/317 [01:17<04:00,  1.05it/s]INFO:root:Starting optimization ...\n",
      " 21%|██        | 66/317 [01:19<04:50,  1.16s/it]INFO:root:Starting optimization ...\n",
      " 21%|██        | 67/317 [01:20<05:15,  1.26s/it]INFO:root:Starting optimization ...\n",
      " 21%|██▏       | 68/317 [01:22<06:00,  1.45s/it]INFO:root:Starting optimization ...\n",
      " 22%|██▏       | 69/317 [01:24<06:32,  1.58s/it]INFO:root:Starting optimization ...\n",
      " 22%|██▏       | 70/317 [01:26<06:48,  1.66s/it]INFO:root:Starting optimization ...\n",
      " 22%|██▏       | 71/317 [01:28<06:54,  1.68s/it]INFO:root:Starting optimization ...\n",
      " 23%|██▎       | 72/317 [01:30<07:41,  1.88s/it]INFO:root:Starting optimization ...\n",
      " 23%|██▎       | 73/317 [01:32<07:54,  1.94s/it]INFO:root:Starting optimization ...\n",
      " 23%|██▎       | 74/317 [01:34<08:00,  1.98s/it]INFO:root:Starting optimization ...\n",
      " 24%|██▎       | 75/317 [01:36<07:52,  1.95s/it]INFO:root:Starting optimization ...\n",
      " 24%|██▍       | 76/317 [01:38<07:54,  1.97s/it]INFO:root:Starting optimization ...\n",
      " 24%|██▍       | 77/317 [01:40<07:57,  1.99s/it]INFO:root:Starting optimization ...\n",
      " 25%|██▍       | 78/317 [01:43<08:44,  2.19s/it]INFO:root:Starting optimization ...\n",
      " 25%|██▍       | 79/317 [01:44<07:47,  1.97s/it]INFO:root:Starting optimization ...\n",
      " 25%|██▌       | 80/317 [01:46<07:56,  2.01s/it]INFO:root:Starting optimization ...\n",
      " 26%|██▌       | 81/317 [01:48<07:34,  1.93s/it]INFO:root:Starting optimization ...\n",
      " 26%|██▌       | 82/317 [01:50<07:51,  2.01s/it]INFO:root:Starting optimization ...\n",
      " 26%|██▌       | 83/317 [01:51<06:49,  1.75s/it]INFO:root:Starting optimization ...\n",
      " 26%|██▋       | 84/317 [01:52<05:29,  1.41s/it]INFO:root:Starting optimization ...\n",
      " 27%|██▋       | 85/317 [01:54<05:30,  1.42s/it]INFO:root:Starting optimization ...\n",
      " 27%|██▋       | 86/317 [01:56<07:06,  1.85s/it]INFO:root:Starting optimization ...\n",
      " 27%|██▋       | 87/317 [01:58<07:09,  1.87s/it]INFO:root:Starting optimization ...\n",
      " 28%|██▊       | 88/317 [02:00<06:56,  1.82s/it]INFO:root:Starting optimization ...\n",
      " 28%|██▊       | 89/317 [02:01<06:24,  1.69s/it]INFO:root:Starting optimization ...\n",
      " 28%|██▊       | 90/317 [02:03<05:53,  1.56s/it]INFO:root:Starting optimization ...\n",
      " 29%|██▊       | 91/317 [02:05<06:25,  1.71s/it]INFO:root:Starting optimization ...\n",
      " 29%|██▉       | 92/317 [02:07<06:51,  1.83s/it]INFO:root:Starting optimization ...\n",
      " 29%|██▉       | 93/317 [02:08<06:28,  1.73s/it]INFO:root:Starting optimization ...\n",
      " 30%|██▉       | 94/317 [02:12<08:55,  2.40s/it]INFO:root:Starting optimization ...\n",
      " 30%|██▉       | 95/317 [02:14<08:22,  2.26s/it]INFO:root:Starting optimization ...\n",
      " 30%|███       | 96/317 [02:16<07:35,  2.06s/it]INFO:root:Starting optimization ...\n",
      " 31%|███       | 97/317 [02:22<12:12,  3.33s/it]INFO:root:Starting optimization ...\n",
      " 31%|███       | 98/317 [02:24<10:51,  2.98s/it]INFO:root:Starting optimization ...\n",
      " 31%|███       | 99/317 [02:26<09:57,  2.74s/it]INFO:root:Starting optimization ...\n",
      " 32%|███▏      | 100/317 [02:29<09:17,  2.57s/it]INFO:root:Starting optimization ...\n",
      " 32%|███▏      | 101/317 [02:30<08:00,  2.23s/it]INFO:root:Starting optimization ...\n",
      " 32%|███▏      | 102/317 [02:32<07:38,  2.13s/it]INFO:root:Starting optimization ...\n",
      " 32%|███▏      | 103/317 [02:33<06:52,  1.93s/it]INFO:root:Starting optimization ...\n",
      " 33%|███▎      | 104/317 [02:35<06:45,  1.90s/it]INFO:root:Starting optimization ...\n",
      " 33%|███▎      | 105/317 [02:37<06:17,  1.78s/it]INFO:root:Starting optimization ...\n",
      " 33%|███▎      | 106/317 [02:38<05:40,  1.62s/it]INFO:root:Starting optimization ...\n",
      " 34%|███▍      | 107/317 [02:39<05:18,  1.52s/it]INFO:root:Starting optimization ...\n",
      " 34%|███▍      | 108/317 [02:40<04:49,  1.38s/it]INFO:root:Starting optimization ...\n",
      " 34%|███▍      | 109/317 [02:41<04:19,  1.25s/it]INFO:root:Starting optimization ...\n",
      " 35%|███▍      | 110/317 [02:42<04:08,  1.20s/it]INFO:root:Starting optimization ...\n",
      " 35%|███▌      | 111/317 [02:44<04:11,  1.22s/it]INFO:root:Starting optimization ...\n",
      " 35%|███▌      | 112/317 [02:45<04:24,  1.29s/it]INFO:root:Starting optimization ...\n",
      " 36%|███▌      | 113/317 [02:47<04:59,  1.47s/it]INFO:root:Starting optimization ...\n",
      " 36%|███▌      | 114/317 [02:49<05:22,  1.59s/it]INFO:root:Starting optimization ...\n",
      " 36%|███▋      | 115/317 [02:51<05:53,  1.75s/it]INFO:root:Starting optimization ...\n",
      " 37%|███▋      | 116/317 [02:53<06:06,  1.82s/it]INFO:root:Starting optimization ...\n",
      " 37%|███▋      | 117/317 [02:55<06:10,  1.85s/it]INFO:root:Starting optimization ...\n",
      " 37%|███▋      | 118/317 [02:57<06:15,  1.89s/it]INFO:root:Starting optimization ...\n",
      " 38%|███▊      | 119/317 [02:59<06:08,  1.86s/it]INFO:root:Starting optimization ...\n",
      " 38%|███▊      | 120/317 [03:01<06:23,  1.95s/it]INFO:root:Starting optimization ...\n",
      " 38%|███▊      | 121/317 [03:03<06:24,  1.96s/it]INFO:root:Starting optimization ...\n",
      " 38%|███▊      | 122/317 [03:05<06:30,  2.00s/it]INFO:root:Starting optimization ...\n",
      " 39%|███▉      | 123/317 [03:07<06:17,  1.94s/it]INFO:root:Starting optimization ...\n",
      " 39%|███▉      | 124/317 [03:09<06:46,  2.11s/it]INFO:root:Starting optimization ...\n",
      " 39%|███▉      | 125/317 [03:11<06:44,  2.11s/it]INFO:root:Starting optimization ...\n",
      " 40%|███▉      | 126/317 [03:14<07:22,  2.32s/it]INFO:root:Starting optimization ...\n",
      " 40%|████      | 127/317 [03:17<07:39,  2.42s/it]INFO:root:Starting optimization ...\n",
      " 40%|████      | 128/317 [03:19<07:44,  2.46s/it]INFO:root:Starting optimization ...\n",
      " 41%|████      | 129/317 [03:22<08:01,  2.56s/it]INFO:root:Starting optimization ...\n",
      " 41%|████      | 130/317 [03:25<07:58,  2.56s/it]INFO:root:Starting optimization ...\n",
      " 41%|████▏     | 131/317 [03:28<08:44,  2.82s/it]INFO:root:Starting optimization ...\n",
      " 42%|████▏     | 132/317 [03:32<10:10,  3.30s/it]INFO:root:Starting optimization ...\n",
      " 42%|████▏     | 133/317 [03:35<09:27,  3.09s/it]INFO:root:Starting optimization ...\n",
      " 42%|████▏     | 134/317 [03:39<10:06,  3.32s/it]INFO:root:Starting optimization ...\n",
      " 43%|████▎     | 135/317 [03:42<09:54,  3.27s/it]INFO:root:Starting optimization ...\n",
      " 43%|████▎     | 136/317 [03:46<10:38,  3.53s/it]INFO:root:Starting optimization ...\n",
      " 43%|████▎     | 137/317 [03:49<10:06,  3.37s/it]INFO:root:Starting optimization ...\n",
      " 44%|████▎     | 138/317 [03:53<10:51,  3.64s/it]INFO:root:Starting optimization ...\n",
      " 44%|████▍     | 139/317 [03:58<11:20,  3.83s/it]INFO:root:Starting optimization ...\n",
      " 44%|████▍     | 140/317 [04:03<12:55,  4.38s/it]INFO:root:Starting optimization ...\n",
      " 44%|████▍     | 141/317 [04:08<13:12,  4.50s/it]INFO:root:Starting optimization ...\n",
      " 45%|████▍     | 142/317 [04:10<10:35,  3.63s/it]INFO:root:Starting optimization ...\n",
      " 45%|████▌     | 143/317 [04:11<08:27,  2.92s/it]INFO:root:Starting optimization ...\n",
      " 45%|████▌     | 144/317 [04:12<06:52,  2.38s/it]INFO:root:Starting optimization ...\n",
      " 46%|████▌     | 145/317 [04:13<05:34,  1.95s/it]INFO:root:Starting optimization ...\n",
      " 46%|████▌     | 146/317 [04:14<04:51,  1.70s/it]INFO:root:Starting optimization ...\n",
      " 46%|████▋     | 147/317 [04:15<04:13,  1.49s/it]INFO:root:Starting optimization ...\n",
      " 47%|████▋     | 148/317 [04:17<04:02,  1.44s/it]INFO:root:Starting optimization ...\n",
      " 47%|████▋     | 149/317 [04:18<03:43,  1.33s/it]INFO:root:Starting optimization ...\n",
      " 47%|████▋     | 150/317 [04:19<03:32,  1.27s/it]INFO:root:Starting optimization ...\n",
      " 48%|████▊     | 151/317 [04:20<03:17,  1.19s/it]INFO:root:Starting optimization ...\n",
      " 48%|████▊     | 152/317 [04:21<03:21,  1.22s/it]INFO:root:Starting optimization ...\n",
      " 48%|████▊     | 153/317 [04:22<03:17,  1.20s/it]INFO:root:Starting optimization ...\n",
      " 49%|████▊     | 154/317 [04:23<03:16,  1.21s/it]INFO:root:Starting optimization ...\n",
      " 49%|████▉     | 155/317 [04:25<03:38,  1.35s/it]INFO:root:Starting optimization ...\n",
      " 49%|████▉     | 156/317 [04:26<03:31,  1.32s/it]INFO:root:Starting optimization ...\n",
      " 50%|████▉     | 157/317 [04:28<03:31,  1.32s/it]INFO:root:Starting optimization ...\n",
      " 50%|████▉     | 158/317 [04:29<03:31,  1.33s/it]INFO:root:Starting optimization ...\n",
      " 50%|█████     | 159/317 [04:30<03:28,  1.32s/it]INFO:root:Starting optimization ...\n",
      " 50%|█████     | 160/317 [04:31<03:12,  1.23s/it]INFO:root:Starting optimization ...\n",
      " 51%|█████     | 161/317 [04:33<03:16,  1.26s/it]INFO:root:Starting optimization ...\n",
      " 51%|█████     | 162/317 [04:34<03:18,  1.28s/it]INFO:root:Starting optimization ...\n",
      " 51%|█████▏    | 163/317 [04:35<03:05,  1.20s/it]INFO:root:Starting optimization ...\n",
      " 52%|█████▏    | 164/317 [04:36<02:54,  1.14s/it]INFO:root:Starting optimization ...\n",
      " 52%|█████▏    | 165/317 [04:37<02:43,  1.08s/it]INFO:root:Starting optimization ...\n",
      " 52%|█████▏    | 166/317 [04:38<02:31,  1.00s/it]INFO:root:Starting optimization ...\n",
      " 53%|█████▎    | 167/317 [04:38<02:15,  1.11it/s]INFO:root:Starting optimization ...\n",
      " 53%|█████▎    | 168/317 [04:39<02:02,  1.22it/s]INFO:root:Starting optimization ...\n",
      " 53%|█████▎    | 169/317 [04:40<01:51,  1.33it/s]INFO:root:Starting optimization ...\n",
      " 54%|█████▎    | 170/317 [04:40<01:48,  1.35it/s]INFO:root:Starting optimization ...\n",
      " 54%|█████▍    | 171/317 [04:41<01:48,  1.35it/s]INFO:root:Starting optimization ...\n",
      " 54%|█████▍    | 172/317 [04:42<01:51,  1.30it/s]INFO:root:Starting optimization ...\n",
      " 55%|█████▍    | 173/317 [04:43<01:48,  1.32it/s]INFO:root:Starting optimization ...\n",
      " 55%|█████▍    | 174/317 [04:44<01:58,  1.21it/s]INFO:root:Starting optimization ...\n",
      " 55%|█████▌    | 175/317 [04:45<02:06,  1.12it/s]INFO:root:Starting optimization ...\n",
      " 56%|█████▌    | 176/317 [04:45<01:55,  1.22it/s]INFO:root:Starting optimization ...\n",
      " 56%|█████▌    | 177/317 [04:46<02:01,  1.16it/s]INFO:root:Starting optimization ...\n",
      " 56%|█████▌    | 178/317 [04:47<01:54,  1.21it/s]INFO:root:Starting optimization ...\n",
      " 56%|█████▋    | 179/317 [04:48<01:43,  1.34it/s]INFO:root:Starting optimization ...\n",
      " 57%|█████▋    | 180/317 [04:48<01:45,  1.29it/s]INFO:root:Starting optimization ...\n",
      " 57%|█████▋    | 181/317 [04:49<01:43,  1.31it/s]INFO:root:Starting optimization ...\n",
      " 57%|█████▋    | 182/317 [04:50<01:59,  1.13it/s]INFO:root:Starting optimization ...\n",
      " 58%|█████▊    | 183/317 [04:51<01:58,  1.13it/s]INFO:root:Starting optimization ...\n",
      " 58%|█████▊    | 184/317 [04:52<02:01,  1.10it/s]INFO:root:Starting optimization ...\n",
      " 58%|█████▊    | 185/317 [04:53<02:03,  1.07it/s]INFO:root:Starting optimization ...\n",
      " 59%|█████▊    | 186/317 [04:54<02:07,  1.03it/s]INFO:root:Starting optimization ...\n",
      " 59%|█████▉    | 187/317 [04:55<02:04,  1.04it/s]INFO:root:Starting optimization ...\n",
      " 59%|█████▉    | 188/317 [04:56<02:01,  1.06it/s]INFO:root:Starting optimization ...\n",
      " 60%|█████▉    | 189/317 [04:57<02:11,  1.03s/it]INFO:root:Starting optimization ...\n",
      " 60%|█████▉    | 190/317 [04:59<02:17,  1.08s/it]INFO:root:Starting optimization ...\n",
      " 60%|██████    | 191/317 [05:00<02:21,  1.12s/it]INFO:root:Starting optimization ...\n",
      " 61%|██████    | 192/317 [05:01<02:30,  1.20s/it]INFO:root:Starting optimization ...\n",
      " 61%|██████    | 193/317 [05:02<02:23,  1.16s/it]INFO:root:Starting optimization ...\n",
      " 61%|██████    | 194/317 [05:04<02:35,  1.26s/it]INFO:root:Starting optimization ...\n",
      " 62%|██████▏   | 195/317 [05:05<02:32,  1.25s/it]INFO:root:Starting optimization ...\n",
      " 62%|██████▏   | 196/317 [05:06<02:24,  1.19s/it]INFO:root:Starting optimization ...\n",
      " 62%|██████▏   | 197/317 [05:07<02:19,  1.16s/it]INFO:root:Starting optimization ...\n",
      " 62%|██████▏   | 198/317 [05:08<02:17,  1.16s/it]INFO:root:Starting optimization ...\n",
      " 63%|██████▎   | 199/317 [05:09<02:20,  1.19s/it]INFO:root:Starting optimization ...\n",
      " 63%|██████▎   | 200/317 [05:10<02:11,  1.13s/it]INFO:root:Starting optimization ...\n",
      " 63%|██████▎   | 201/317 [05:12<02:13,  1.15s/it]INFO:root:Starting optimization ...\n",
      " 64%|██████▎   | 202/317 [05:13<02:07,  1.10s/it]INFO:root:Starting optimization ...\n",
      " 64%|██████▍   | 203/317 [05:14<02:02,  1.07s/it]INFO:root:Starting optimization ...\n",
      " 64%|██████▍   | 204/317 [05:14<01:51,  1.01it/s]INFO:root:Starting optimization ...\n",
      " 65%|██████▍   | 205/317 [05:15<01:43,  1.08it/s]INFO:root:Starting optimization ...\n",
      " 65%|██████▍   | 206/317 [05:16<01:40,  1.11it/s]INFO:root:Starting optimization ...\n",
      " 65%|██████▌   | 207/317 [05:17<01:34,  1.16it/s]INFO:root:Starting optimization ...\n",
      " 66%|██████▌   | 208/317 [05:18<01:34,  1.15it/s]INFO:root:Starting optimization ...\n",
      " 66%|██████▌   | 209/317 [05:18<01:29,  1.21it/s]INFO:root:Starting optimization ...\n",
      " 66%|██████▌   | 210/317 [05:19<01:30,  1.19it/s]INFO:root:Starting optimization ...\n",
      " 67%|██████▋   | 211/317 [05:20<01:18,  1.35it/s]INFO:root:Starting optimization ...\n",
      " 67%|██████▋   | 212/317 [05:20<01:07,  1.56it/s]INFO:root:Starting optimization ...\n",
      " 67%|██████▋   | 213/317 [05:21<01:06,  1.56it/s]INFO:root:Starting optimization ...\n",
      " 68%|██████▊   | 214/317 [05:21<00:58,  1.77it/s]INFO:root:Starting optimization ...\n",
      " 68%|██████▊   | 215/317 [05:22<01:05,  1.56it/s]INFO:root:Starting optimization ...\n",
      " 68%|██████▊   | 216/317 [05:23<01:06,  1.52it/s]INFO:root:Starting optimization ...\n",
      " 68%|██████▊   | 217/317 [05:23<01:07,  1.48it/s]INFO:root:Starting optimization ...\n",
      " 69%|██████▉   | 218/317 [05:24<01:02,  1.59it/s]INFO:root:Starting optimization ...\n",
      " 69%|██████▉   | 219/317 [05:24<00:56,  1.73it/s]INFO:root:Starting optimization ...\n",
      " 69%|██████▉   | 220/317 [05:25<00:52,  1.84it/s]INFO:root:Starting optimization ...\n",
      " 70%|██████▉   | 221/317 [05:25<00:51,  1.87it/s]INFO:root:Starting optimization ...\n",
      " 70%|███████   | 222/317 [05:26<00:49,  1.92it/s]INFO:root:Starting optimization ...\n",
      " 70%|███████   | 223/317 [05:27<00:52,  1.79it/s]INFO:root:Starting optimization ...\n",
      " 71%|███████   | 224/317 [05:27<00:49,  1.87it/s]INFO:root:Starting optimization ...\n",
      " 71%|███████   | 225/317 [05:28<00:47,  1.94it/s]INFO:root:Starting optimization ...\n",
      " 71%|███████▏  | 226/317 [05:28<00:54,  1.67it/s]INFO:root:Starting optimization ...\n",
      " 72%|███████▏  | 227/317 [05:29<00:56,  1.60it/s]INFO:root:Starting optimization ...\n",
      " 72%|███████▏  | 228/317 [05:30<00:59,  1.50it/s]INFO:root:Starting optimization ...\n",
      " 72%|███████▏  | 229/317 [05:31<01:00,  1.46it/s]INFO:root:Starting optimization ...\n",
      " 73%|███████▎  | 230/317 [05:31<01:00,  1.44it/s]INFO:root:Starting optimization ...\n",
      " 73%|███████▎  | 231/317 [05:32<01:01,  1.40it/s]INFO:root:Starting optimization ...\n",
      " 73%|███████▎  | 232/317 [05:33<01:02,  1.36it/s]INFO:root:Starting optimization ...\n",
      " 74%|███████▎  | 233/317 [05:33<01:00,  1.39it/s]INFO:root:Starting optimization ...\n",
      " 74%|███████▍  | 234/317 [05:34<01:04,  1.28it/s]INFO:root:Starting optimization ...\n",
      " 74%|███████▍  | 235/317 [05:35<01:05,  1.25it/s]INFO:root:Starting optimization ...\n",
      " 74%|███████▍  | 236/317 [05:36<00:58,  1.39it/s]INFO:root:Starting optimization ...\n",
      " 75%|███████▍  | 237/317 [05:37<00:58,  1.37it/s]INFO:root:Starting optimization ...\n",
      " 75%|███████▌  | 238/317 [05:37<01:02,  1.27it/s]INFO:root:Starting optimization ...\n",
      " 75%|███████▌  | 239/317 [05:38<01:06,  1.17it/s]INFO:root:Starting optimization ...\n",
      " 76%|███████▌  | 240/317 [05:39<01:06,  1.16it/s]INFO:root:Starting optimization ...\n",
      " 76%|███████▌  | 241/317 [05:40<01:09,  1.09it/s]INFO:root:Starting optimization ...\n",
      " 76%|███████▋  | 242/317 [05:41<01:03,  1.18it/s]INFO:root:Starting optimization ...\n",
      " 77%|███████▋  | 243/317 [05:42<01:01,  1.21it/s]INFO:root:Starting optimization ...\n",
      " 77%|███████▋  | 244/317 [05:43<01:00,  1.21it/s]INFO:root:Starting optimization ...\n",
      " 77%|███████▋  | 245/317 [05:44<01:07,  1.06it/s]INFO:root:Starting optimization ...\n",
      " 78%|███████▊  | 246/317 [05:45<01:04,  1.09it/s]INFO:root:Starting optimization ...\n",
      " 78%|███████▊  | 247/317 [05:46<01:20,  1.15s/it]INFO:root:Starting optimization ...\n",
      " 78%|███████▊  | 248/317 [05:48<01:18,  1.14s/it]INFO:root:Starting optimization ...\n",
      " 79%|███████▊  | 249/317 [05:49<01:24,  1.25s/it]INFO:root:Starting optimization ...\n",
      " 79%|███████▉  | 250/317 [05:51<01:34,  1.41s/it]INFO:root:Starting optimization ...\n",
      " 79%|███████▉  | 251/317 [05:52<01:33,  1.42s/it]INFO:root:Starting optimization ...\n",
      " 79%|███████▉  | 252/317 [05:54<01:35,  1.47s/it]INFO:root:Starting optimization ...\n",
      " 80%|███████▉  | 253/317 [05:56<01:55,  1.80s/it]INFO:root:Starting optimization ...\n",
      " 80%|████████  | 254/317 [05:58<01:43,  1.64s/it]INFO:root:Starting optimization ...\n",
      " 80%|████████  | 255/317 [06:00<01:51,  1.80s/it]INFO:root:Starting optimization ...\n",
      " 81%|████████  | 256/317 [06:03<02:20,  2.30s/it]INFO:root:Starting optimization ...\n",
      " 81%|████████  | 257/317 [06:05<02:12,  2.21s/it]INFO:root:Starting optimization ...\n",
      " 81%|████████▏ | 258/317 [06:08<02:13,  2.26s/it]INFO:root:Starting optimization ...\n",
      " 82%|████████▏ | 259/317 [06:10<02:13,  2.31s/it]INFO:root:Starting optimization ...\n",
      " 82%|████████▏ | 260/317 [06:12<02:03,  2.17s/it]INFO:root:Starting optimization ...\n",
      " 82%|████████▏ | 261/317 [06:14<02:01,  2.17s/it]INFO:root:Starting optimization ...\n",
      " 83%|████████▎ | 262/317 [06:15<01:44,  1.90s/it]INFO:root:Starting optimization ...\n",
      " 83%|████████▎ | 263/317 [06:17<01:42,  1.90s/it]INFO:root:Starting optimization ...\n",
      " 83%|████████▎ | 264/317 [06:19<01:32,  1.74s/it]INFO:root:Starting optimization ...\n",
      " 84%|████████▎ | 265/317 [06:20<01:24,  1.63s/it]INFO:root:Starting optimization ...\n",
      " 84%|████████▍ | 266/317 [06:22<01:25,  1.67s/it]INFO:root:Starting optimization ...\n",
      " 84%|████████▍ | 267/317 [06:24<01:37,  1.94s/it]INFO:root:Starting optimization ...\n",
      " 85%|████████▍ | 268/317 [06:27<01:50,  2.25s/it]INFO:root:Starting optimization ...\n",
      " 85%|████████▍ | 269/317 [06:30<01:59,  2.49s/it]INFO:root:Starting optimization ...\n",
      " 85%|████████▌ | 270/317 [06:34<02:15,  2.88s/it]INFO:root:Starting optimization ...\n",
      " 85%|████████▌ | 271/317 [06:36<02:03,  2.68s/it]INFO:root:Starting optimization ...\n",
      " 86%|████████▌ | 272/317 [06:39<01:56,  2.59s/it]INFO:root:Starting optimization ...\n",
      " 86%|████████▌ | 273/317 [06:40<01:37,  2.21s/it]INFO:root:Starting optimization ...\n",
      " 86%|████████▋ | 274/317 [06:41<01:22,  1.92s/it]INFO:root:Starting optimization ...\n",
      " 87%|████████▋ | 275/317 [06:43<01:15,  1.80s/it]INFO:root:Starting optimization ...\n",
      " 87%|████████▋ | 276/317 [06:44<01:05,  1.61s/it]INFO:root:Starting optimization ...\n",
      " 87%|████████▋ | 277/317 [06:45<01:02,  1.55s/it]INFO:root:Starting optimization ...\n",
      " 88%|████████▊ | 278/317 [06:47<01:00,  1.56s/it]INFO:root:Starting optimization ...\n",
      " 88%|████████▊ | 279/317 [06:49<00:59,  1.58s/it]INFO:root:Starting optimization ...\n",
      " 88%|████████▊ | 280/317 [06:50<00:58,  1.59s/it]INFO:root:Starting optimization ...\n",
      " 89%|████████▊ | 281/317 [06:52<00:53,  1.49s/it]INFO:root:Starting optimization ...\n",
      " 89%|████████▉ | 282/317 [06:53<00:51,  1.46s/it]INFO:root:Starting optimization ...\n",
      " 89%|████████▉ | 283/317 [06:55<00:50,  1.50s/it]INFO:root:Starting optimization ...\n",
      " 90%|████████▉ | 284/317 [06:56<00:52,  1.60s/it]INFO:root:Starting optimization ...\n",
      " 90%|████████▉ | 285/317 [06:58<00:53,  1.66s/it]INFO:root:Starting optimization ...\n",
      " 90%|█████████ | 286/317 [07:00<00:51,  1.65s/it]INFO:root:Starting optimization ...\n",
      " 91%|█████████ | 287/317 [07:03<01:00,  2.03s/it]INFO:root:Starting optimization ...\n",
      " 91%|█████████ | 288/317 [07:06<01:13,  2.52s/it]INFO:root:Starting optimization ...\n",
      " 91%|█████████ | 289/317 [07:08<01:04,  2.30s/it]INFO:root:Starting optimization ...\n",
      " 91%|█████████▏| 290/317 [07:10<01:00,  2.26s/it]INFO:root:Starting optimization ...\n",
      " 92%|█████████▏| 291/317 [07:12<00:57,  2.20s/it]INFO:root:Starting optimization ...\n",
      " 92%|█████████▏| 292/317 [07:15<00:55,  2.21s/it]INFO:root:Starting optimization ...\n",
      " 92%|█████████▏| 293/317 [07:17<00:56,  2.35s/it]INFO:root:Starting optimization ...\n",
      " 93%|█████████▎| 294/317 [07:20<00:57,  2.51s/it]INFO:root:Starting optimization ...\n",
      " 93%|█████████▎| 295/317 [07:23<00:54,  2.47s/it]INFO:root:Starting optimization ...\n",
      " 93%|█████████▎| 296/317 [07:25<00:51,  2.45s/it]INFO:root:Starting optimization ...\n",
      " 94%|█████████▎| 297/317 [07:28<00:49,  2.50s/it]INFO:root:Starting optimization ...\n",
      " 94%|█████████▍| 298/317 [07:30<00:48,  2.55s/it]INFO:root:Starting optimization ...\n",
      " 94%|█████████▍| 299/317 [07:33<00:48,  2.70s/it]INFO:root:Starting optimization ...\n",
      " 95%|█████████▍| 300/317 [07:36<00:45,  2.69s/it]INFO:root:Starting optimization ...\n",
      " 95%|█████████▍| 301/317 [07:39<00:44,  2.79s/it]INFO:root:Starting optimization ...\n",
      " 95%|█████████▌| 302/317 [07:41<00:39,  2.66s/it]INFO:root:Starting optimization ...\n",
      " 96%|█████████▌| 303/317 [07:44<00:36,  2.59s/it]INFO:root:Starting optimization ...\n",
      " 96%|█████████▌| 304/317 [07:46<00:31,  2.45s/it]INFO:root:Starting optimization ...\n",
      " 96%|█████████▌| 305/317 [07:48<00:29,  2.43s/it]INFO:root:Starting optimization ...\n",
      " 97%|█████████▋| 306/317 [07:50<00:24,  2.21s/it]INFO:root:Starting optimization ...\n",
      " 97%|█████████▋| 307/317 [07:52<00:20,  2.08s/it]INFO:root:Starting optimization ...\n",
      " 97%|█████████▋| 308/317 [07:53<00:17,  1.98s/it]INFO:root:Starting optimization ...\n",
      " 97%|█████████▋| 309/317 [07:55<00:15,  1.91s/it]INFO:root:Starting optimization ...\n",
      " 98%|█████████▊| 310/317 [07:57<00:12,  1.81s/it]INFO:root:Starting optimization ...\n",
      " 98%|█████████▊| 311/317 [07:58<00:10,  1.69s/it]INFO:root:Starting optimization ...\n",
      " 98%|█████████▊| 312/317 [08:00<00:08,  1.68s/it]INFO:root:Starting optimization ...\n",
      " 99%|█████████▊| 313/317 [08:01<00:06,  1.55s/it]INFO:root:Starting optimization ...\n",
      " 99%|█████████▉| 314/317 [08:03<00:04,  1.58s/it]INFO:root:Starting optimization ...\n",
      " 99%|█████████▉| 315/317 [08:04<00:03,  1.57s/it]INFO:root:Starting optimization ...\n",
      "100%|█████████▉| 316/317 [08:06<00:01,  1.53s/it]INFO:root:Starting optimization ...\n",
      "100%|██████████| 317/317 [08:07<00:00,  1.54s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_gene</th>\n",
       "      <th>egene</th>\n",
       "      <th>weight_col</th>\n",
       "      <th>pv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000206195</td>\n",
       "      <td>ENSG00000206195</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.468064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.498752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.308303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000131100</td>\n",
       "      <td>ENSG00000131100</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.207885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>ENSG00000205560</td>\n",
       "      <td>ENSG00000205560</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.583235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.274153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>ENSG00000205559</td>\n",
       "      <td>ENSG00000205559</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.502091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.429815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.391919</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         burden_gene            egene weight_col        pv\n",
       "0    ENSG00000206195  ENSG00000206195   maf_beta  1.000000\n",
       "1    ENSG00000177663  ENSG00000177663   maf_beta  0.468064\n",
       "2    ENSG00000069998  ENSG00000069998   maf_beta  0.498752\n",
       "3    ENSG00000093072  ENSG00000093072   maf_beta  0.308303\n",
       "4    ENSG00000131100  ENSG00000131100   maf_beta  0.207885\n",
       "..               ...              ...        ...       ...\n",
       "312  ENSG00000205560  ENSG00000205560   maf_beta  0.583235\n",
       "313  ENSG00000100288  ENSG00000100288   maf_beta  0.274153\n",
       "314  ENSG00000205559  ENSG00000205559   maf_beta  0.502091\n",
       "315  ENSG00000100299  ENSG00000100299   maf_beta  0.429815\n",
       "316  ENSG00000079974  ENSG00000079974   maf_beta  0.391919\n",
       "\n",
       "[317 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for gene, row in tqdm(dd.C.var.iterrows(), total=dd.shape[3]):\n",
    "    Y = gaussianize(dd.G.obsm[pb_gex_key][[gene]].values.astype(\"float\") + 1e-5 * np.random.randn(dd.shape[0], 1))\n",
    "\n",
    "    start = max(0, row.start - cis_window)\n",
    "    end = row.end + cis_window\n",
    "    _G = dd.G[:, (dd.G.var.pos < end)]\n",
    "    _G = _G[:, (_G.var.pos > start)]\n",
    "    _G = _G[:, (_G.X.std(0) != 0)]\n",
    "\n",
    "    rdict = run_skat_test(_G, Y, F, gene)\n",
    "    results.append(rdict)\n",
    "\n",
    "rdf_skat = pd.DataFrame(results)\n",
    "rdf_skat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "burden_gene",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "egene",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "weight_col",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "pv",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "5117051e-e37c-4a6f-8453-406d1279504a",
       "rows": [
        [
         "0",
         "ENSG00000128322",
         "ENSG00000128322",
         "maf_beta",
         "0.34263064984649105"
        ],
        [
         "1",
         "ENSG00000093072",
         "ENSG00000093072",
         "maf_beta",
         "0.9772793701482388"
        ],
        [
         "2",
         "ENSG00000261202",
         "ENSG00000261202",
         "maf_beta",
         "0.9676151128040275"
        ],
        [
         "3",
         "ENSG00000183785",
         "ENSG00000183785",
         "maf_beta",
         "0.9996911458193686"
        ],
        [
         "4",
         "ENSG00000185721",
         "ENSG00000185721",
         "maf_beta",
         "0.33241222238226076"
        ],
        [
         "5",
         "ENSG00000241878",
         "ENSG00000241878",
         "maf_beta",
         "0.14654523868126224"
        ],
        [
         "6",
         "ENSG00000287817",
         "ENSG00000287817",
         "maf_beta",
         "0.009897743084311905"
        ],
        [
         "7",
         "ENSG00000100138",
         "ENSG00000100138",
         "maf_beta",
         "0.21398003576746194"
        ],
        [
         "8",
         "ENSG00000213888",
         "ENSG00000213888",
         "maf_beta",
         "0.058928192377704636"
        ],
        [
         "9",
         "ENSG00000270022",
         "ENSG00000270022",
         "maf_beta",
         "0.41299908821913545"
        ],
        [
         "10",
         "ENSG00000100106",
         "ENSG00000100106",
         "maf_beta",
         "0.4443510941692246"
        ],
        [
         "11",
         "ENSG00000286990",
         "ENSG00000286990",
         "maf_beta",
         "0.0363396897715329"
        ],
        [
         "12",
         "ENSG00000205593",
         "ENSG00000205593",
         "maf_beta",
         "0.4280847883219484"
        ],
        [
         "13",
         "ENSG00000253352",
         "ENSG00000253352",
         "maf_beta",
         "0.6178962174551161"
        ],
        [
         "14",
         "ENSG00000241484",
         "ENSG00000241484",
         "maf_beta",
         "0.5315383492499491"
        ],
        [
         "15",
         "ENSG00000100393",
         "ENSG00000100393",
         "maf_beta",
         "0.41374175923663437"
        ],
        [
         "16",
         "ENSG00000184470",
         "ENSG00000184470",
         "maf_beta",
         "0.1814180644347423"
        ],
        [
         "17",
         "ENSG00000128191",
         "ENSG00000128191",
         "maf_beta",
         "0.006046673854188556"
        ],
        [
         "18",
         "ENSG00000128271",
         "ENSG00000128271",
         "maf_beta",
         "0.9357026358361222"
        ],
        [
         "19",
         "ENSG00000215193",
         "ENSG00000215193",
         "maf_beta",
         "0.10966112495012614"
        ],
        [
         "20",
         "ENSG00000211679",
         "ENSG00000211679",
         "maf_beta",
         "0.3576079641540679"
        ],
        [
         "21",
         "ENSG00000100342",
         "ENSG00000100342",
         "maf_beta",
         "0.3221649008660922"
        ],
        [
         "22",
         "ENSG00000280383",
         "ENSG00000280383",
         "maf_beta",
         "0.8136845604877789"
        ],
        [
         "23",
         "ENSG00000177096",
         "ENSG00000177096",
         "maf_beta",
         "0.9846610555179294"
        ],
        [
         "24",
         "ENSG00000182858",
         "ENSG00000182858",
         "maf_beta",
         "0.6203719464580876"
        ],
        [
         "25",
         "ENSG00000100298",
         "ENSG00000100298",
         "maf_beta",
         "0.3596365389978715"
        ],
        [
         "26",
         "ENSG00000273032",
         "ENSG00000273032",
         "maf_beta",
         "0.8887710143996503"
        ],
        [
         "27",
         "ENSG00000185838",
         "ENSG00000185838",
         "maf_beta",
         "0.05383357079465001"
        ],
        [
         "28",
         "ENSG00000284194",
         "ENSG00000284194",
         "maf_beta",
         "0.6391962007858571"
        ],
        [
         "29",
         "ENSG00000211677",
         "ENSG00000211677",
         "maf_beta",
         "0.8637413576753483"
        ],
        [
         "30",
         "ENSG00000205559",
         "ENSG00000205559",
         "maf_beta",
         "0.22925715354293263"
        ],
        [
         "31",
         "ENSG00000100387",
         "ENSG00000100387",
         "maf_beta",
         "0.0815021510274564"
        ],
        [
         "32",
         "ENSG00000234928",
         "ENSG00000234928",
         "maf_beta",
         "0.9457490127084993"
        ],
        [
         "33",
         "ENSG00000100417",
         "ENSG00000100417",
         "maf_beta",
         "0.18477983080513505"
        ],
        [
         "34",
         "ENSG00000184708",
         "ENSG00000184708",
         "maf_beta",
         "0.17315826340644513"
        ],
        [
         "35",
         "ENSG00000100372",
         "ENSG00000100372",
         "maf_beta",
         "0.7321534663924815"
        ],
        [
         "36",
         "ENSG00000184949",
         "ENSG00000184949",
         "maf_beta",
         "0.10126474611464897"
        ],
        [
         "37",
         "ENSG00000100207",
         "ENSG00000100207",
         "maf_beta",
         "0.5948335056405235"
        ],
        [
         "38",
         "ENSG00000167074",
         "ENSG00000167074",
         "maf_beta",
         "0.004521402063233149"
        ],
        [
         "39",
         "ENSG00000100154",
         "ENSG00000100154",
         "maf_beta",
         "0.0001950604523666133"
        ],
        [
         "40",
         "ENSG00000100294",
         "ENSG00000100294",
         "maf_beta",
         "0.5112223522961837"
        ],
        [
         "41",
         "ENSG00000182541",
         "ENSG00000182541",
         "maf_beta",
         "0.7533220769525311"
        ],
        [
         "42",
         "ENSG00000100290",
         "ENSG00000100290",
         "maf_beta",
         "0.17355495144331878"
        ],
        [
         "43",
         "ENSG00000100034",
         "ENSG00000100034",
         "maf_beta",
         "0.2442196304086388"
        ],
        [
         "44",
         "ENSG00000240972",
         "ENSG00000240972",
         "maf_beta",
         "0.0015366716432209104"
        ],
        [
         "45",
         "ENSG00000100271",
         "ENSG00000100271",
         "maf_beta",
         "0.14558232427575124"
        ],
        [
         "46",
         "ENSG00000100253",
         "ENSG00000100253",
         "maf_beta",
         "0.9623897867866679"
        ],
        [
         "47",
         "ENSG00000138964",
         "ENSG00000138964",
         "maf_beta",
         "0.0393399938360236"
        ],
        [
         "48",
         "ENSG00000279278",
         "ENSG00000279278",
         "maf_beta",
         "0.9806918659036726"
        ],
        [
         "49",
         "ENSG00000100239",
         "ENSG00000100239",
         "maf_beta",
         "0.6876782353571935"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 413
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_gene</th>\n",
       "      <th>egene</th>\n",
       "      <th>weight_col</th>\n",
       "      <th>pv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000128322</td>\n",
       "      <td>ENSG00000128322</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.342631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.977279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000261202</td>\n",
       "      <td>ENSG00000261202</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.967615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000183785</td>\n",
       "      <td>ENSG00000183785</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.999691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000185721</td>\n",
       "      <td>ENSG00000185721</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.332412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENSG00000186716</td>\n",
       "      <td>ENSG00000186716</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.604184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENSG00000198355</td>\n",
       "      <td>ENSG00000198355</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.971767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>ENSG00000273096</td>\n",
       "      <td>ENSG00000273096</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.235938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>ENSG00000100292</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.445208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>ENSG00000099968</td>\n",
       "      <td>ENSG00000099968</td>\n",
       "      <td>maf_beta</td>\n",
       "      <td>0.344851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>413 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         burden_gene            egene weight_col        pv\n",
       "0    ENSG00000128322  ENSG00000128322   maf_beta  0.342631\n",
       "1    ENSG00000093072  ENSG00000093072   maf_beta  0.977279\n",
       "2    ENSG00000261202  ENSG00000261202   maf_beta  0.967615\n",
       "3    ENSG00000183785  ENSG00000183785   maf_beta  0.999691\n",
       "4    ENSG00000185721  ENSG00000185721   maf_beta  0.332412\n",
       "..               ...              ...        ...       ...\n",
       "408  ENSG00000186716  ENSG00000186716   maf_beta  0.604184\n",
       "409  ENSG00000198355  ENSG00000198355   maf_beta  0.971767\n",
       "410  ENSG00000273096  ENSG00000273096   maf_beta  0.235938\n",
       "411  ENSG00000100292  ENSG00000100292   maf_beta  0.445208\n",
       "412  ENSG00000099968  ENSG00000099968   maf_beta  0.344851\n",
       "\n",
       "[413 rows x 4 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "single_cell_base6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
