{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/nasif12/home_if12/l_mai/.conda/envs/mm_scgenetics/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "from pathlib import Path\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import cellink as cl\n",
    "from cellink.tl._burden_testing import *\n",
    "from scipy.stats import beta\n",
    "import pickle\n",
    "import numpy as np\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mai's test notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test only with chr22 data \n",
    "base_data_dir = Path(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/\")\n",
    "scdata_path = base_data_dir / \"input_data/chr22_OneK1K_cohort_gene_expression_matrix_14_celltypes_w_gene_locations.h5ad.gz\"\n",
    "gdata_dir = \"/data/ceph/hdd/project/node_09/sys_gen_students/2024_2025/project04_rare_variant_sc/input_data/filter_vcf_r08/\"\n",
    "DNA_LM_upstream = base_data_dir/ \"input_data/annotations/onek1k_inf_scores_upstream_model.tsv\"\n",
    "DNA_LM_downstream = base_data_dir/ \"input_data/annotations/onek1k_inf_scores_downstream_model.tsv\"\n",
    "vep_scores = base_data_dir/ \"input_data/annotations/onek1k1_chr22_variants_annotated_vep.txt\"\n",
    "\n",
    "zarr_file = os.path.join(gdata_dir, f\"chr22.dose.filtered.R2_0.8.vcz\")\n",
    "#eigenvec = pd.read_csv(base_data_dir / \"input_data/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned.eigenvec\", sep = ' ')\n",
    "scdata = sc.read_h5ad(scdata_path)\n",
    "gdata = cl.io.read_sgkit_zarr(zarr_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-01-19 18:27:20,439] INFO:cellink.tl._annotate_snps_genotype_data: renaming id column #Uploaded_variation into snp_id\n",
      "[2025-01-19 18:27:20,606] INFO:cellink.tl._annotate_snps_genotype_data: Subsetting annotations to variants that are in gdata\n",
      "[2025-01-19 18:27:20,606] INFO:cellink.tl._annotate_snps_genotype_data: 0 with missing annotation\n",
      "[2025-01-19 18:27:21,194] INFO:cellink.tl._annotate_snps_genotype_data: Index(['snp_id', 'Location', 'Allele', 'Gene', 'Feature', 'Feature_type',\n",
      "       'Consequence', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE',\n",
      "       'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen',\n",
      "       'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF',\n",
      "       'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF',\n",
      "       'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'CADD_PHRED',\n",
      "       'CADD_RAW', 'TSSDistance'],\n",
      "      dtype='object')\n",
      "[2025-01-19 18:27:21,558] INFO:cellink.tl.utils: Exploding column Consequence\n",
      "[2025-01-19 18:27:22,364] INFO:cellink.tl.utils: Number of original rows: 182552.\n",
      " Number of exploded rows: 200116\n",
      "[2025-01-19 18:27:22,388] INFO:cellink.tl.utils: Making dummies from column Consequence\n",
      "[2025-01-19 18:27:22,717] INFO:cellink.tl._annotate_snps_genotype_data: Changing dtype of categorical columns ['snp_id', 'Location', 'Allele', 'Gene', 'Feature', 'Feature_type', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen', 'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF', 'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'TSSDistance']\n",
      "[2025-01-19 18:27:23,047] INFO:cellink.tl._annotate_snps_genotype_data: Expanding annotations from multiple contexts per variant\n",
      "[2025-01-19 18:27:23,048] INFO:cellink.tl._annotate_snps_genotype_data: getting unique counts\n",
      "Columns with multiple values per variant: Index(['Gene', 'Feature', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS',\n",
      "       'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen', 'TSSDistance',\n",
      "       'Consequence_3_prime_UTR_variant', 'Consequence_5_prime_UTR_variant',\n",
      "       'Consequence_NMD_transcript_variant',\n",
      "       'Consequence_coding_sequence_variant',\n",
      "       'Consequence_downstream_gene_variant',\n",
      "       'Consequence_incomplete_terminal_codon_variant',\n",
      "       'Consequence_intron_variant', 'Consequence_mature_miRNA_variant',\n",
      "       'Consequence_missense_variant',\n",
      "       'Consequence_non_coding_transcript_exon_variant',\n",
      "       'Consequence_non_coding_transcript_variant',\n",
      "       'Consequence_splice_acceptor_variant',\n",
      "       'Consequence_splice_donor_5th_base_variant',\n",
      "       'Consequence_splice_donor_region_variant',\n",
      "       'Consequence_splice_donor_variant',\n",
      "       'Consequence_splice_polypyrimidine_tract_variant',\n",
      "       'Consequence_splice_region_variant', 'Consequence_stop_gained',\n",
      "       'Consequence_stop_retained_variant', 'Consequence_synonymous_variant',\n",
      "       'Consequence_upstream_gene_variant'],\n",
      "      dtype='object')\n",
      "[2025-01-19 18:31:26,007] INFO:cellink.tl._annotate_snps_genotype_data: Maximum number of distinct annotations per variant 8\n",
      "[2025-01-19 18:32:56,766] INFO:cellink.tl._annotate_snps_genotype_data: merging non-expanded annos into first data frame\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1034 × 143083\n",
       "    var: 'chrom', 'pos', 'a0', 'a1', 'AF', 'ER2', 'maf', 'R2', 'contig', 'id', 'id_mask', 'quality'\n",
       "    varm: 'annotations_0', 'annotations_1', 'annotations_2', 'annotations_3', 'annotations_4', 'annotations_5', 'annotations_6', 'annotations_7'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.tl.add_vep_annos_to_gdata(vep_scores, gdata,\n",
    "                             cols_to_explode=[\"Consequence\"],\n",
    "                             cols_to_dummy=[\"Consequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_snp_id(DNA_LM):\n",
    "    DNA_LM['snp_id'] = DNA_LM['Chromosome'] + \"_\" + DNA_LM['pos'].astype(str) + \"_\" + DNA_LM['ref'] + \"_\" + DNA_LM['alt']\n",
    "    DNA_LM['snp_id'] = DNA_LM['snp_id'].str.replace('chr', '')\n",
    "    \n",
    "    # Set 'snap_id' as the index\n",
    "    DNA_LM.set_index('snp_id', inplace=True)\n",
    "    return DNA_LM\n",
    "\n",
    "def reverse_and_update_snp_ids(gdata_df, dna_df):\n",
    "    updated_index = []\n",
    "    \n",
    "    for snp_id in dna_df.index:\n",
    "        if snp_id in gdata_df.index:\n",
    "            updated_index.append(snp_id)\n",
    "        else:\n",
    "            chrom, pos, ref, alt = snp_id.split(\"_\")\n",
    "            reversed_snp_id = f\"{chrom}_{pos}_{alt}_{ref}\"  # Reverse ref and alt\n",
    "            \n",
    "            # Check if reversed_snp_id exists in data_df\n",
    "            if reversed_snp_id in gdata_df.index:\n",
    "                #print(f\"Reversing {snp_id} to {reversed_snp_id}\")\n",
    "                updated_index.append(reversed_snp_id)\n",
    "            else:\n",
    "                print(f\"Error, unknown snp_id {snp_id}\")\n",
    "                updated_index.append(snp_id)\n",
    "    \n",
    "    # Update DNA_LM's index\n",
    "    dna_df.index = updated_index\n",
    "    print(\"\\nUpdated DNA_LM index:\")\n",
    "    print(dna_df.index)\n",
    "    return dna_df\n",
    "\n",
    "def add_maf_annotation(gdata):\n",
    "    weighted_snp_maf = beta.pdf(gdata.var[\"maf\"], 1, 25)\n",
    "    gdata.varm[\"annotations_0\"][\"MAF_beta_1.25\"] = weighted_snp_maf\n",
    "    return gdata\n",
    "\n",
    "\n",
    "def add_DNA_LM(gdata, file, chromosome, colname):\n",
    "\n",
    "    DNA_LM = pd.read_csv(file,\n",
    "                sep = '\\t')\n",
    "    DNA_LM = add_snp_id(DNA_LM)\n",
    "    DNA_LM = reverse_and_update_snp_ids(gdata.varm[\"annotations_0\"],\n",
    "                                        DNA_LM[DNA_LM[\"Chromosome\"]==f\"chr{chromosome}\"])\n",
    "\n",
    "    gdata.varm[\"annotations_0\"][colname] = DNA_LM[\"influence_score\"].reindex(gdata.varm[\"annotations_0\"].index)\n",
    "\n",
    "    # Rename the merged column if needed\n",
    "    gdata.varm[\"annotations_0\"].rename(columns={\"influence_score\": colname}, inplace=True)\n",
    "    return gdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "add maf annotation to gdata \n",
      "add DNA_LM annotation to gdata \n",
      "\n",
      "Updated DNA_LM index:\n",
      "Index(['22_17069391_G_A', '22_17093288_C_T', '22_17102725_A_G',\n",
      "       '22_17115621_T_A', '22_17117130_G_A', '22_17128890_A_G',\n",
      "       '22_17202942_T_G', '22_17227050_G_T', '22_17274178_A_T',\n",
      "       '22_17276558_C_T',\n",
      "       ...\n",
      "       '22_51171783_C_T', '22_51172439_G_A', '22_51172880_G_A',\n",
      "       '22_51175039_A_C', '22_51175382_A_G', '22_51175798_C_T',\n",
      "       '22_51177291_C_T', '22_51180520_C_T', '22_51180534_A_G',\n",
      "       '22_51185359_G_A'],\n",
      "      dtype='object', length=46826)\n",
      "\n",
      "Updated DNA_LM index:\n",
      "Index(['22_17069391_G_A', '22_17093288_C_T', '22_17102725_A_G',\n",
      "       '22_17115621_T_A', '22_17117130_G_A', '22_17128890_A_G',\n",
      "       '22_17202942_T_G', '22_17227050_G_T', '22_17274178_A_T',\n",
      "       '22_17276558_C_T',\n",
      "       ...\n",
      "       '22_51171783_C_T', '22_51172439_G_A', '22_51172880_G_A',\n",
      "       '22_51175039_A_C', '22_51175382_A_G', '22_51175798_C_T',\n",
      "       '22_51177291_C_T', '22_51180520_C_T', '22_51180534_A_G',\n",
      "       '22_51185359_G_A'],\n",
      "      dtype='object', length=46826)\n"
     ]
    }
   ],
   "source": [
    "# add maf annotaion to gdata\n",
    "print(f\"add maf annotation to gdata \")\n",
    "gdata = add_maf_annotation(gdata)\n",
    "\n",
    "# add DNA_LM annotations (downstream and upstream models) to gdata \n",
    "print(f\"add DNA_LM annotation to gdata \")\n",
    "gdata = add_DNA_LM(gdata, file=DNA_LM_upstream, chromosome=\"22\", colname='DNA_LM_up')\n",
    "gdata = add_DNA_LM(gdata, file=DNA_LM_downstream, chromosome=\"22\", colname='DNA_LM_down')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>DNA_LM_mixed</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>9350.801</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>22.916096</td>\n",
       "      <td>43612.917035</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>10775.218</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>28.387657</td>\n",
       "      <td>52127.930292</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>8327.426</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>21.604692</td>\n",
       "      <td>39069.213569</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>9439.675</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>32.699742</td>\n",
       "      <td>44983.557785</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>9054.896</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>23.019223</td>\n",
       "      <td>42525.702648</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>9227.155</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>27.382034</td>\n",
       "      <td>42677.295555</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>ENSG00000267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>9411.798</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>25.363997</td>\n",
       "      <td>44142.005875</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>ENSG00000267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>9862.022</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>23.919521</td>\n",
       "      <td>44060.525919</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>ENSG00000267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>8392.390</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>24.546299</td>\n",
       "      <td>40866.836483</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>ENSG00000267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>9015.698</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>26.508232</td>\n",
       "      <td>42943.342890</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>ENSG00000267338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9810 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  DNA_LM_mixed  \\\n",
       "id                                                                           \n",
       "1_1          9350.801  13.224558    22.916096   43612.917035     13.224558   \n",
       "2_2         10775.218  16.520484    28.387657   52127.930292     16.520484   \n",
       "3_3          8327.426  12.240323    21.604692   39069.213569     12.240323   \n",
       "4_4          9439.675  18.346526    32.699742   44983.557785     18.346526   \n",
       "6_6          9054.896  13.340017    23.019223   42525.702648     13.340017   \n",
       "...               ...        ...          ...            ...           ...   \n",
       "1077_1078    9227.155  15.636985    27.382034   42677.295555     15.636985   \n",
       "1078_1079    9411.798  14.626046    25.363997   44142.005875     14.626046   \n",
       "1079_1080    9862.022  13.470969    23.919521   44060.525919     13.470969   \n",
       "1080_1081    8392.390  13.858525    24.546299   40866.836483     13.858525   \n",
       "1081_1082    9015.698  15.057642    26.508232   42943.342890     15.057642   \n",
       "\n",
       "                    Geneid  \n",
       "id                          \n",
       "1_1        ENSG00000233866  \n",
       "2_2        ENSG00000233866  \n",
       "3_3        ENSG00000233866  \n",
       "4_4        ENSG00000233866  \n",
       "6_6        ENSG00000233866  \n",
       "...                    ...  \n",
       "1077_1078  ENSG00000267338  \n",
       "1078_1079  ENSG00000267338  \n",
       "1079_1080  ENSG00000267338  \n",
       "1080_1081  ENSG00000267338  \n",
       "1081_1082  ENSG00000267338  \n",
       "\n",
       "[9810 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_res = pd.read_pickle(base_data_dir/\"output/chr22_computed_burdens_test_10_genes.pkl\")\n",
    "test_res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Onek1k chr 22 with gene location for test purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1272489 × 32738\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'pool', 'individual', 'percent.mt', 'latent', 'nCount_SCT', 'nFeature_SCT', 'cell_type', 'cell_label', 'sex', 'age'\n",
       "    var: 'GeneSymbol', 'features', 'chromosome', 'start', 'end'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scdata_full_path = base_data_dir / \"input_data/OneK1K_cohort_gene_expression_matrix_14_celltypes_w_gene_locations.h5ad.gz\"\n",
    "scdata = sc.read_h5ad(scdata_full_path)\n",
    "scdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 1272489 × 655\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'pool', 'individual', 'percent.mt', 'latent', 'nCount_SCT', 'nFeature_SCT', 'cell_type', 'cell_label', 'sex', 'age'\n",
       "    var: 'GeneSymbol', 'features', 'chromosome', 'start', 'end'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scdata_22 = scdata[:, scdata.var[\"chromosome\"] == \"22\"]\n",
    "scdata_22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>features</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000233866</th>\n",
       "      <td>LA16c-4G1.3</td>\n",
       "      <td>LA16c-4G1.3</td>\n",
       "      <td>22</td>\n",
       "      <td>15915800.0</td>\n",
       "      <td>15914721.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000225255</th>\n",
       "      <td>LA16c-83F12.6</td>\n",
       "      <td>LA16c-83F12.6</td>\n",
       "      <td>22</td>\n",
       "      <td>15741297.0</td>\n",
       "      <td>15779680.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198062</th>\n",
       "      <td>POTEH</td>\n",
       "      <td>POTEH</td>\n",
       "      <td>22</td>\n",
       "      <td>15690026.0</td>\n",
       "      <td>15721631.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000236666</th>\n",
       "      <td>POTEH-AS1</td>\n",
       "      <td>POTEH-AS1</td>\n",
       "      <td>22</td>\n",
       "      <td>15703403.0</td>\n",
       "      <td>15699361.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000230471</th>\n",
       "      <td>LA16c-2F2.8</td>\n",
       "      <td>LA16c-2F2.8</td>\n",
       "      <td>22</td>\n",
       "      <td>15625205.0</td>\n",
       "      <td>15557577.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251322</th>\n",
       "      <td>SHANK3</td>\n",
       "      <td>SHANK3</td>\n",
       "      <td>22</td>\n",
       "      <td>50674408.0</td>\n",
       "      <td>50733212.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000225929</th>\n",
       "      <td>AC000036.4</td>\n",
       "      <td>AC000036.4</td>\n",
       "      <td>22</td>\n",
       "      <td>50738173.0</td>\n",
       "      <td>50735813.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000100312</th>\n",
       "      <td>ACR</td>\n",
       "      <td>ACR</td>\n",
       "      <td>22</td>\n",
       "      <td>50738196.0</td>\n",
       "      <td>50745339.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000254499</th>\n",
       "      <td>AC002056.5</td>\n",
       "      <td>AC002056.5</td>\n",
       "      <td>22</td>\n",
       "      <td>50743520.0</td>\n",
       "      <td>50740593.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000079974</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>RABL2B</td>\n",
       "      <td>22</td>\n",
       "      <td>50783667.0</td>\n",
       "      <td>50767501.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>655 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GeneSymbol        features chromosome       start  \\\n",
       "Geneid                                                                   \n",
       "ENSG00000233866     LA16c-4G1.3     LA16c-4G1.3         22  15915800.0   \n",
       "ENSG00000225255   LA16c-83F12.6   LA16c-83F12.6         22  15741297.0   \n",
       "ENSG00000198062           POTEH           POTEH         22  15690026.0   \n",
       "ENSG00000236666       POTEH-AS1       POTEH-AS1         22  15703403.0   \n",
       "ENSG00000230471     LA16c-2F2.8     LA16c-2F2.8         22  15625205.0   \n",
       "...                         ...             ...        ...         ...   \n",
       "ENSG00000251322          SHANK3          SHANK3         22  50674408.0   \n",
       "ENSG00000225929      AC000036.4      AC000036.4         22  50738173.0   \n",
       "ENSG00000100312             ACR             ACR         22  50738196.0   \n",
       "ENSG00000254499      AC002056.5      AC002056.5         22  50743520.0   \n",
       "ENSG00000079974          RABL2B          RABL2B         22  50783667.0   \n",
       "\n",
       "                        end  \n",
       "Geneid                       \n",
       "ENSG00000233866  15914721.0  \n",
       "ENSG00000225255  15779680.0  \n",
       "ENSG00000198062  15721631.0  \n",
       "ENSG00000236666  15699361.0  \n",
       "ENSG00000230471  15557577.0  \n",
       "...                     ...  \n",
       "ENSG00000251322  50733212.0  \n",
       "ENSG00000225929  50735813.0  \n",
       "ENSG00000100312  50745339.0  \n",
       "ENSG00000254499  50740593.0  \n",
       "ENSG00000079974  50767501.0  \n",
       "\n",
       "[655 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scdata_22.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "scdata_22.write(base_data_dir / \"input_data/chr22_OneK1K_cohort_gene_expression_matrix_14_celltypes_w_gene_locations.h5ad.gz\", compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test distance calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#preprocess scdata\n",
    "sc.pp.normalize_total(scdata, target_sum=1e4)  # Normalize total counts per cell\n",
    "sc.pp.log1p(scdata)  # Apply log-transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Consequence_start_lost</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>gnomADe_OTH_AF</th>\n",
       "      <th>Consequence_intergenic_variant</th>\n",
       "      <th>Location</th>\n",
       "      <th>gnomADe_AFR_AF</th>\n",
       "      <th>Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>Consequence_splice_acceptor_variant</th>\n",
       "      <th>Consequence_splice_donor_5th_base_variant</th>\n",
       "      <th>Consequence_splice_donor_region_variant</th>\n",
       "      <th>Consequence_splice_donor_variant</th>\n",
       "      <th>Consequence_splice_polypyrimidine_tract_variant</th>\n",
       "      <th>Consequence_splice_region_variant</th>\n",
       "      <th>Consequence_stop_gained</th>\n",
       "      <th>Consequence_stop_retained_variant</th>\n",
       "      <th>Consequence_synonymous_variant</th>\n",
       "      <th>Consequence_upstream_gene_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_16849573_A_G</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>22:16849573</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16849971_A_T</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>22:16849971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16850437_G_A</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>22:16850437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16851225_C_T</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>22:16851225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16851356_C_T</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>22:16851356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51202748_A_G</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>22:51202748</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51208568_G_T</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>22:51208568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51211031_A_G</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>22:51211031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51213613_C_T</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>22:51213613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51216564_T_C</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>22:51216564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143083 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Consequence_start_lost PHENO  gnomADe_AF Feature_type  \\\n",
       "snp_id                                                                   \n",
       "22_16849573_A_G                       0     -         NaN            -   \n",
       "22_16849971_A_T                       0     -         NaN            -   \n",
       "22_16850437_G_A                       0     -         NaN            -   \n",
       "22_16851225_C_T                       0     -         NaN            -   \n",
       "22_16851356_C_T                       0     -         NaN            -   \n",
       "...                                 ...   ...         ...          ...   \n",
       "22_51202748_A_G                       0     -         NaN   Transcript   \n",
       "22_51208568_G_T                       0     -         NaN   Transcript   \n",
       "22_51211031_A_G                       0     -         NaN   Transcript   \n",
       "22_51213613_C_T                       0     -         NaN   Transcript   \n",
       "22_51216564_T_C                       0     -         NaN   Transcript   \n",
       "\n",
       "                CLIN_SIG  gnomADe_OTH_AF  Consequence_intergenic_variant  \\\n",
       "snp_id                                                                     \n",
       "22_16849573_A_G        -             NaN                               1   \n",
       "22_16849971_A_T        -             NaN                               1   \n",
       "22_16850437_G_A        -             NaN                               1   \n",
       "22_16851225_C_T        -             NaN                               1   \n",
       "22_16851356_C_T        -             NaN                               1   \n",
       "...                  ...             ...                             ...   \n",
       "22_51202748_A_G        -             NaN                               0   \n",
       "22_51208568_G_T        -             NaN                               0   \n",
       "22_51211031_A_G        -             NaN                               0   \n",
       "22_51213613_C_T        -             NaN                               0   \n",
       "22_51216564_T_C        -             NaN                               0   \n",
       "\n",
       "                    Location  gnomADe_AFR_AF Allele  ...  \\\n",
       "snp_id                                               ...   \n",
       "22_16849573_A_G  22:16849573             NaN      G  ...   \n",
       "22_16849971_A_T  22:16849971             NaN      T  ...   \n",
       "22_16850437_G_A  22:16850437             NaN      A  ...   \n",
       "22_16851225_C_T  22:16851225             NaN      T  ...   \n",
       "22_16851356_C_T  22:16851356             NaN      T  ...   \n",
       "...                      ...             ...    ...  ...   \n",
       "22_51202748_A_G  22:51202748             NaN      G  ...   \n",
       "22_51208568_G_T  22:51208568             NaN      T  ...   \n",
       "22_51211031_A_G  22:51211031             NaN      G  ...   \n",
       "22_51213613_C_T  22:51213613             NaN      T  ...   \n",
       "22_51216564_T_C  22:51216564             NaN      C  ...   \n",
       "\n",
       "                 Consequence_splice_acceptor_variant  \\\n",
       "snp_id                                                 \n",
       "22_16849573_A_G                                    0   \n",
       "22_16849971_A_T                                    0   \n",
       "22_16850437_G_A                                    0   \n",
       "22_16851225_C_T                                    0   \n",
       "22_16851356_C_T                                    0   \n",
       "...                                              ...   \n",
       "22_51202748_A_G                                    0   \n",
       "22_51208568_G_T                                    0   \n",
       "22_51211031_A_G                                    0   \n",
       "22_51213613_C_T                                    0   \n",
       "22_51216564_T_C                                    0   \n",
       "\n",
       "                 Consequence_splice_donor_5th_base_variant  \\\n",
       "snp_id                                                       \n",
       "22_16849573_A_G                                          0   \n",
       "22_16849971_A_T                                          0   \n",
       "22_16850437_G_A                                          0   \n",
       "22_16851225_C_T                                          0   \n",
       "22_16851356_C_T                                          0   \n",
       "...                                                    ...   \n",
       "22_51202748_A_G                                          0   \n",
       "22_51208568_G_T                                          0   \n",
       "22_51211031_A_G                                          0   \n",
       "22_51213613_C_T                                          0   \n",
       "22_51216564_T_C                                          0   \n",
       "\n",
       "                Consequence_splice_donor_region_variant  \\\n",
       "snp_id                                                    \n",
       "22_16849573_A_G                                       0   \n",
       "22_16849971_A_T                                       0   \n",
       "22_16850437_G_A                                       0   \n",
       "22_16851225_C_T                                       0   \n",
       "22_16851356_C_T                                       0   \n",
       "...                                                 ...   \n",
       "22_51202748_A_G                                       0   \n",
       "22_51208568_G_T                                       0   \n",
       "22_51211031_A_G                                       0   \n",
       "22_51213613_C_T                                       0   \n",
       "22_51216564_T_C                                       0   \n",
       "\n",
       "                 Consequence_splice_donor_variant  \\\n",
       "snp_id                                              \n",
       "22_16849573_A_G                                 0   \n",
       "22_16849971_A_T                                 0   \n",
       "22_16850437_G_A                                 0   \n",
       "22_16851225_C_T                                 0   \n",
       "22_16851356_C_T                                 0   \n",
       "...                                           ...   \n",
       "22_51202748_A_G                                 0   \n",
       "22_51208568_G_T                                 0   \n",
       "22_51211031_A_G                                 0   \n",
       "22_51213613_C_T                                 0   \n",
       "22_51216564_T_C                                 0   \n",
       "\n",
       "                 Consequence_splice_polypyrimidine_tract_variant  \\\n",
       "snp_id                                                             \n",
       "22_16849573_A_G                                                0   \n",
       "22_16849971_A_T                                                0   \n",
       "22_16850437_G_A                                                0   \n",
       "22_16851225_C_T                                                0   \n",
       "22_16851356_C_T                                                0   \n",
       "...                                                          ...   \n",
       "22_51202748_A_G                                                0   \n",
       "22_51208568_G_T                                                0   \n",
       "22_51211031_A_G                                                0   \n",
       "22_51213613_C_T                                                0   \n",
       "22_51216564_T_C                                                0   \n",
       "\n",
       "                 Consequence_splice_region_variant  Consequence_stop_gained  \\\n",
       "snp_id                                                                        \n",
       "22_16849573_A_G                                  0                        0   \n",
       "22_16849971_A_T                                  0                        0   \n",
       "22_16850437_G_A                                  0                        0   \n",
       "22_16851225_C_T                                  0                        0   \n",
       "22_16851356_C_T                                  0                        0   \n",
       "...                                            ...                      ...   \n",
       "22_51202748_A_G                                  0                        0   \n",
       "22_51208568_G_T                                  0                        0   \n",
       "22_51211031_A_G                                  0                        0   \n",
       "22_51213613_C_T                                  0                        0   \n",
       "22_51216564_T_C                                  0                        0   \n",
       "\n",
       "                 Consequence_stop_retained_variant  \\\n",
       "snp_id                                               \n",
       "22_16849573_A_G                                  0   \n",
       "22_16849971_A_T                                  0   \n",
       "22_16850437_G_A                                  0   \n",
       "22_16851225_C_T                                  0   \n",
       "22_16851356_C_T                                  0   \n",
       "...                                            ...   \n",
       "22_51202748_A_G                                  0   \n",
       "22_51208568_G_T                                  0   \n",
       "22_51211031_A_G                                  0   \n",
       "22_51213613_C_T                                  0   \n",
       "22_51216564_T_C                                  0   \n",
       "\n",
       "                Consequence_synonymous_variant  \\\n",
       "snp_id                                           \n",
       "22_16849573_A_G                              0   \n",
       "22_16849971_A_T                              0   \n",
       "22_16850437_G_A                              0   \n",
       "22_16851225_C_T                              0   \n",
       "22_16851356_C_T                              0   \n",
       "...                                        ...   \n",
       "22_51202748_A_G                              0   \n",
       "22_51208568_G_T                              0   \n",
       "22_51211031_A_G                              0   \n",
       "22_51213613_C_T                              0   \n",
       "22_51216564_T_C                              0   \n",
       "\n",
       "                 Consequence_upstream_gene_variant  \n",
       "snp_id                                              \n",
       "22_16849573_A_G                                  0  \n",
       "22_16849971_A_T                                  0  \n",
       "22_16850437_G_A                                  0  \n",
       "22_16851225_C_T                                  0  \n",
       "22_16851356_C_T                                  0  \n",
       "...                                            ...  \n",
       "22_51202748_A_G                                  0  \n",
       "22_51208568_G_T                                  0  \n",
       "22_51211031_A_G                                  0  \n",
       "22_51213613_C_T                                  0  \n",
       "22_51216564_T_C                                  0  \n",
       "\n",
       "[143083 rows x 58 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdata.varm[\"annotations_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdata_anno0 = gdata.varm[\"annotations_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_burden(gd_gene, weight_col):\n",
    "    this_weights = np.array(gd_gene.varm['annotations_0'][weight_col])\n",
    "    g_weigthed = gd_gene.X * this_weights\n",
    "    this_burdens = np.nansum(g_weigthed, axis = 1) #TODO implement alternative weighting functions\n",
    "    return this_burdens\n",
    "\n",
    "def _calc_tss_distance_per_gene(variants_df,\n",
    "                                gene_start,\n",
    "                                gene_end,\n",
    "                                col_name_distance,\n",
    "                                col_name_saige=\"\"):\n",
    "    \"\"\"\n",
    "    Calculate absolute TSS Distance.\n",
    "\n",
    "    Parameters:\n",
    "        variants_df (pd.DataFrame): DataFrame with SNP ID as index and contains variant position in column \"Position\"\n",
    "        gene_start (int): Start of Gene aka TS\n",
    "        col_name_distance (str): name of TSS column\n",
    "        col_name_saige (str, optional): determines whether saige formula should be calculated as well.\n",
    "\n",
    "    Returns:\n",
    "        distances_df (pd.DataFrame): Dataframe with tss distance per variant\n",
    "    \"\"\"\n",
    "    distances = {\"snp_id\": [], col_name_distance: []}\n",
    "    saige_distances = []\n",
    "    #print(variants_df.index)\n",
    "    \n",
    "    if gene_start > gene_end:\n",
    "        gene_start, gene_end = gene_end, gene_start  # Reverse strand handling\n",
    "\n",
    "    for i, row in variants_df.iterrows():\n",
    "        if gene_start <= row[\"Position\"] <= gene_end:\n",
    "            distances[\"snp_id\"].append(i)\n",
    "            distances[col_name_distance].append(0)  # Variant is within the gene\n",
    "\n",
    "        # Calculate absolute distances to start and end positions\n",
    "        distance_to_start = abs(row[\"Position\"] - gene_start)  # upstream\n",
    "        distance_to_end = abs(row[\"Position\"] - gene_end)  # downstream\n",
    "\n",
    "        # Get the minimum distance = correct distance\n",
    "        distance = min(distance_to_start, distance_to_end)\n",
    "        distances[\"snp_id\"].append(i)\n",
    "        distances[col_name_distance].append(distance)\n",
    "\n",
    "        # calculate saige if parameter is set\n",
    "        if col_name_saige != \"\":\n",
    "            distance_saige = np.exp(-1e-5 * distance)\n",
    "            saige_distances.append(distance_saige)\n",
    "\n",
    "    # add saige to final dataframe if parameter is set\n",
    "    if col_name_saige != \"\":\n",
    "        distances[col_name_saige] = saige_distances\n",
    "\n",
    "    # return dataframe with tss distance per variant\n",
    "    distance_df = pd.DataFrame(distances)\n",
    "    distance_df.set_index(\"snp_id\", inplace=True)\n",
    "    #print(distance_df)\n",
    "    return distance_df\n",
    "\n",
    "\n",
    "def _find_snps_near_gene(gdata,\n",
    "                         gene_chrom,\n",
    "                         gene_start,\n",
    "                         gene_end,\n",
    "                         bp_range=10000):\n",
    "    \"\"\"\n",
    "    Finds SNPs within a specified range of a gene's location.\n",
    "\n",
    "    Parameters:\n",
    "        gdata (pd.DataFrame): DataFrame with a 'Location' column in the format \"chromosome:position\".\n",
    "        gene_chrom, gene_start, gene_end (int)\n",
    "        bp_range (int): Range in base pairs to search upstream and downstream.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: With data of SNPs within the specified range.\n",
    "    \"\"\"\n",
    "    # Parse the gene location\n",
    "    #import ipdb; ipdb.set_trace()\n",
    "\n",
    "    # Extract chromosome and position from the SNPs\n",
    "    gene_chrom = str(gene_chrom)\n",
    "    gdata_df = gdata.copy()\n",
    "    gdata_df[['Chromosome', 'Position']] = gdata_df['Location'].str.split(':', expand=True)\n",
    "    gdata_df['Position'] = gdata_df['Position'].astype(int)\n",
    "\n",
    "    # Filter for SNPs within the range\n",
    "    # snps_in_range = gdata_df[\n",
    "    #     (gdata_df['Chromosome'] == gene_chrom) &\n",
    "    #     (gdata_df['Position'] >= gene_start - bp_range) &\n",
    "    #     (gdata_df['Position'] <= gene_end + bp_range)\n",
    "    # ]\n",
    "\n",
    "    if gene_start < gene_end:\n",
    "        # forward strand\n",
    "        snps_upstream = gdata_df[\n",
    "            (gdata_df['Chromosome'] == gene_chrom) &\n",
    "            (gdata_df['Position'] >= gene_start - bp_range)\n",
    "        ]\n",
    "        snps_downstream = gdata_df[\n",
    "            (gdata_df['Chromosome'] == gene_chrom) &\n",
    "            (gdata_df['Position'] <= gene_start + bp_range)\n",
    "        ]\n",
    "    else:\n",
    "        # reverse strand\n",
    "        snps_upstream = gdata_df[\n",
    "            (gdata_df['Chromosome'] == gene_chrom) &\n",
    "            (gdata_df['Position'] >= gene_start + bp_range)\n",
    "        ]\n",
    "        snps_downstream = gdata_df[\n",
    "            (gdata_df['Chromosome'] == gene_chrom) &\n",
    "            (gdata_df['Position'] <= gene_start - bp_range)\n",
    "        ]\n",
    "\n",
    "    # return snps_in_range.index\n",
    "    return snps_upstream, snps_downstream\n",
    "\n",
    "\n",
    "def _compute_burdens_for_gene(this_gd,\n",
    "                              this_gene,\n",
    "                              gene_chrom,\n",
    "                              gene_start,\n",
    "                              gene_end,\n",
    "                              weight_cols,\n",
    "                              annotation_varm=\"annotations_0\",\n",
    "                              window_size=100000,\n",
    "                              DNA_LM_up=\"\",\n",
    "                              DNA_LM_down=\"\",\n",
    "                              DNA_LM_mixed=\"DNA_LM_mixed\",\n",
    "                              GENE_TSS_DISTANCE=\"\",\n",
    "                              GENE_TSS_DISTANCE_SAIGE=\"\"):\n",
    "    \"\"\"\n",
    "    Compute burdenscores for a given gene and given annotations\n",
    "\n",
    "    Parameters:\n",
    "        this_gd (pd.DataFrame): ddata.gdata.\n",
    "        this_gene (str): Ensemble ID.\n",
    "        gene_chrom, gene_start, gene_end (int)\n",
    "        weight_cols (list): colnames of variant annotations to compute burden scores for.\n",
    "        annotation_varm (str): key for pd.DataFrame (gdata.varm[key])\n",
    "        window_size (int)\n",
    "        DNA_LM_up (str): colname for DNA_LM upstream model\n",
    "                    if empty, mixed model is not computed\n",
    "        DNA_LM_down (str): colname for DNA_LM downstream model\n",
    "                    if empty, mixed model is not computed\n",
    "        DNA_LM_mixed (str): name of mixed model column\n",
    "        GENE_TSS_DISTANCE (str): name for tss distance column\n",
    "        GENE_TSS_DISTANCE_SAIGE (str): name for tss distance saige column\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame containing burden scores for this_gene across the weight_cols\n",
    "    \"\"\"\n",
    "    # Return a dataframe with None entries for weight cols, if gene location is nan (this means it could not be found in ensembl)\n",
    "    if np.isnan(gene_chrom):\n",
    "        print(f\"Failed to retrieve location for gene {this_gene}. No Burden scores computed.\")\n",
    "        # Create a DataFrame with None for all the weight columns\n",
    "        empty_burdens = pd.DataFrame(\n",
    "            None,\n",
    "            index=this_gd.obs.index,  # Assuming these are the sample indices\n",
    "            columns=weight_cols\n",
    "        )\n",
    "        # Add the Geneid column\n",
    "        empty_burdens[\"Geneid\"] = this_gene\n",
    "        return empty_burdens\n",
    "\n",
    "    # Filter the variants using the SNP location and gene location\n",
    "    this_vars_up_df, this_vars_down_df = _find_snps_near_gene(this_gd.varm[annotation_varm], gene_chrom, gene_start, gene_end, bp_range=window_size)\n",
    "\n",
    "    # get snps IDs\n",
    "    this_vars_down = this_vars_down_df.index\n",
    "    this_vars_up = this_vars_up_df.index\n",
    "\n",
    "    gd_gene = this_gd[:, this_vars_up.append(this_vars_down).unique()].copy()\n",
    "\n",
    "    # if mixed model is computed, add column for DNA_LM mixed model\n",
    "    if DNA_LM_up != \"\":\n",
    "        # Add the \"DNA_LM_mixed\" column to the annotations_0 DataFrame\n",
    "        gd_gene.varm[\"annotations_0\"][DNA_LM_mixed] = np.nan  # Initialize the column with NaN\n",
    "\n",
    "        # Assign values to the \"DNA_LM_mixed\" column based on the conditions\n",
    "        gd_gene.varm[\"annotations_0\"].loc[this_vars_up, DNA_LM_mixed] = gd_gene.varm[\"annotations_0\"].loc[this_vars_up, DNA_LM_up]\n",
    "        gd_gene.varm[\"annotations_0\"].loc[this_vars_down, DNA_LM_mixed] = gd_gene.varm[\"annotations_0\"].loc[this_vars_down, DNA_LM_down]\n",
    "\n",
    "    if GENE_TSS_DISTANCE_SAIGE != \"\": # calc GENE_TSS_DISTANCE and GENE_TSS_DISTANCE_SAIGE\n",
    "        gd_gene.varm[\"annotations_0\"][GENE_TSS_DISTANCE] = np.nan # Initialize the column with NaN\n",
    "        gd_gene.varm[\"annotations_0\"][GENE_TSS_DISTANCE_SAIGE] = np.nan  # Initialize the column with NaN\n",
    "        # calculate GENE_TSS_DISTANCE and GENE_TSS_DISTANCE_SAIGE independent of up or downstream\n",
    "        all_variants = pd.concat([this_vars_up_df, this_vars_down_df], axis=0)\n",
    "        distances = _calc_tss_distance_per_gene(all_variants, gene_start, gene_end, GENE_TSS_DISTANCE, GENE_TSS_DISTANCE_SAIGE)\n",
    "        #print(distances)\n",
    "        #print(distances.index)\n",
    "        # add GENE_TSS_DISTANCE and GENE_TSS_DISTANCE_SAIGE to annotation 0\n",
    "        gd_gene.varm[\"annotations_0\"].loc[distances.index, GENE_TSS_DISTANCE] = distances[GENE_TSS_DISTANCE]\n",
    "        gd_gene.varm[\"annotations_0\"].loc[distances.index, GENE_TSS_DISTANCE_SAIGE] = distances[GENE_TSS_DISTANCE_SAIGE]\n",
    "    elif GENE_TSS_DISTANCE != \"\":  # calc only GENE_TSS_DISTANCE\n",
    "        gd_gene.varm[\"annotations_0\"][GENE_TSS_DISTANCE] = np.nan  # Initialize the column with NaN\n",
    "        # calculate tss distance independent of up or downstream\n",
    "        all_variants = pd.concat([this_vars_up_df, this_vars_down_df])\n",
    "        tss_distances = _calc_tss_distance_per_gene(all_variants, gene_start, gene_end, GENE_TSS_DISTANCE)\n",
    "        # add to gd_gene\n",
    "        gd_gene.varm[\"annotations_0\"].loc[tss_distances.index, GENE_TSS_DISTANCE] = tss_distances[GENE_TSS_DISTANCE]\n",
    "\n",
    "    all_burdens_this_gene = []\n",
    "    for weight_col in weight_cols:\n",
    "        this_burden = _get_burden(gd_gene, weight_col)\n",
    "        all_burdens_this_gene.append(this_burden)\n",
    "\n",
    "    all_burdens_this_gene = np.stack(all_burdens_this_gene, axis=1)\n",
    "    all_burdens_this_gene = pd.DataFrame(all_burdens_this_gene, index=gd_gene.obs.index, columns=weight_cols)\n",
    "    all_burdens_this_gene[\"Geneid\"] = this_gene\n",
    "\n",
    "    return all_burdens_this_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "this_gd = gdata[:, gdata.var[\"maf\"] < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_TSS_DISTANCE</th>\n",
       "      <th>GENE_TSS_DISTANCE_SAIGE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>5.974337e+10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>6.255511e+10</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>5.302204e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>6.431704e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>5.468808e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096_845_2</th>\n",
       "      <td>5.767224e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100_914_2</th>\n",
       "      <td>5.312065e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102_932_2</th>\n",
       "      <td>6.263424e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103_926_2</th>\n",
       "      <td>6.648737e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104_920_2</th>\n",
       "      <td>5.762447e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1034 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE_TSS_DISTANCE  GENE_TSS_DISTANCE_SAIGE           Geneid\n",
       "id                                                                     \n",
       "1_1              5.974337e+10                 0.000005  ENSG00000233866\n",
       "2_2              6.255511e+10                 0.000030  ENSG00000233866\n",
       "3_3              5.302204e+10                 0.000003  ENSG00000233866\n",
       "4_4              6.431704e+10                 0.000004  ENSG00000233866\n",
       "6_6              5.468808e+10                 0.000004  ENSG00000233866\n",
       "...                       ...                      ...              ...\n",
       "1096_845_2       5.767224e+10                 0.000003  ENSG00000233866\n",
       "1100_914_2       5.312065e+10                 0.000004  ENSG00000233866\n",
       "1102_932_2       6.263424e+10                 0.000003  ENSG00000233866\n",
       "1103_926_2       6.648737e+10                 0.000004  ENSG00000233866\n",
       "1104_920_2       5.762447e+10                 0.000003  ENSG00000233866\n",
       "\n",
       "[1034 rows x 3 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_compute_burdens_for_gene(this_gd, \"ENSG00000233866\", 22, 15915800.0, 15914721.0, [\"GENE_TSS_DISTANCE\",\"GENE_TSS_DISTANCE_SAIGE\"], \"annotations_0\", 10000,GENE_TSS_DISTANCE=\"GENE_TSS_DISTANCE\", GENE_TSS_DISTANCE_SAIGE=\"GENE_TSS_DISTANCE_SAIGE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_TSS_DISTANCE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>5.974337e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>6.255511e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>5.302204e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>6.431704e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>5.468808e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096_845_2</th>\n",
       "      <td>5.767224e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100_914_2</th>\n",
       "      <td>5.312065e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102_932_2</th>\n",
       "      <td>6.263424e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103_926_2</th>\n",
       "      <td>6.648737e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104_920_2</th>\n",
       "      <td>5.762447e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1034 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE_TSS_DISTANCE           Geneid\n",
       "id                                            \n",
       "1_1              5.974337e+10  ENSG00000233866\n",
       "2_2              6.255511e+10  ENSG00000233866\n",
       "3_3              5.302204e+10  ENSG00000233866\n",
       "4_4              6.431704e+10  ENSG00000233866\n",
       "6_6              5.468808e+10  ENSG00000233866\n",
       "...                       ...              ...\n",
       "1096_845_2       5.767224e+10  ENSG00000233866\n",
       "1100_914_2       5.312065e+10  ENSG00000233866\n",
       "1102_932_2       6.263424e+10  ENSG00000233866\n",
       "1103_926_2       6.648737e+10  ENSG00000233866\n",
       "1104_920_2       5.762447e+10  ENSG00000233866\n",
       "\n",
       "[1034 rows x 2 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_compute_burdens_for_gene(this_gd, \"ENSG00000233866\", 22, 15915800.0, 15914721.0, [\"GENE_TSS_DISTANCE\"], \"annotations_0\", 10000,GENE_TSS_DISTANCE=\"GENE_TSS_DISTANCE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_TSS_DISTANCE_SAIGE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>0.000030</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096_845_2</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100_914_2</th>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102_932_2</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103_926_2</th>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104_920_2</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1034 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE_TSS_DISTANCE_SAIGE           Geneid\n",
       "id                                                  \n",
       "1_1                        0.000005  ENSG00000233866\n",
       "2_2                        0.000030  ENSG00000233866\n",
       "3_3                        0.000003  ENSG00000233866\n",
       "4_4                        0.000004  ENSG00000233866\n",
       "6_6                        0.000004  ENSG00000233866\n",
       "...                             ...              ...\n",
       "1096_845_2                 0.000003  ENSG00000233866\n",
       "1100_914_2                 0.000004  ENSG00000233866\n",
       "1102_932_2                 0.000003  ENSG00000233866\n",
       "1103_926_2                 0.000004  ENSG00000233866\n",
       "1104_920_2                 0.000003  ENSG00000233866\n",
       "\n",
       "[1034 rows x 2 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_compute_burdens_for_gene(this_gd, \"ENSG00000233866\", 22, 15915800.0, 15914721.0, [\"GENE_TSS_DISTANCE_SAIGE\"], \"annotations_0\", 10000,GENE_TSS_DISTANCE_SAIGE=\"GENE_TSS_DISTANCE_SAIGE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_burdens(ddata,\n",
    "                    max_af=0.05,\n",
    "                    weight_cols=[\"DISTANCE\", \"CADD_PHRED\"],\n",
    "                    annotations_varm=\"annotations_0\",\n",
    "                    window_size=100000,\n",
    "                    DNA_LM_up=\"\",\n",
    "                    DNA_LM_down=\"\",\n",
    "                    DNA_LM_mixed=\"DNA_LM_mixed\",\n",
    "                    GENE_TSS_DISTANCE=\"\",\n",
    "                    GENE_TSS_DISTANCE_SAIGE=\"\"):\n",
    "    \"\"\"Compute gene burdens for each gene and sample using different variant annotations\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    ddata : ddata\n",
    "        _description_\n",
    "    max_af : float, optional\n",
    "        maximum variant minor allele frequency, by default 0.05\n",
    "    weight_cols : list, optional\n",
    "        variant annotations used for weighting (columns of gdata.varm.annotations_0), by default [\"DISTANCE\", \"CADD_PHRED\"]\n",
    "    annotations_varm: str, optional\n",
    "        key for gdata.varm dataframe (eg: gdata.varm[\"annotations_0\"])\n",
    "    window_size: int, optional\n",
    "        range around gene TSS, in which variants are regarded for gene burden scores\n",
    "    DNA_LM_up: str, optional, if DNA_LM mixed model should be computed\n",
    "        colname of DNA_LM score upstream model\n",
    "    DNA_LM_down: str, optional, if DNA_LM mixed model should be computed\n",
    "        colname of DNA_LM score downstream model\n",
    "    DNA_LM_mixed: str, optional\n",
    "        colname of DNA_LM score mixed model\n",
    "    GENE_TSS_DISTANCE: str, optional\n",
    "        colname for TSS distances and also flag to compute its burden \n",
    "    GENE_TSS_DISTANCE_SAIGE:str, optional\n",
    "        colname for TSS distances using saige formula and also flag to compute its burden \n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        gene burdens for all genes, individuals and all annotations in weightcols\n",
    "    \"\"\"\n",
    "    if (DNA_LM_up != \"\" and DNA_LM_down == \"\") or (DNA_LM_up == \"\" and DNA_LM_down != \"\"):\n",
    "        raise ValueError(\"If you want to compute the burden scores using the DNA_LM mixed model, you must set both DNA_LM_up and DNA_LM_down, else leave both empty.\")\n",
    "\n",
    "    this_gd = ddata.gdata.copy()\n",
    "    this_ad = ddata.adata.copy()\n",
    "    this_gd = this_gd[:, this_gd.var[\"maf\"] < max_af]\n",
    "    all_burdens = []\n",
    "\n",
    "    if not all(col in this_ad.var.columns for col in [\"chromosome\", \"start\", \"end\"]):\n",
    "        # compute all the gene locations\n",
    "        this_ad.var[['chromosome', 'start', 'end']] = this_ad.var.index.to_series().apply(\n",
    "            lambda x: pd.Series(_get_gene_location(x))\n",
    "        )\n",
    "\n",
    "    # add mixed model to the weight cols for which burden score is computed\n",
    "    if DNA_LM_up != \"\" and DNA_LM_down != \"\" and DNA_LM_mixed not in weight_cols:\n",
    "        weight_cols.append(DNA_LM_mixed)\n",
    "        if DNA_LM_up not in weight_cols:\n",
    "            weight_cols.append(DNA_LM_up)\n",
    "        if DNA_LM_down not in weight_cols:\n",
    "            weight_cols.append(DNA_LM_down)\n",
    "\n",
    "    # add tss distance to the weight cols to initialize burden score computations\n",
    "    if GENE_TSS_DISTANCE != \"\" and GENE_TSS_DISTANCE not in weight_cols:\n",
    "        weight_cols.append(GENE_TSS_DISTANCE)\n",
    "\n",
    "    # add tss distance saige to the weight cols to initialize burden score computations\n",
    "    if GENE_TSS_DISTANCE_SAIGE != \"\":\n",
    "        if GENE_TSS_DISTANCE_SAIGE not in weight_cols:\n",
    "            weight_cols.append(GENE_TSS_DISTANCE_SAIGE)\n",
    "        if GENE_TSS_DISTANCE == \"\":  # if saige is set then tss distance has to be calculated too\n",
    "            GENE_TSS_DISTANCE = \"GENE_TSS_DISTANCE\"\n",
    "\n",
    "    for gene in tqdm(this_ad.var.index[0:1]):\n",
    "        gene_chrom = int(this_ad.var.loc[gene, \"chromosome\"])\n",
    "        gene_start = int(this_ad.var.loc[gene, \"start\"])\n",
    "        gene_end = int(this_ad.var.loc[gene, \"end\"])\n",
    "\n",
    "        this_b = _compute_burdens_for_gene(this_gd, gene, gene_chrom, gene_start, gene_end, weight_cols, annotations_varm, window_size, DNA_LM_up, DNA_LM_down, DNA_LM_mixed, GENE_TSS_DISTANCE, GENE_TSS_DISTANCE_SAIGE)\n",
    "        all_burdens.append(this_b)\n",
    "\n",
    "    all_burdens = pd.concat(all_burdens)\n",
    "\n",
    "    return all_burdens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-01-19 18:33:15,519] INFO:cellink._core.donordata: Keeping 981/1034 donors\n",
      "[2025-01-19 18:33:15,520] INFO:cellink._core.donordata: Dropping 53/1034 donors from genetic data\n",
      "[2025-01-19 18:33:15,520] INFO:cellink._core.donordata: Dropping 0/981 donors from single-cell data\n"
     ]
    }
   ],
   "source": [
    "data = cl.DonorData(adata=scdata, gdata=gdata, donor_key_in_sc_adata=\"individual\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000233866\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000233866\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "00%|██████████| 1/1 [00:05<00:00,  5.22s/it]"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>DNA_LM_mixed</th>\n",
       "      <th>GENE_TSS_DISTANCE</th>\n",
       "      <th>GENE_TSS_DISTANCE_SAIGE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>9350.801</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>22.916096</td>\n",
       "      <td>43612.917035</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>5.974337e+10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>10775.218</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>28.387657</td>\n",
       "      <td>52127.930292</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>6.255511e+10</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>8327.426</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>21.604692</td>\n",
       "      <td>39069.213569</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>5.302204e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>9439.675</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>32.699742</td>\n",
       "      <td>44983.557785</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>6.431704e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>9054.896</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>23.019223</td>\n",
       "      <td>42525.702648</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>5.468808e+10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>9227.155</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>27.382034</td>\n",
       "      <td>42677.295555</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>5.507287e+10</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>9411.798</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>25.363997</td>\n",
       "      <td>44142.005875</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>6.204754e+10</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>9862.022</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>23.919521</td>\n",
       "      <td>44060.525919</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>6.135349e+10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>8392.390</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>24.546299</td>\n",
       "      <td>40866.836483</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>5.485095e+10</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>9015.698</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>26.508232</td>\n",
       "      <td>42943.342890</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>6.015882e+10</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>981 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  DNA_LM_mixed  \\\n",
       "id                                                                           \n",
       "1_1          9350.801  13.224558    22.916096   43612.917035     13.224558   \n",
       "2_2         10775.218  16.520484    28.387657   52127.930292     16.520484   \n",
       "3_3          8327.426  12.240323    21.604692   39069.213569     12.240323   \n",
       "4_4          9439.675  18.346526    32.699742   44983.557785     18.346526   \n",
       "6_6          9054.896  13.340017    23.019223   42525.702648     13.340017   \n",
       "...               ...        ...          ...            ...           ...   \n",
       "1077_1078    9227.155  15.636985    27.382034   42677.295555     15.636985   \n",
       "1078_1079    9411.798  14.626046    25.363997   44142.005875     14.626046   \n",
       "1079_1080    9862.022  13.470969    23.919521   44060.525919     13.470969   \n",
       "1080_1081    8392.390  13.858525    24.546299   40866.836483     13.858525   \n",
       "1081_1082    9015.698  15.057642    26.508232   42943.342890     15.057642   \n",
       "\n",
       "           GENE_TSS_DISTANCE  GENE_TSS_DISTANCE_SAIGE           Geneid  \n",
       "id                                                                      \n",
       "1_1             5.974337e+10                 0.000005  ENSG00000233866  \n",
       "2_2             6.255511e+10                 0.000030  ENSG00000233866  \n",
       "3_3             5.302204e+10                 0.000003  ENSG00000233866  \n",
       "4_4             6.431704e+10                 0.000004  ENSG00000233866  \n",
       "6_6             5.468808e+10                 0.000004  ENSG00000233866  \n",
       "...                      ...                      ...              ...  \n",
       "1077_1078       5.507287e+10                 0.000006  ENSG00000233866  \n",
       "1078_1079       6.204754e+10                 0.000096  ENSG00000233866  \n",
       "1079_1080       6.135349e+10                 0.000005  ENSG00000233866  \n",
       "1080_1081       5.485095e+10                 0.000003  ENSG00000233866  \n",
       "1081_1082       6.015882e+10                 0.000054  ENSG00000233866  \n",
       "\n",
       "[981 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_burdens(data, max_af=0.05, weight_cols=[\"CADD_PHRED\", \"DNA_LM_up\", \"DNA_LM_down\", \"MAF_beta_1.25\"], window_size=100000, DNA_LM_up=\"DNA_LM_up\", DNA_LM_down=\"DNA_LM_down\",GENE_TSS_DISTANCE=\"GENE_TSS_DISTANCE\", GENE_TSS_DISTANCE_SAIGE=\"GENE_TSS_DISTANCE_SAIGE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "00%|██████████| 1/1 [00:04<00:00,  4.84s/it]"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>DNA_LM_mixed</th>\n",
       "      <th>GENE_TSS_DISTANCE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>9350.801</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>22.916096</td>\n",
       "      <td>43612.917035</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>5.974337e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>10775.218</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>28.387657</td>\n",
       "      <td>52127.930292</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>6.255511e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>8327.426</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>21.604692</td>\n",
       "      <td>39069.213569</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>5.302204e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>9439.675</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>32.699742</td>\n",
       "      <td>44983.557785</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>6.431704e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>9054.896</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>23.019223</td>\n",
       "      <td>42525.702648</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>5.468808e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>9227.155</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>27.382034</td>\n",
       "      <td>42677.295555</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>5.507287e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>9411.798</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>25.363997</td>\n",
       "      <td>44142.005875</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>6.204754e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>9862.022</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>23.919521</td>\n",
       "      <td>44060.525919</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>6.135349e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>8392.390</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>24.546299</td>\n",
       "      <td>40866.836483</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>5.485095e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>9015.698</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>26.508232</td>\n",
       "      <td>42943.342890</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>6.015882e+10</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>981 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  DNA_LM_mixed  \\\n",
       "id                                                                           \n",
       "1_1          9350.801  13.224558    22.916096   43612.917035     13.224558   \n",
       "2_2         10775.218  16.520484    28.387657   52127.930292     16.520484   \n",
       "3_3          8327.426  12.240323    21.604692   39069.213569     12.240323   \n",
       "4_4          9439.675  18.346526    32.699742   44983.557785     18.346526   \n",
       "6_6          9054.896  13.340017    23.019223   42525.702648     13.340017   \n",
       "...               ...        ...          ...            ...           ...   \n",
       "1077_1078    9227.155  15.636985    27.382034   42677.295555     15.636985   \n",
       "1078_1079    9411.798  14.626046    25.363997   44142.005875     14.626046   \n",
       "1079_1080    9862.022  13.470969    23.919521   44060.525919     13.470969   \n",
       "1080_1081    8392.390  13.858525    24.546299   40866.836483     13.858525   \n",
       "1081_1082    9015.698  15.057642    26.508232   42943.342890     15.057642   \n",
       "\n",
       "           GENE_TSS_DISTANCE           Geneid  \n",
       "id                                             \n",
       "1_1             5.974337e+10  ENSG00000233866  \n",
       "2_2             6.255511e+10  ENSG00000233866  \n",
       "3_3             5.302204e+10  ENSG00000233866  \n",
       "4_4             6.431704e+10  ENSG00000233866  \n",
       "6_6             5.468808e+10  ENSG00000233866  \n",
       "...                      ...              ...  \n",
       "1077_1078       5.507287e+10  ENSG00000233866  \n",
       "1078_1079       6.204754e+10  ENSG00000233866  \n",
       "1079_1080       6.135349e+10  ENSG00000233866  \n",
       "1080_1081       5.485095e+10  ENSG00000233866  \n",
       "1081_1082       6.015882e+10  ENSG00000233866  \n",
       "\n",
       "[981 rows x 7 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_burdens(data, max_af=0.05, weight_cols=[\"CADD_PHRED\", \"DNA_LM_up\", \"DNA_LM_down\", \"MAF_beta_1.25\"], window_size=100000, DNA_LM_up=\"DNA_LM_up\", DNA_LM_down=\"DNA_LM_down\",GENE_TSS_DISTANCE=\"GENE_TSS_DISTANCE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "00%|██████████| 1/1 [00:04<00:00,  4.95s/it]"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>DNA_LM_mixed</th>\n",
       "      <th>GENE_TSS_DISTANCE_SAIGE</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>9350.801</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>22.916096</td>\n",
       "      <td>43612.917035</td>\n",
       "      <td>13.224558</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>10775.218</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>28.387657</td>\n",
       "      <td>52127.930292</td>\n",
       "      <td>16.520484</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>8327.426</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>21.604692</td>\n",
       "      <td>39069.213569</td>\n",
       "      <td>12.240323</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>9439.675</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>32.699742</td>\n",
       "      <td>44983.557785</td>\n",
       "      <td>18.346526</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>9054.896</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>23.019223</td>\n",
       "      <td>42525.702648</td>\n",
       "      <td>13.340017</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>9227.155</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>27.382034</td>\n",
       "      <td>42677.295555</td>\n",
       "      <td>15.636985</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>9411.798</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>25.363997</td>\n",
       "      <td>44142.005875</td>\n",
       "      <td>14.626046</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>9862.022</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>23.919521</td>\n",
       "      <td>44060.525919</td>\n",
       "      <td>13.470969</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>8392.390</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>24.546299</td>\n",
       "      <td>40866.836483</td>\n",
       "      <td>13.858525</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>9015.698</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>26.508232</td>\n",
       "      <td>42943.342890</td>\n",
       "      <td>15.057642</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>981 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  DNA_LM_mixed  \\\n",
       "id                                                                           \n",
       "1_1          9350.801  13.224558    22.916096   43612.917035     13.224558   \n",
       "2_2         10775.218  16.520484    28.387657   52127.930292     16.520484   \n",
       "3_3          8327.426  12.240323    21.604692   39069.213569     12.240323   \n",
       "4_4          9439.675  18.346526    32.699742   44983.557785     18.346526   \n",
       "6_6          9054.896  13.340017    23.019223   42525.702648     13.340017   \n",
       "...               ...        ...          ...            ...           ...   \n",
       "1077_1078    9227.155  15.636985    27.382034   42677.295555     15.636985   \n",
       "1078_1079    9411.798  14.626046    25.363997   44142.005875     14.626046   \n",
       "1079_1080    9862.022  13.470969    23.919521   44060.525919     13.470969   \n",
       "1080_1081    8392.390  13.858525    24.546299   40866.836483     13.858525   \n",
       "1081_1082    9015.698  15.057642    26.508232   42943.342890     15.057642   \n",
       "\n",
       "           GENE_TSS_DISTANCE_SAIGE           Geneid  \n",
       "id                                                   \n",
       "1_1                       0.000005  ENSG00000233866  \n",
       "2_2                       0.000030  ENSG00000233866  \n",
       "3_3                       0.000003  ENSG00000233866  \n",
       "4_4                       0.000004  ENSG00000233866  \n",
       "6_6                       0.000004  ENSG00000233866  \n",
       "...                            ...              ...  \n",
       "1077_1078                 0.000006  ENSG00000233866  \n",
       "1078_1079                 0.000096  ENSG00000233866  \n",
       "1079_1080                 0.000005  ENSG00000233866  \n",
       "1080_1081                 0.000003  ENSG00000233866  \n",
       "1081_1082                 0.000054  ENSG00000233866  \n",
       "\n",
       "[981 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_burdens(data, max_af=0.05, weight_cols=[\"CADD_PHRED\", \"DNA_LM_up\", \"DNA_LM_down\", \"MAF_beta_1.25\"], window_size=100000, DNA_LM_up=\"DNA_LM_up\", DNA_LM_down=\"DNA_LM_down\", GENE_TSS_DISTANCE_SAIGE=\"GENE_TSS_DISTANCE_SAIGE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_gene</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>burden_type</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>0.541026</td>\n",
       "      <td>0.000362</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>CADD_PHRED</td>\n",
       "      <td>0.285429</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>DNA_LM_influence_score</td>\n",
       "      <td>0.341302</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>0.365920</td>\n",
       "      <td>0.000537</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000237438</td>\n",
       "      <td>ENSG00000237438</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>0.321480</td>\n",
       "      <td>0.000314</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>0.958578</td>\n",
       "      <td>-0.000063</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>CADD_PHRED</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1870</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>DNA_LM_influence_score</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1871</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD8 NC</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1872 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          burden_gene      target_gene             burden_type    pvalue  \\\n",
       "0     ENSG00000100181  ENSG00000100181                DISTANCE  0.541026   \n",
       "1     ENSG00000100181  ENSG00000100181              CADD_PHRED  0.285429   \n",
       "2     ENSG00000100181  ENSG00000100181  DNA_LM_influence_score  0.341302   \n",
       "3     ENSG00000100181  ENSG00000100181           MAF_beta_1.25  0.365920   \n",
       "4     ENSG00000237438  ENSG00000237438                DISTANCE  0.321480   \n",
       "...               ...              ...                     ...       ...   \n",
       "1867  ENSG00000100299  ENSG00000100299           MAF_beta_1.25  0.958578   \n",
       "1868  ENSG00000079974  ENSG00000079974                DISTANCE  1.000000   \n",
       "1869  ENSG00000079974  ENSG00000079974              CADD_PHRED  1.000000   \n",
       "1870  ENSG00000079974  ENSG00000079974  DNA_LM_influence_score  1.000000   \n",
       "1871  ENSG00000079974  ENSG00000079974           MAF_beta_1.25  1.000000   \n",
       "\n",
       "          beta cell_type  chrom  \n",
       "0     0.000362    CD8 NC     22  \n",
       "1     0.000635    CD8 NC     22  \n",
       "2     0.000566    CD8 NC     22  \n",
       "3     0.000537    CD8 NC     22  \n",
       "4     0.000314    CD8 NC     22  \n",
       "...        ...       ...    ...  \n",
       "1867 -0.000063    CD8 NC     22  \n",
       "1868       NaN    CD8 NC     22  \n",
       "1869       NaN    CD8 NC     22  \n",
       "1870       NaN    CD8 NC     22  \n",
       "1871       NaN    CD8 NC     22  \n",
       "\n",
       "[1872 rows x 7 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.read_csv(base_data_dir/\"output/example_output/burden_CellTypeCD8-NC_Chrom22.csv\")\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_res = pd.read_pickle(base_data_dir/\"output/all_results_DNA_LM_and_MAF.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_gene</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>burden_type</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pvalue_corrected</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>0.766616</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>CD8 ET</td>\n",
       "      <td>22</td>\n",
       "      <td>19376.982274</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>CADD_PHRED</td>\n",
       "      <td>0.840065</td>\n",
       "      <td>-0.000085</td>\n",
       "      <td>CD8 ET</td>\n",
       "      <td>22</td>\n",
       "      <td>21233.491502</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>DNA_LM_influence_score</td>\n",
       "      <td>0.596951</td>\n",
       "      <td>-0.000224</td>\n",
       "      <td>CD8 ET</td>\n",
       "      <td>22</td>\n",
       "      <td>15088.544000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>0.896485</td>\n",
       "      <td>-0.000055</td>\n",
       "      <td>CD8 ET</td>\n",
       "      <td>22</td>\n",
       "      <td>22659.563673</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237438</td>\n",
       "      <td>ENSG00000237438</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>0.521166</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>CD8 ET</td>\n",
       "      <td>22</td>\n",
       "      <td>13172.988603</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>ENSG00000100288</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>0.093014</td>\n",
       "      <td>-0.030436</td>\n",
       "      <td>Erythrocytes</td>\n",
       "      <td>22</td>\n",
       "      <td>2351.011086</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>DISTANCE</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Erythrocytes</td>\n",
       "      <td>22</td>\n",
       "      <td>25276.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>CADD_PHRED</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Erythrocytes</td>\n",
       "      <td>22</td>\n",
       "      <td>25276.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>DNA_LM_influence_score</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Erythrocytes</td>\n",
       "      <td>22</td>\n",
       "      <td>25276.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>MAF_beta_1.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Erythrocytes</td>\n",
       "      <td>22</td>\n",
       "      <td>25276.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25276 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        burden_gene      target_gene             burden_type    pvalue  \\\n",
       "0   ENSG00000100181  ENSG00000100181                DISTANCE  0.766616   \n",
       "1   ENSG00000100181  ENSG00000100181              CADD_PHRED  0.840065   \n",
       "2   ENSG00000100181  ENSG00000100181  DNA_LM_influence_score  0.596951   \n",
       "3   ENSG00000100181  ENSG00000100181           MAF_beta_1.25  0.896485   \n",
       "0   ENSG00000237438  ENSG00000237438                DISTANCE  0.521166   \n",
       "..              ...              ...                     ...       ...   \n",
       "3   ENSG00000100288  ENSG00000100288           MAF_beta_1.25  0.093014   \n",
       "0   ENSG00000079974  ENSG00000079974                DISTANCE  1.000000   \n",
       "1   ENSG00000079974  ENSG00000079974              CADD_PHRED  1.000000   \n",
       "2   ENSG00000079974  ENSG00000079974  DNA_LM_influence_score  1.000000   \n",
       "3   ENSG00000079974  ENSG00000079974           MAF_beta_1.25  1.000000   \n",
       "\n",
       "        beta     cell_type chrom  pvalue_corrected  significant  \n",
       "0   0.000125        CD8 ET    22      19376.982274        False  \n",
       "1  -0.000085        CD8 ET    22      21233.491502        False  \n",
       "2  -0.000224        CD8 ET    22      15088.544000        False  \n",
       "3  -0.000055        CD8 ET    22      22659.563673        False  \n",
       "0   0.000112        CD8 ET    22      13172.988603        False  \n",
       "..       ...           ...   ...               ...          ...  \n",
       "3  -0.030436  Erythrocytes    22       2351.011086        False  \n",
       "0        NaN  Erythrocytes    22      25276.000000        False  \n",
       "1        NaN  Erythrocytes    22      25276.000000        False  \n",
       "2        NaN  Erythrocytes    22      25276.000000        False  \n",
       "3        NaN  Erythrocytes    22      25276.000000        False  \n",
       "\n",
       "[25276 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-mm_scgenetics]",
   "language": "python",
   "name": "conda-env-.conda-mm_scgenetics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
