{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9992e6bf-f6b8-4023-ac28-81c07c8aa289",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df6f91ac-d01b-42c8-90d7-8080fda37793",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/nasif12/home_if12/l_mai/.conda/envs/mm_scgenetics/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import cellink as cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "33c88950-3429-418e-be8f-049a6cba5002",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cellink.tl._burden_testing import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d6fed0fb-3f3f-4588-ab05-6e379a04f74f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_col = \"cell_label\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68891e79-3ec6-4ccb-8e03-3c1f89eaaddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_data_dir = Path(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/\")\n",
    "input_dir = base_data_dir / \"input_data\"\n",
    "output_dir = base_data_dir / \"output\"\n",
    "## defining the path to the data\n",
    "scdata_path = input_dir / \"chr22_OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad.gz\"\n",
    "gdata_path = input_dir / \"plink/\"\n",
    "dump_path = base_data_dir / \"output/example_output\"\n",
    "vep_anno_file = input_dir /  \"all_variants_annotated_fixed_ref_reversed_id.txt\"\n",
    "DNA_LM_file = input_dir/ \"annotations/onek1k_inf_scores_upstream_model.tsv\"\n",
    "## defining target chromosome\n",
    "target_chromosome = \"22\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "05cb5d18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data\n",
    "data = pd.read_pickle(output_dir/\"data_preprocessed.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "642f9025",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gnomADe_EAS_AF</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence_intergenic_variant</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>Consequence_start_lost</th>\n",
       "      <th>gnomADe_AMR_AF</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>gnomADe_NFE_AF</th>\n",
       "      <th>...</th>\n",
       "      <th>Consequence_splice_donor_5th_base_variant</th>\n",
       "      <th>Consequence_splice_donor_region_variant</th>\n",
       "      <th>Consequence_splice_donor_variant</th>\n",
       "      <th>Consequence_splice_polypyrimidine_tract_variant</th>\n",
       "      <th>Consequence_splice_region_variant</th>\n",
       "      <th>Consequence_stop_gained</th>\n",
       "      <th>Consequence_synonymous_variant</th>\n",
       "      <th>Consequence_upstream_gene_variant</th>\n",
       "      <th>DNA_LM_influence_score</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_16849573_G_A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs112435201</td>\n",
       "      <td>-</td>\n",
       "      <td>0.433139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053059</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16849971_T_A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs7287956</td>\n",
       "      <td>-</td>\n",
       "      <td>0.442607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.076044</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16850437_A_G</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs5748209</td>\n",
       "      <td>-</td>\n",
       "      <td>0.369731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.054154</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16851225_T_C</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs5746874</td>\n",
       "      <td>-</td>\n",
       "      <td>0.393139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.109348</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16851356_T_C</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs5748581</td>\n",
       "      <td>-</td>\n",
       "      <td>0.377289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.116335</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51202748_G_A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs9616963</td>\n",
       "      <td>-</td>\n",
       "      <td>0.550962</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033990</td>\n",
       "      <td>4.648319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51208568_T_G</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs148425445</td>\n",
       "      <td>-</td>\n",
       "      <td>0.102244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.031510</td>\n",
       "      <td>4.706528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51211031_G_A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs9616968</td>\n",
       "      <td>-</td>\n",
       "      <td>0.153297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.024377</td>\n",
       "      <td>5.008209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51213613_T_C</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs34726907</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.394121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.024270</td>\n",
       "      <td>0.825022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51216564_C_T</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs9616970</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.113869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.020555</td>\n",
       "      <td>0.825022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96257 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gnomADe_EAS_AF Feature_type  Consequence_intergenic_variant  \\\n",
       "snp_id                                                                         \n",
       "22_16849573_G_A             NaN            -                               1   \n",
       "22_16849971_T_A             NaN            -                               1   \n",
       "22_16850437_A_G             NaN            -                               1   \n",
       "22_16851225_T_C             NaN            -                               1   \n",
       "22_16851356_T_C             NaN            -                               1   \n",
       "...                         ...          ...                             ...   \n",
       "22_51202748_G_A             NaN   Transcript                               0   \n",
       "22_51208568_T_G             NaN   Transcript                               0   \n",
       "22_51211031_G_A             NaN   Transcript                               0   \n",
       "22_51213613_T_C             NaN   Transcript                               0   \n",
       "22_51216564_C_T             NaN   Transcript                               0   \n",
       "\n",
       "                PHENO  Consequence_start_lost  gnomADe_AMR_AF  \\\n",
       "snp_id                                                          \n",
       "22_16849573_G_A     -                       0             NaN   \n",
       "22_16849971_T_A     -                       0             NaN   \n",
       "22_16850437_A_G     -                       0             NaN   \n",
       "22_16851225_T_C     -                       0             NaN   \n",
       "22_16851356_T_C     -                       0             NaN   \n",
       "...               ...                     ...             ...   \n",
       "22_51202748_G_A     -                       0             NaN   \n",
       "22_51208568_T_G     -                       0             NaN   \n",
       "22_51211031_G_A     -                       0             NaN   \n",
       "22_51213613_T_C     -                       0             NaN   \n",
       "22_51216564_C_T     -                       0             NaN   \n",
       "\n",
       "                Existing_variation CLIN_SIG  CADD_RAW  gnomADe_NFE_AF  ...  \\\n",
       "snp_id                                                                 ...   \n",
       "22_16849573_G_A        rs112435201        -  0.433139             NaN  ...   \n",
       "22_16849971_T_A          rs7287956        -  0.442607             NaN  ...   \n",
       "22_16850437_A_G          rs5748209        -  0.369731             NaN  ...   \n",
       "22_16851225_T_C          rs5746874        -  0.393139             NaN  ...   \n",
       "22_16851356_T_C          rs5748581        -  0.377289             NaN  ...   \n",
       "...                            ...      ...       ...             ...  ...   \n",
       "22_51202748_G_A          rs9616963        -  0.550962             NaN  ...   \n",
       "22_51208568_T_G        rs148425445        -  0.102244             NaN  ...   \n",
       "22_51211031_G_A          rs9616968        -  0.153297             NaN  ...   \n",
       "22_51213613_T_C         rs34726907        - -0.394121             NaN  ...   \n",
       "22_51216564_C_T          rs9616970        - -0.113869             NaN  ...   \n",
       "\n",
       "                 Consequence_splice_donor_5th_base_variant  \\\n",
       "snp_id                                                       \n",
       "22_16849573_G_A                                          0   \n",
       "22_16849971_T_A                                          0   \n",
       "22_16850437_A_G                                          0   \n",
       "22_16851225_T_C                                          0   \n",
       "22_16851356_T_C                                          0   \n",
       "...                                                    ...   \n",
       "22_51202748_G_A                                          0   \n",
       "22_51208568_T_G                                          0   \n",
       "22_51211031_G_A                                          0   \n",
       "22_51213613_T_C                                          0   \n",
       "22_51216564_C_T                                          0   \n",
       "\n",
       "                 Consequence_splice_donor_region_variant  \\\n",
       "snp_id                                                     \n",
       "22_16849573_G_A                                        0   \n",
       "22_16849971_T_A                                        0   \n",
       "22_16850437_A_G                                        0   \n",
       "22_16851225_T_C                                        0   \n",
       "22_16851356_T_C                                        0   \n",
       "...                                                  ...   \n",
       "22_51202748_G_A                                        0   \n",
       "22_51208568_T_G                                        0   \n",
       "22_51211031_G_A                                        0   \n",
       "22_51213613_T_C                                        0   \n",
       "22_51216564_C_T                                        0   \n",
       "\n",
       "                 Consequence_splice_donor_variant  \\\n",
       "snp_id                                              \n",
       "22_16849573_G_A                                 0   \n",
       "22_16849971_T_A                                 0   \n",
       "22_16850437_A_G                                 0   \n",
       "22_16851225_T_C                                 0   \n",
       "22_16851356_T_C                                 0   \n",
       "...                                           ...   \n",
       "22_51202748_G_A                                 0   \n",
       "22_51208568_T_G                                 0   \n",
       "22_51211031_G_A                                 0   \n",
       "22_51213613_T_C                                 0   \n",
       "22_51216564_C_T                                 0   \n",
       "\n",
       "                Consequence_splice_polypyrimidine_tract_variant  \\\n",
       "snp_id                                                            \n",
       "22_16849573_G_A                                               0   \n",
       "22_16849971_T_A                                               0   \n",
       "22_16850437_A_G                                               0   \n",
       "22_16851225_T_C                                               0   \n",
       "22_16851356_T_C                                               0   \n",
       "...                                                         ...   \n",
       "22_51202748_G_A                                               0   \n",
       "22_51208568_T_G                                               0   \n",
       "22_51211031_G_A                                               0   \n",
       "22_51213613_T_C                                               0   \n",
       "22_51216564_C_T                                               0   \n",
       "\n",
       "                Consequence_splice_region_variant Consequence_stop_gained  \\\n",
       "snp_id                                                                      \n",
       "22_16849573_G_A                                 0                       0   \n",
       "22_16849971_T_A                                 0                       0   \n",
       "22_16850437_A_G                                 0                       0   \n",
       "22_16851225_T_C                                 0                       0   \n",
       "22_16851356_T_C                                 0                       0   \n",
       "...                                           ...                     ...   \n",
       "22_51202748_G_A                                 0                       0   \n",
       "22_51208568_T_G                                 0                       0   \n",
       "22_51211031_G_A                                 0                       0   \n",
       "22_51213613_T_C                                 0                       0   \n",
       "22_51216564_C_T                                 0                       0   \n",
       "\n",
       "                 Consequence_synonymous_variant  \\\n",
       "snp_id                                            \n",
       "22_16849573_G_A                               0   \n",
       "22_16849971_T_A                               0   \n",
       "22_16850437_A_G                               0   \n",
       "22_16851225_T_C                               0   \n",
       "22_16851356_T_C                               0   \n",
       "...                                         ...   \n",
       "22_51202748_G_A                               0   \n",
       "22_51208568_T_G                               0   \n",
       "22_51211031_G_A                               0   \n",
       "22_51213613_T_C                               0   \n",
       "22_51216564_C_T                               0   \n",
       "\n",
       "                 Consequence_upstream_gene_variant  DNA_LM_influence_score  \\\n",
       "snp_id                                                                       \n",
       "22_16849573_G_A                                  0                0.053059   \n",
       "22_16849971_T_A                                  0                0.076044   \n",
       "22_16850437_A_G                                  0                0.054154   \n",
       "22_16851225_T_C                                  0                0.109348   \n",
       "22_16851356_T_C                                  0                0.116335   \n",
       "...                                            ...                     ...   \n",
       "22_51202748_G_A                                  0                0.033990   \n",
       "22_51208568_T_G                                  0                0.031510   \n",
       "22_51211031_G_A                                  0                0.024377   \n",
       "22_51213613_T_C                                  0                0.024270   \n",
       "22_51216564_C_T                                  0                0.020555   \n",
       "\n",
       "                 MAF_beta_1.25  \n",
       "snp_id                          \n",
       "22_16849573_G_A       0.000227  \n",
       "22_16849971_T_A       0.000227  \n",
       "22_16850437_A_G       0.000227  \n",
       "22_16851225_T_C       0.000227  \n",
       "22_16851356_T_C       0.000227  \n",
       "...                        ...  \n",
       "22_51202748_G_A       4.648319  \n",
       "22_51208568_T_G       4.706528  \n",
       "22_51211031_G_A       5.008209  \n",
       "22_51213613_T_C       0.825022  \n",
       "22_51216564_C_T       0.825022  \n",
       "\n",
       "[96257 rows x 59 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.gdata.varm[\"annotations_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2861f6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.gdata.varm[\"annotations_0\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed2e976-2306-427d-9993-c81d13ae7db7",
   "metadata": {},
   "source": [
    "### Understanding burden computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "28a6881e-a583-4c35-805e-99ec5213e1d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['22_51197576_C_G', '22_51198868_C_T', '22_51198906_A_G',\n",
       "       '22_51198998_A_C', '22_51202748_G_A', '22_51208568_T_G',\n",
       "       '22_51211031_G_A', '22_51213613_T_C', '22_51216564_C_T'],\n",
       "      dtype='object', name='snp_id')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "this_vars = data.gdata.varm[\"annotations_0\"][data.gdata.varm[\"annotations_0\"][\"Gene\"] == \"ENSG00000184319\"].index\n",
    "this_vars\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4f22f316-96e6-4baa-a574-4413d8abf23e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "snp_id\n",
       "22_16849573_G_A                  -\n",
       "22_16849971_T_A                  -\n",
       "22_16850437_A_G                  -\n",
       "22_16851225_T_C                  -\n",
       "22_16851356_T_C                  -\n",
       "                        ...       \n",
       "22_51202748_G_A    ENSG00000184319\n",
       "22_51208568_T_G    ENSG00000184319\n",
       "22_51211031_G_A    ENSG00000184319\n",
       "22_51213613_T_C    ENSG00000184319\n",
       "22_51216564_C_T    ENSG00000184319\n",
       "Name: Gene, Length: 96257, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.gdata.varm[\"annotations_0\"][\"Gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "50794ce2-d5f3-4df5-9295-de55fbc3dce2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "snp_id\n",
       "22_16849573_G_A    22:16849573\n",
       "22_16849971_T_A    22:16849971\n",
       "22_16850437_A_G    22:16850437\n",
       "22_16851225_T_C    22:16851225\n",
       "22_16851356_T_C    22:16851356\n",
       "                      ...     \n",
       "22_51202748_G_A    22:51202748\n",
       "22_51208568_T_G    22:51208568\n",
       "22_51211031_G_A    22:51211031\n",
       "22_51213613_T_C    22:51213613\n",
       "22_51216564_C_T    22:51216564\n",
       "Name: Location, Length: 96257, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.gdata.varm[\"annotations_0\"][\"Location\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a442310a-2cb8-456c-bb4c-7c060e6cb6f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## we need the location of the gene-> choose vars based on window\n",
    "#todo: test +/- strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1fc29f0-2bd7-4d60-9094-3f2c48456148",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def get_gene_location(ensembl_id):\n",
    "    url = f\"https://rest.ensembl.org/lookup/id/{ensembl_id}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    if response.ok:\n",
    "        data = response.json()\n",
    "        #print(data)\n",
    "        if data.get('strand') == 1: # forward\n",
    "            return f\"{data.get('seq_region_name')}:{data.get('start')}-{data.get('end')}\"\n",
    "        else: # reverse\n",
    "            return f\"{data.get('seq_region_name')}:{data.get('end')}-{data.get('start')}\"\n",
    "    else:\n",
    "        return f\"Error: {response.status_code}, {response.text}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9c833361-b861-4eac-a0da-239e13ff11e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22:50798655-50799123\n",
      "11:116837622-116835751\n"
     ]
    }
   ],
   "source": [
    "#eg: forward\n",
    "gene_id = \"ENSG00000184319\"\n",
    "location = get_gene_location(gene_id)\n",
    "print(location)\n",
    "\n",
    "#eg: rev\n",
    "gene_id = \"ENSG00000118137\"\n",
    "location = get_gene_location(gene_id)\n",
    "print(location)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee1a582d",
   "metadata": {},
   "source": [
    "# Check results of burden testing\n",
    "with window size 100k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "df0d8046",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_pickle(output_dir/\"all_results_DNA_LM_and_MAF_100k.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be2a5fed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_influence_score</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>13629.0</td>\n",
       "      <td>61.807</td>\n",
       "      <td>0.342296</td>\n",
       "      <td>186.380891</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>54694.0</td>\n",
       "      <td>330.785</td>\n",
       "      <td>1.901447</td>\n",
       "      <td>1011.361912</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>11470.0</td>\n",
       "      <td>50.978</td>\n",
       "      <td>0.202178</td>\n",
       "      <td>118.862321</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>10698.0</td>\n",
       "      <td>45.424</td>\n",
       "      <td>0.155111</td>\n",
       "      <td>98.863939</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>10019.0</td>\n",
       "      <td>42.980</td>\n",
       "      <td>0.185957</td>\n",
       "      <td>130.097863</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>660213 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           DISTANCE  CADD_PHRED  DNA_LM_influence_score  MAF_beta_1.25  \\\n",
       "id                                                                       \n",
       "1_1             0.0       0.000                0.000000       0.000000   \n",
       "2_2             0.0       0.000                0.000000       0.000000   \n",
       "3_3             0.0       0.000                0.000000       0.000000   \n",
       "4_4             0.0       0.000                0.000000       0.000000   \n",
       "6_6             0.0       0.000                0.000000       0.000000   \n",
       "...             ...         ...                     ...            ...   \n",
       "1077_1078   13629.0      61.807                0.342296     186.380891   \n",
       "1078_1079   54694.0     330.785                1.901447    1011.361912   \n",
       "1079_1080   11470.0      50.978                0.202178     118.862321   \n",
       "1080_1081   10698.0      45.424                0.155111      98.863939   \n",
       "1081_1082   10019.0      42.980                0.185957     130.097863   \n",
       "\n",
       "                    Geneid  \n",
       "id                          \n",
       "1_1        ENSG00000198445  \n",
       "2_2        ENSG00000198445  \n",
       "3_3        ENSG00000198445  \n",
       "4_4        ENSG00000198445  \n",
       "6_6        ENSG00000198445  \n",
       "...                    ...  \n",
       "1077_1078  ENSG00000079974  \n",
       "1078_1079  ENSG00000079974  \n",
       "1079_1080  ENSG00000079974  \n",
       "1080_1081  ENSG00000079974  \n",
       "1081_1082  ENSG00000079974  \n",
       "\n",
       "[660213 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "62532eca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DISTANCE', 'CADD_PHRED', 'DNA_LM_influence_score', 'MAF_beta_1.25',\n",
       "       'Geneid'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1dc56b7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total GeneIds: 673\n",
      "zero value GeneIds: 256\n"
     ]
    }
   ],
   "source": [
    "# all unique geneIds\n",
    "unique_geneids = res['Geneid'].unique()\n",
    "print(f\"Total GeneIds: {str(len(unique_geneids))}\")\n",
    "zero_geneids = res[(res['DISTANCE'] == 0.0) & (res['CADD_PHRED'] == 0.0) & (res['DNA_LM_influence_score'] == 0.0) & (res['MAF_beta_1.25'] == 0.0)]['Geneid'].unique()\n",
    "print(f\"zero value GeneIds: {str(len(zero_geneids))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8143824c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35\n"
     ]
    }
   ],
   "source": [
    "# these are all genes where the ensembl id did not return a gene location\n",
    "missing_gene = []\n",
    "with open(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/missing_genes_for_burden_calc.txt\", \"r\") as file:\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        if line not in missing_gene:\n",
    "            missing_gene.append(line)\n",
    "\n",
    "print(len(missing_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4c77601c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000273203',\n",
       " 'ENSG00000272675',\n",
       " 'ENSG00000188424',\n",
       " 'ENSG00000223579',\n",
       " 'ENSG00000234252',\n",
       " 'ENSG00000225070',\n",
       " 'ENSG00000198477',\n",
       " 'ENSG00000270041',\n",
       " 'ENSG00000272955',\n",
       " 'ENSG00000260065']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_gene[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "07d9636b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "225"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check how many zero value geneId that are not in missing genes\n",
    "missing = [item for item in zero_geneids if item not in missing_gene]\n",
    "len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ca50f387-a042-43ae-b7ae-7822ead9c3ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check how many zero value geneId that are  in missing genes\n",
    "overlap = [item for item in zero_geneids if item in missing_gene]\n",
    "len(overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fcc3894f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_influence_score</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSG00000198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>13629.0</td>\n",
       "      <td>61.807</td>\n",
       "      <td>0.342296</td>\n",
       "      <td>186.380891</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>54694.0</td>\n",
       "      <td>330.785</td>\n",
       "      <td>1.901447</td>\n",
       "      <td>1011.361912</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>11470.0</td>\n",
       "      <td>50.978</td>\n",
       "      <td>0.202178</td>\n",
       "      <td>118.862321</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>10698.0</td>\n",
       "      <td>45.424</td>\n",
       "      <td>0.155111</td>\n",
       "      <td>98.863939</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>10019.0</td>\n",
       "      <td>42.980</td>\n",
       "      <td>0.185957</td>\n",
       "      <td>130.097863</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>625878 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           DISTANCE  CADD_PHRED  DNA_LM_influence_score  MAF_beta_1.25  \\\n",
       "id                                                                       \n",
       "1_1             0.0       0.000                0.000000       0.000000   \n",
       "2_2             0.0       0.000                0.000000       0.000000   \n",
       "3_3             0.0       0.000                0.000000       0.000000   \n",
       "4_4             0.0       0.000                0.000000       0.000000   \n",
       "6_6             0.0       0.000                0.000000       0.000000   \n",
       "...             ...         ...                     ...            ...   \n",
       "1077_1078   13629.0      61.807                0.342296     186.380891   \n",
       "1078_1079   54694.0     330.785                1.901447    1011.361912   \n",
       "1079_1080   11470.0      50.978                0.202178     118.862321   \n",
       "1080_1081   10698.0      45.424                0.155111      98.863939   \n",
       "1081_1082   10019.0      42.980                0.185957     130.097863   \n",
       "\n",
       "                    Geneid  \n",
       "id                          \n",
       "1_1        ENSG00000198445  \n",
       "2_2        ENSG00000198445  \n",
       "3_3        ENSG00000198445  \n",
       "4_4        ENSG00000198445  \n",
       "6_6        ENSG00000198445  \n",
       "...                    ...  \n",
       "1077_1078  ENSG00000079974  \n",
       "1078_1079  ENSG00000079974  \n",
       "1079_1080  ENSG00000079974  \n",
       "1080_1081  ENSG00000079974  \n",
       "1081_1082  ENSG00000079974  \n",
       "\n",
       "[625878 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_res = res[~res['Geneid'].isin(missing_gene)]\n",
    "filtered_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6b06cef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "660213\n",
      "625878\n"
     ]
    }
   ],
   "source": [
    "print(len(res.index))\n",
    "print(len(filtered_res.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9e83d2ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object saved to all_results_DNA_LM_and_MAF_100k_filtered.pkl\n"
     ]
    }
   ],
   "source": [
    "# Save data object\n",
    "import pickle \n",
    "with open(output_dir/\"all_results_DNA_LM_and_MAF_100k_filtered.pkl\", \"wb\") as f:\n",
    "    pickle.dump(filtered_res, f)\n",
    "\n",
    "print(\"AnnData object saved to all_results_DNA_LM_and_MAF_100k_filtered.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a76f5f0e",
   "metadata": {},
   "source": [
    "## check if ensembl ids of missing locations have something in common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7fef27e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6814d667",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene_info(ensembl_id):\n",
    "    # Base URL for Ensembl REST API\n",
    "    base_url = \"https://rest.ensembl.org/lookup/id/\"\n",
    "    # Send a GET request to fetch gene information\n",
    "    response = requests.get(f\"{base_url}{ensembl_id}\", params={\"content-type\": \"application/json\"})\n",
    "\n",
    "    # Check if the gene exists and if it is deprecated\n",
    "    print(\"got response\")\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "\n",
    "        if \"status\" in data and data[\"status\"] == \"deprecated\":\n",
    "            print(f\"Gene {ensembl_id} is deprecated.\")\n",
    "        else:\n",
    "            print(f\"Gene {ensembl_id} is valid.\")\n",
    "        \n",
    "        return data\n",
    "    else:\n",
    "        print(f\"Gene {ensembl_id} not found.\")\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ce22436f-29e9-4d07-bb96-90fae2f1f394",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'error': 'Bad Request. Check the ID format or URL structure.'}\n"
     ]
    }
   ],
   "source": [
    "def get_ensembl_id_info(ensembl_id):\n",
    "    url = f\"https://rest.ensembl.org/lookup/id/{ensembl_id}?expand=1\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    response = requests.get(url, headers=headers)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        return response.json()  # Returns ID information\n",
    "    elif response.status_code == 404:\n",
    "        return {\"error\": \"ID not found. It may be deprecated or invalid.\"}\n",
    "    elif response.status_code == 400:\n",
    "        return {\"error\": \"Bad Request. Check the ID format or URL structure.\"}\n",
    "    else:\n",
    "        return {\"error\": f\"Unexpected error: {response.status_code}\"}\n",
    "\n",
    "# Example usage\n",
    "ensembl_id = \"ENSG00000273203\"\n",
    "info = get_ensembl_id_info(ensembl_id)\n",
    "print(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "854e93ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "got response\n",
      "Gene ENSG00000273203 not found.\n"
     ]
    }
   ],
   "source": [
    "gene_info = fetch_gene_info(\"ENSG00000273203\")\n",
    "\n",
    "if gene_info:\n",
    "    print(\"Gene Information:\", gene_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a8ade5cf-ef74-4966-b698-152a58899ffa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Information: {'error': 'Unexpected error: 400'}\n"
     ]
    }
   ],
   "source": [
    "gene_info = get_ensembl_id_info(\"ENSG00000273203\")\n",
    "\n",
    "if gene_info:\n",
    "    print(\"Gene Information:\", gene_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9016e03-521e-4516-b082-e4929eba8f73",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-mm_scgenetics]",
   "language": "python",
   "name": "conda-env-.conda-mm_scgenetics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
