{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54b334f9-50f0-4f8b-a935-9d5e15aacf77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b89c8695-5a52-43ac-b079-2457198383a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/mzb_scgenetics/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import cellink as cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "15db3c0a-c54d-4746-9343-a69c65e18cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cellink.tl._annotate_snps_genotype_data import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e9a1b4dd-ec83-45d0-bcc4-5ac0d84ba362",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cellink.tl._burden_testing import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68c64c54-db98-46a7-8d00-ba6fe2e2ed1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths\n",
    "base_data_dir = Path(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/\")\n",
    "scdata_path = base_data_dir / \"input_data/OneK1K_cohort_gene_expression_matrix_14_celltypes_w_gene_locations.h5ad.gz\"\n",
    "gdata_dir = Path(\"/data/ceph/hdd/project/node_09/sys_gen_students/2024_2025/project04_rare_variant_sc/input_data/filter_vcf_r08/\")\n",
    "\n",
    "# READ FILES\n",
    "zarr_file = gdata_dir / \"chr22.dose.filtered.R2_0.8.vcz\"\n",
    "#zarr_file = os.path.join(gdata_dir, f\"chr{str(args.chromosome)}.dose.filtered.R2_0.8.vcz\")\n",
    "eigenvec = pd.read_csv(base_data_dir / \"input_data/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned.eigenvec\", sep = ' ')\n",
    "DNA_LM_upstream = base_data_dir/ \"input_data/annotations/onek1k_inf_scores_upstream_model.tsv\"\n",
    "DNA_LM_downstream = base_data_dir/ \"input_data/annotations/onek1k_inf_scores_downstream_model.tsv\"\n",
    "vep_scores = base_data_dir/ \"input_data/annotations/onek1k1_all_variants_annotated_vep.txt\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3200d270-42f1-4155-a680-038d247e0764",
   "metadata": {},
   "outputs": [],
   "source": [
    "scdata = sc.read_h5ad(scdata_path)\n",
    "gdata = cl.io.read_sgkit_zarr(zarr_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "27960a37-7b66-4555-9298-dde025d183f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp/l_back/ipykernel_1952352/1980214688.py:7: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annos = pd.read_csv(vep_scores, sep=\"\\t\", skiprows=_get_vep_start_row(vep_scores))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomADe_FIN_AF</th>\n",
       "      <th>gnomADe_NFE_AF</th>\n",
       "      <th>gnomADe_OTH_AF</th>\n",
       "      <th>gnomADe_SAS_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>TSSDistance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_715265_C/T</td>\n",
       "      <td>1:715265</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>ENST00000434264</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1.266</td>\n",
       "      <td>-0.115786</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_715265_C/T</td>\n",
       "      <td>1:715265</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>ENST00000428504</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1.266</td>\n",
       "      <td>-0.115786</td>\n",
       "      <td>1259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_715367_A/G</td>\n",
       "      <td>1:715367</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>ENST00000428504</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>2.493</td>\n",
       "      <td>-0.008796</td>\n",
       "      <td>1361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_715367_A/G</td>\n",
       "      <td>1:715367</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>ENST00000434264</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>2.493</td>\n",
       "      <td>-0.008796</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_717485_C/A</td>\n",
       "      <td>1:717485</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>ENST00000428504</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.400</td>\n",
       "      <td>-0.287018</td>\n",
       "      <td>3479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14044678</th>\n",
       "      <td>9_141072019_G/A</td>\n",
       "      <td>9:141072019</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000159247</td>\n",
       "      <td>ENST00000508529</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>6.198</td>\n",
       "      <td>0.224585</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14044679</th>\n",
       "      <td>9_141073623_G/A</td>\n",
       "      <td>9:141073623</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000159247</td>\n",
       "      <td>ENST00000508529</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>3.765</td>\n",
       "      <td>0.070335</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14044680</th>\n",
       "      <td>9_141084200_G/A</td>\n",
       "      <td>9:141084200</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>7.328</td>\n",
       "      <td>0.309063</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14044681</th>\n",
       "      <td>9_141089144_C/T</td>\n",
       "      <td>9:141089144</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000237419</td>\n",
       "      <td>ENST00000428088</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.029</td>\n",
       "      <td>-0.648286</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14044682</th>\n",
       "      <td>9_141101939_C/T</td>\n",
       "      <td>9:141101939</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000233013</td>\n",
       "      <td>ENST00000540522</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.025</td>\n",
       "      <td>-0.667585</td>\n",
       "      <td>4698</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14044683 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #Uploaded_variation     Location Allele             Gene  \\\n",
       "0               1_715265_C/T     1:715265      T  ENSG00000237491   \n",
       "1               1_715265_C/T     1:715265      T  ENSG00000228327   \n",
       "2               1_715367_A/G     1:715367      G  ENSG00000228327   \n",
       "3               1_715367_A/G     1:715367      G  ENSG00000237491   \n",
       "4               1_717485_C/A     1:717485      A  ENSG00000228327   \n",
       "...                      ...          ...    ...              ...   \n",
       "14044678     9_141072019_G/A  9:141072019      A  ENSG00000159247   \n",
       "14044679     9_141073623_G/A  9:141073623      A  ENSG00000159247   \n",
       "14044680     9_141084200_G/A  9:141084200      A                -   \n",
       "14044681     9_141089144_C/T  9:141089144      T  ENSG00000237419   \n",
       "14044682     9_141101939_C/T  9:141101939      T  ENSG00000233013   \n",
       "\n",
       "                  Feature Feature_type  \\\n",
       "0         ENST00000434264   Transcript   \n",
       "1         ENST00000428504   Transcript   \n",
       "2         ENST00000428504   Transcript   \n",
       "3         ENST00000434264   Transcript   \n",
       "4         ENST00000428504   Transcript   \n",
       "...                   ...          ...   \n",
       "14044678  ENST00000508529   Transcript   \n",
       "14044679  ENST00000508529   Transcript   \n",
       "14044680                -            -   \n",
       "14044681  ENST00000428088   Transcript   \n",
       "14044682  ENST00000540522   Transcript   \n",
       "\n",
       "                                           Consequence cDNA_position  \\\n",
       "0         intron_variant,non_coding_transcript_variant             -   \n",
       "1                                upstream_gene_variant             -   \n",
       "2                                upstream_gene_variant             -   \n",
       "3         intron_variant,non_coding_transcript_variant             -   \n",
       "4                                upstream_gene_variant             -   \n",
       "...                                                ...           ...   \n",
       "14044678                       downstream_gene_variant             -   \n",
       "14044679                       downstream_gene_variant             -   \n",
       "14044680                            intergenic_variant             -   \n",
       "14044681                       downstream_gene_variant             -   \n",
       "14044682                         upstream_gene_variant             -   \n",
       "\n",
       "         CDS_position Protein_position  ... gnomADe_FIN_AF gnomADe_NFE_AF  \\\n",
       "0                   -                -  ...              -              -   \n",
       "1                   -                -  ...              -              -   \n",
       "2                   -                -  ...              -              -   \n",
       "3                   -                -  ...              -              -   \n",
       "4                   -                -  ...              -              -   \n",
       "...               ...              ...  ...            ...            ...   \n",
       "14044678            -                -  ...              -              -   \n",
       "14044679            -                -  ...              -              -   \n",
       "14044680            -                -  ...              -              -   \n",
       "14044681            -                -  ...              -              -   \n",
       "14044682            -                -  ...              -              -   \n",
       "\n",
       "         gnomADe_OTH_AF gnomADe_SAS_AF CLIN_SIG SOMATIC PHENO CADD_PHRED  \\\n",
       "0                     -              -        -       -     -      1.266   \n",
       "1                     -              -        -       -     -      1.266   \n",
       "2                     -              -        -       -     -      2.493   \n",
       "3                     -              -        -       -     -      2.493   \n",
       "4                     -              -        -       -     -      0.400   \n",
       "...                 ...            ...      ...     ...   ...        ...   \n",
       "14044678              -              -        -       -     -      6.198   \n",
       "14044679              -              -        -       -     -      3.765   \n",
       "14044680              -              -        -       -     -      7.328   \n",
       "14044681              -              -        -       -     -      0.029   \n",
       "14044682              -              -        -       -     -      0.025   \n",
       "\n",
       "          CADD_RAW TSSDistance  \n",
       "0        -0.115786           -  \n",
       "1        -0.115786        1259  \n",
       "2        -0.008796        1361  \n",
       "3        -0.008796           -  \n",
       "4        -0.287018        3479  \n",
       "...            ...         ...  \n",
       "14044678  0.224585           -  \n",
       "14044679  0.070335           -  \n",
       "14044680  0.309063           -  \n",
       "14044681 -0.648286           -  \n",
       "14044682 -0.667585        4698  \n",
       "\n",
       "[14044683 rows x 37 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cellink.tl.utils import (\n",
    "    _add_dummy_cols,\n",
    "    _explode_columns,\n",
    "    _get_vep_start_row,\n",
    ")\n",
    "\n",
    "annos = pd.read_csv(vep_scores, sep=\"\\t\", skiprows=_get_vep_start_row(vep_scores))\n",
    "annos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1fb123a4-cb6d-4ccc-b24a-153baf7a11ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_annos = annos[annos[\"Location\"].str.split(\":\").str[0] == \"22\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "964840e8-b34f-4eeb-8d68-c4e18ffe7a80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_annos.loc[7828337, \"DISTANCE\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f3523873-1e77-428d-a351-e4b2762d3f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_annos.to_csv(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/input_data/annotations/onek1k1_chr22_variants_annotated_vep.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "99a6eb2d-93ea-414b-840b-8cfc4fa5a622",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-01-18 17:02:33,450] INFO:cellink.tl._annotate_snps_genotype_data: renaming id column #Uploaded_variation into snp_id\n",
      "[2025-01-18 17:02:33,584] INFO:cellink.tl._annotate_snps_genotype_data: Subsetting annotations to variants that are in gdata\n",
      "[2025-01-18 17:02:33,585] INFO:cellink.tl._annotate_snps_genotype_data: 0 with missing annotation\n",
      "[2025-01-18 17:02:34,037] INFO:cellink.tl._annotate_snps_genotype_data: Index(['snp_id', 'Location', 'Allele', 'Gene', 'Feature', 'Feature_type',\n",
      "       'Consequence', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE',\n",
      "       'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen',\n",
      "       'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF',\n",
      "       'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF',\n",
      "       'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'CADD_PHRED',\n",
      "       'CADD_RAW', 'TSSDistance'],\n",
      "      dtype='object')\n",
      "[2025-01-18 17:02:34,403] INFO:cellink.tl.utils: Exploding column Consequence\n",
      "[2025-01-18 17:02:35,050] INFO:cellink.tl.utils: Number of original rows: 182552.\n",
      " Number of exploded rows: 200116\n",
      "[2025-01-18 17:02:35,064] INFO:cellink.tl.utils: Making dummies from column Consequence\n",
      "[2025-01-18 17:02:35,300] INFO:cellink.tl._annotate_snps_genotype_data: Changing dtype of categorical columns ['snp_id', 'Location', 'Allele', 'Gene', 'Feature', 'Feature_type', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen', 'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF', 'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'TSSDistance']\n",
      "[2025-01-18 17:02:35,657] INFO:cellink.tl._annotate_snps_genotype_data: Expanding annotations from multiple contexts per variant\n",
      "[2025-01-18 17:02:35,658] INFO:cellink.tl._annotate_snps_genotype_data: getting unique counts\n",
      "Columns with multiple values per variant: Index(['Gene', 'Feature', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS',\n",
      "       'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen', 'TSSDistance',\n",
      "       'Consequence_3_prime_UTR_variant', 'Consequence_5_prime_UTR_variant',\n",
      "       'Consequence_NMD_transcript_variant',\n",
      "       'Consequence_coding_sequence_variant',\n",
      "       'Consequence_downstream_gene_variant',\n",
      "       'Consequence_incomplete_terminal_codon_variant',\n",
      "       'Consequence_intron_variant', 'Consequence_mature_miRNA_variant',\n",
      "       'Consequence_missense_variant',\n",
      "       'Consequence_non_coding_transcript_exon_variant',\n",
      "       'Consequence_non_coding_transcript_variant',\n",
      "       'Consequence_splice_acceptor_variant',\n",
      "       'Consequence_splice_donor_5th_base_variant',\n",
      "       'Consequence_splice_donor_region_variant',\n",
      "       'Consequence_splice_donor_variant',\n",
      "       'Consequence_splice_polypyrimidine_tract_variant',\n",
      "       'Consequence_splice_region_variant', 'Consequence_stop_gained',\n",
      "       'Consequence_stop_retained_variant', 'Consequence_synonymous_variant',\n",
      "       'Consequence_upstream_gene_variant'],\n",
      "      dtype='object')\n",
      "[2025-01-18 17:06:59,008] INFO:cellink.tl._annotate_snps_genotype_data: Maximum number of distinct annotations per variant 8\n",
      "[2025-01-18 17:08:38,034] INFO:cellink.tl._annotate_snps_genotype_data: merging non-expanded annos into first data frame\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1034 × 143083\n",
       "    var: 'chrom', 'pos', 'a0', 'a1', 'AF', 'ER2', 'maf', 'R2', 'contig', 'id', 'id_mask', 'quality'\n",
       "    varm: 'annotations_0', 'annotations_1', 'annotations_2', 'annotations_3', 'annotations_4', 'annotations_5', 'annotations_6', 'annotations_7'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_chr22=\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/input_data/annotations/onek1k1_chr22_variants_annotated_vep.txt\"\n",
    "cl.tl.add_vep_annos_to_gdata(vep_chr22, gdata,\n",
    "                             cols_to_explode=[\"Consequence\"],\n",
    "                             cols_to_dummy=[\"Consequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "01323a65-f2d7-4e55-89c4-3e7e764bc3d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(scdata.var.loc['ENSG00000243485', \"start\"])#[gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e4cf36ba-3114-42db-b09a-18a85dce239e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>features</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>28589.0</td>\n",
       "      <td>31109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>37595.0</td>\n",
       "      <td>34553.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>1</td>\n",
       "      <td>91105.0</td>\n",
       "      <td>89551.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215635</th>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268590</th>\n",
       "      <td>BAGE5</td>\n",
       "      <td>BAGE5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251180</th>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215616</th>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GeneSymbol       features chromosome    start      end\n",
       "Geneid                                                                    \n",
       "ENSG00000243485     MIR1302-10     MIR1302-10          1  28589.0  31109.0\n",
       "ENSG00000237613        FAM138A        FAM138A          1  37595.0  34553.0\n",
       "ENSG00000186092          OR4F5          OR4F5          1  65419.0  71585.0\n",
       "ENSG00000238009   RP11-34P13.7   RP11-34P13.7        NaN      NaN      NaN\n",
       "ENSG00000239945   RP11-34P13.8   RP11-34P13.8          1  91105.0  89551.0\n",
       "...                        ...            ...        ...      ...      ...\n",
       "ENSG00000215635     AC145205.1     AC145205.1        NaN      NaN      NaN\n",
       "ENSG00000268590          BAGE5          BAGE5        NaN      NaN      NaN\n",
       "ENSG00000251180     CU459201.1     CU459201.1        NaN      NaN      NaN\n",
       "ENSG00000215616     AC002321.2     AC002321.2        NaN      NaN      NaN\n",
       "ENSG00000215611     AC002321.1     AC002321.1        NaN      NaN      NaN\n",
       "\n",
       "[32738 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scdata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ec969405-a677-420b-ad56-78da7ce7cac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_res_chr22=pd.read_pickle(\"/s/project/sys_gen_students/2024_2025/project04_rare_variant_sc/output/chr22_all_results_DNA_LM_down_and_MAF_100k.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5ec58261-0a68-4732-9dce-6d636da5cd5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>1377789.0</td>\n",
       "      <td>9350.801</td>\n",
       "      <td>69.814536</td>\n",
       "      <td>133.275103</td>\n",
       "      <td>43612.917035</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>1395621.0</td>\n",
       "      <td>10775.218</td>\n",
       "      <td>77.605564</td>\n",
       "      <td>151.365002</td>\n",
       "      <td>52127.930292</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>1319479.0</td>\n",
       "      <td>8327.426</td>\n",
       "      <td>62.418011</td>\n",
       "      <td>119.240087</td>\n",
       "      <td>39069.213569</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>1434398.0</td>\n",
       "      <td>9439.675</td>\n",
       "      <td>65.065089</td>\n",
       "      <td>129.037760</td>\n",
       "      <td>44983.557785</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>1428912.0</td>\n",
       "      <td>9054.896</td>\n",
       "      <td>66.864204</td>\n",
       "      <td>127.860172</td>\n",
       "      <td>42525.702648</td>\n",
       "      <td>ENSG00000233866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>1422923.0</td>\n",
       "      <td>9165.348</td>\n",
       "      <td>63.239808</td>\n",
       "      <td>124.158291</td>\n",
       "      <td>42490.914665</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>1366478.0</td>\n",
       "      <td>9077.036</td>\n",
       "      <td>66.221163</td>\n",
       "      <td>129.565329</td>\n",
       "      <td>43114.049369</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>1595183.0</td>\n",
       "      <td>9811.044</td>\n",
       "      <td>72.452906</td>\n",
       "      <td>139.280093</td>\n",
       "      <td>43941.663597</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>1323326.0</td>\n",
       "      <td>8346.966</td>\n",
       "      <td>61.066604</td>\n",
       "      <td>119.626789</td>\n",
       "      <td>40767.972544</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1564430.0</td>\n",
       "      <td>8972.718</td>\n",
       "      <td>65.330756</td>\n",
       "      <td>128.064908</td>\n",
       "      <td>42813.245027</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>642555 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            DISTANCE  CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  \\\n",
       "id                                                                        \n",
       "1_1        1377789.0    9350.801  69.814536   133.275103   43612.917035   \n",
       "2_2        1395621.0   10775.218  77.605564   151.365002   52127.930292   \n",
       "3_3        1319479.0    8327.426  62.418011   119.240087   39069.213569   \n",
       "4_4        1434398.0    9439.675  65.065089   129.037760   44983.557785   \n",
       "6_6        1428912.0    9054.896  66.864204   127.860172   42525.702648   \n",
       "...              ...         ...        ...          ...            ...   \n",
       "1077_1078  1422923.0    9165.348  63.239808   124.158291   42490.914665   \n",
       "1078_1079  1366478.0    9077.036  66.221163   129.565329   43114.049369   \n",
       "1079_1080  1595183.0    9811.044  72.452906   139.280093   43941.663597   \n",
       "1080_1081  1323326.0    8346.966  61.066604   119.626789   40767.972544   \n",
       "1081_1082  1564430.0    8972.718  65.330756   128.064908   42813.245027   \n",
       "\n",
       "                    Geneid  \n",
       "id                          \n",
       "1_1        ENSG00000233866  \n",
       "2_2        ENSG00000233866  \n",
       "3_3        ENSG00000233866  \n",
       "4_4        ENSG00000233866  \n",
       "6_6        ENSG00000233866  \n",
       "...                    ...  \n",
       "1077_1078  ENSG00000079974  \n",
       "1078_1079  ENSG00000079974  \n",
       "1079_1080  ENSG00000079974  \n",
       "1080_1081  ENSG00000079974  \n",
       "1081_1082  ENSG00000079974  \n",
       "\n",
       "[642555 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_res_chr22[(all_res_chr22[\"CADD_PHRED\"]!=0) & (~np.isnan(all_res_chr22[\"CADD_PHRED\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b6adb1c-462c-438f-b4bb-74aa8c7bb87b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2249d7-3680-4d25-b191-3815b8c0f7a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8308ec81-6ce0-49e9-9733-e747c1330718",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PASS</th>\n",
       "      <th>GENOTYPED</th>\n",
       "      <th>GENOTYPED_ONLY</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_16849573_A_G</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_16849971_A_T</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PASS  GENOTYPED  GENOTYPED_ONLY\n",
       "snp_id                                          \n",
       "22_16849573_A_G  True      False           False\n",
       "22_16849971_A_T  True      False           False"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdata[:, [\"22_16849573_A_G\",\"22_16849971_A_T\"]].varm[\"filter\"]\n",
    "#this_weights = np.array(gdata[:, [\"22_16849573_A_G\",\"22_16849971_A_T\"]].varm['annotations_0'][weight_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "dfb9cb75-16bb-4910-a60f-5070a55b0fe5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>DNA_LM_up</th>\n",
       "      <th>DNA_LM_down</th>\n",
       "      <th>MAF_beta_1.25</th>\n",
       "      <th>Geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077_1078</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078_1079</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079_1080</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080_1081</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32115978 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           DISTANCE  CADD_PHRED  DNA_LM_up  DNA_LM_down  MAF_beta_1.25  \\\n",
       "id                                                                       \n",
       "1_1             0.0         0.0        0.0          0.0            0.0   \n",
       "2_2             0.0         0.0        0.0          0.0            0.0   \n",
       "3_3             0.0         0.0        0.0          0.0            0.0   \n",
       "4_4             0.0         0.0        0.0          0.0            0.0   \n",
       "6_6             0.0         0.0        0.0          0.0            0.0   \n",
       "...             ...         ...        ...          ...            ...   \n",
       "1077_1078       NaN         NaN        NaN          NaN            NaN   \n",
       "1078_1079       NaN         NaN        NaN          NaN            NaN   \n",
       "1079_1080       NaN         NaN        NaN          NaN            NaN   \n",
       "1080_1081       NaN         NaN        NaN          NaN            NaN   \n",
       "1081_1082       NaN         NaN        NaN          NaN            NaN   \n",
       "\n",
       "                    Geneid  \n",
       "id                          \n",
       "1_1        ENSG00000243485  \n",
       "2_2        ENSG00000243485  \n",
       "3_3        ENSG00000243485  \n",
       "4_4        ENSG00000243485  \n",
       "6_6        ENSG00000243485  \n",
       "...                    ...  \n",
       "1077_1078  ENSG00000215611  \n",
       "1078_1079  ENSG00000215611  \n",
       "1079_1080  ENSG00000215611  \n",
       "1080_1081  ENSG00000215611  \n",
       "1081_1082  ENSG00000215611  \n",
       "\n",
       "[32115978 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_res_chr22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c9190a9-47e0-4379-ad53-0500d2d6e974",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-mzb_scgenetics]",
   "language": "python",
   "name": "conda-env-anaconda-mzb_scgenetics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
