{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import warnings\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "from pathlib import Path\n",
    "\n",
    "from cellink.io import read_sgkit_zarr\n",
    "from cellink.tl import get_best_eqtl_on_single_gene\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "logger = logging.getLogger(__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEBUG = False\n",
    "TARGET_CHROMOSOME = \"22\"\n",
    "THRESHOLD = 0.05\n",
    "## running the eqtl test\n",
    "CIS_WINDOW = 1_000_000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/lollo/Work/hackathon/data/Yazar_OneK1K/gene_annotation.csv /home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/filter_zarr_r08/chr22.dose.filtered.R2_0.8.vcz /home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad.gz\n"
     ]
    }
   ],
   "source": [
    "## paths\n",
    "DATA = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K\")\n",
    "# DATA = Path(\"/Users/jan.engelmann/projects/sc-eqtl/data\")\n",
    "\n",
    "vcf_file_path = DATA / \"OneK1K_imputation_post_qc_r2_08/filter_vcf_r08/chr22.dose.filtered.R2_0.8.vcf.gz\"\n",
    "\n",
    "zarr_path = vcf_file_path.parent.parent / \"filter_zarr_r08\"\n",
    "zarr_path.mkdir(exist_ok=True)\n",
    "\n",
    "icf_file_path = zarr_path / vcf_file_path.with_suffix(\".icf\").name\n",
    "zarr_file_path = (zarr_path / vcf_file_path.stem).with_suffix(\".vcz\")\n",
    "\n",
    "gene_annotation_path = DATA / \"gene_annotation.csv\"\n",
    "\n",
    "if DEBUG:\n",
    "    scdata_path = DATA / \"debug_OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad\"\n",
    "    scdata_annotated_path = DATA / \"debug_OneK1K_cohort_gene_expression_matrix_14_celltypes_filtered_annotated.h5ad\"\n",
    "else:\n",
    "    scdata_path = DATA / \"OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad.gz\"\n",
    "    scdata_annotated_path = DATA / \"OneK1K_cohort_gene_expression_matrix_14_celltypes_filtered_annotated.h5ad.gz\"\n",
    "\n",
    "dump_path = \"/home/lollo/Work/hackathon/dump/eqtl_results.csv\"\n",
    "\n",
    "print(gene_annotation_path, zarr_file_path, scdata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>features</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215635</th>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>ENSG00000215635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268590</th>\n",
       "      <td>BAGE5</td>\n",
       "      <td>BAGE5</td>\n",
       "      <td>ENSG00000268590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251180</th>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>ENSG00000251180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215616</th>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>ENSG00000215616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GeneSymbol       features  ensembl_gene_id\n",
       "Geneid                                                        \n",
       "ENSG00000243485     MIR1302-10     MIR1302-10  ENSG00000243485\n",
       "ENSG00000237613        FAM138A        FAM138A  ENSG00000237613\n",
       "ENSG00000186092          OR4F5          OR4F5  ENSG00000186092\n",
       "ENSG00000238009   RP11-34P13.7   RP11-34P13.7  ENSG00000238009\n",
       "ENSG00000239945   RP11-34P13.8   RP11-34P13.8  ENSG00000239945\n",
       "...                        ...            ...              ...\n",
       "ENSG00000215635     AC145205.1     AC145205.1  ENSG00000215635\n",
       "ENSG00000268590          BAGE5          BAGE5  ENSG00000268590\n",
       "ENSG00000251180     CU459201.1     CU459201.1  ENSG00000251180\n",
       "ENSG00000215616     AC002321.2     AC002321.2  ENSG00000215616\n",
       "ENSG00000215611     AC002321.1     AC002321.1  ENSG00000215611\n",
       "\n",
       "[32738 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## reading single cell data\n",
    "scdata = ad.read_h5ad(scdata_path)\n",
    "## adding column needed to match the annotated data\n",
    "scdata.var[\"ensembl_gene_id\"] = scdata.var_names\n",
    "scdata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000229483</td>\n",
       "      <td>23743974</td>\n",
       "      <td>23744736</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000232849</td>\n",
       "      <td>93708910</td>\n",
       "      <td>93710179</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000229558</td>\n",
       "      <td>23993110</td>\n",
       "      <td>24002818</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000224394</td>\n",
       "      <td>94470677</td>\n",
       "      <td>94488397</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000232977</td>\n",
       "      <td>24040710</td>\n",
       "      <td>24061603</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61264</th>\n",
       "      <td>ENSG00000228242</td>\n",
       "      <td>14186223</td>\n",
       "      <td>14189784</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61265</th>\n",
       "      <td>ENSG00000150867</td>\n",
       "      <td>22823778</td>\n",
       "      <td>23003484</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61266</th>\n",
       "      <td>ENSG00000255021</td>\n",
       "      <td>14313873</td>\n",
       "      <td>14345345</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61267</th>\n",
       "      <td>ENSG00000251576</td>\n",
       "      <td>14389951</td>\n",
       "      <td>14394068</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61270</th>\n",
       "      <td>ENSG00000082898</td>\n",
       "      <td>61704984</td>\n",
       "      <td>61765761</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id  start_position  end_position chromosome_name\n",
       "4      ENSG00000229483        23743974      23744736              13\n",
       "7      ENSG00000232849        93708910      93710179              13\n",
       "11     ENSG00000229558        23993110      24002818              13\n",
       "13     ENSG00000224394        94470677      94488397              13\n",
       "15     ENSG00000232977        24040710      24061603              13\n",
       "...                ...             ...           ...             ...\n",
       "61264  ENSG00000228242        14186223      14189784               3\n",
       "61265  ENSG00000150867        22823778      23003484              10\n",
       "61266  ENSG00000255021        14313873      14345345               3\n",
       "61267  ENSG00000251576        14389951      14394068               3\n",
       "61270  ENSG00000082898        61704984      61765761               2\n",
       "\n",
       "[32738 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## reading annotation data\n",
    "annotation_df = pd.read_csv(gene_annotation_path).loc[\n",
    "    :, [\"ensembl_gene_id\", \"start_position\", \"end_position\", \"chromosome_name\"]\n",
    "]\n",
    "annotation_df = annotation_df.loc[annotation_df.ensembl_gene_id.isin(scdata.var_names)]\n",
    "annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>features</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>89295</td>\n",
       "      <td>133566</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>89551</td>\n",
       "      <td>91105</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215635</th>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>AC145205.1</td>\n",
       "      <td>44537</td>\n",
       "      <td>47211</td>\n",
       "      <td>GL000204.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268590</th>\n",
       "      <td>BAGE5</td>\n",
       "      <td>BAGE5</td>\n",
       "      <td>899</td>\n",
       "      <td>2487</td>\n",
       "      <td>GL000237.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251180</th>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>CU459201.1</td>\n",
       "      <td>12836</td>\n",
       "      <td>34543</td>\n",
       "      <td>GL000242.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215616</th>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>AC002321.2</td>\n",
       "      <td>3905</td>\n",
       "      <td>22432</td>\n",
       "      <td>GL000201.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>AC002321.1</td>\n",
       "      <td>26348</td>\n",
       "      <td>27637</td>\n",
       "      <td>GL000201.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GeneSymbol       features  start     end       chrom\n",
       "Geneid                                                                  \n",
       "ENSG00000243485     MIR1302-10     MIR1302-10  29554   31109           1\n",
       "ENSG00000237613        FAM138A        FAM138A  34554   36081           1\n",
       "ENSG00000186092          OR4F5          OR4F5  69091   70008           1\n",
       "ENSG00000238009   RP11-34P13.7   RP11-34P13.7  89295  133566           1\n",
       "ENSG00000239945   RP11-34P13.8   RP11-34P13.8  89551   91105           1\n",
       "...                        ...            ...    ...     ...         ...\n",
       "ENSG00000215635     AC145205.1     AC145205.1  44537   47211  GL000204.1\n",
       "ENSG00000268590          BAGE5          BAGE5    899    2487  GL000237.1\n",
       "ENSG00000251180     CU459201.1     CU459201.1  12836   34543  GL000242.1\n",
       "ENSG00000215616     AC002321.2     AC002321.2   3905   22432  GL000201.1\n",
       "ENSG00000215611     AC002321.1     AC002321.1  26348   27637  GL000201.1\n",
       "\n",
       "[32738 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## merging the scdata.var df with the annotations\n",
    "merged_df = scdata.var.merge(annotation_df, on=\"ensembl_gene_id\")\n",
    "merged_df = merged_df.rename(\n",
    "    columns={\"ensembl_gene_id\": \"Geneid\", \"start_position\": \"start\", \"end_position\": \"end\", \"chromosome_name\": \"chrom\"}\n",
    ")\n",
    "merged_df.index = merged_df[\"Geneid\"]\n",
    "merged_df = merged_df.drop([\"Geneid\"], axis=1)\n",
    "scdata.var = merged_df\n",
    "scdata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1034 × 143083\n",
       "    var: 'chrom', 'pos', 'a0', 'a1', 'AF', 'ER2', 'maf', 'R2', 'contig', 'id', 'id_mask', 'quality'\n",
       "    varm: 'filter'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## reading genetic data\n",
    "gdata = read_sgkit_zarr(zarr_file_path)\n",
    "gdata.obs = gdata.obs.set_index(\"id\")\n",
    "gdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_to_analyse = [\"CD4 ET\", \"CD4 NC\", \"B IN\", \"B Mem\", \"Plasma\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "673327cca8144c9e9ce3717456867eb0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/440 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8bbca6458f034ce7b13dec92cb34b369",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/515 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c9302e7d19fa4f3988eb0b5c65c86f68",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/455 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "79b8f194cf2946998c203e8249d65b0b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/426 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1e62e30546ee48e88ec02a3ac6a5a80b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/350 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## constructing result output\n",
    "results = []\n",
    "# for cell_label in scdata.obs.cell_label.unique():\n",
    "for cell_label in cells_to_analyse:\n",
    "    ## subsetting the scdata over the current cell_label\n",
    "    scdata_cell = scdata[scdata.obs.cell_label == cell_label]\n",
    "    ## normalizing counts\n",
    "    sc.pp.normalize_total(scdata_cell)\n",
    "    sc.pp.log1p(scdata_cell)\n",
    "    sc.pp.normalize_total(scdata_cell)\n",
    "    scdata_cell = scdata_cell[:, scdata_cell.var[\"chrom\"] == TARGET_CHROMOSOME]\n",
    "    ## aggregating the data\n",
    "    pbdata = sc.get.aggregate(scdata_cell, \"individual\", \"mean\")\n",
    "    gdata_cell = gdata[pbdata.obs.index]\n",
    "    pbdata.X = pbdata.layers[\"mean\"]\n",
    "    ## sanity check (we have all the individuals from both data sources)\n",
    "    assert (pbdata.obs.index == gdata_cell.obs.index).all()\n",
    "    ## filter out genes that are expressed in less than ten individuals\n",
    "    sc.pp.filter_genes(pbdata, min_cells=10)\n",
    "    ## performing eqtl tests over the genes with chrom 22\n",
    "    for target_gene in tqdm(pbdata.var_names):\n",
    "        eqtl_results = get_best_eqtl_on_single_gene(pbdata, gdata_cell, target_gene, CIS_WINDOW)\n",
    "        if len(eqtl_results) == 0:\n",
    "            print(\n",
    "                f\"No matches found for the current combination of gene {target_gene} and {cell_label} using window of {cis_window}\"\n",
    "            )\n",
    "            eqtl_results = {\"target_gene\": target_gene, \"no_tested_variants\": 0}\n",
    "        eqtl_results[\"cell_label\"] = cell_label\n",
    "        results.append(eqtl_results)\n",
    "    ## constructing output DataFrame\n",
    "    eqtl_results_df = pd.DataFrame(results)\n",
    "    ## saving the resulting dataframe\n",
    "    eqtl_results_df.to_csv(dump_path, index=False)\n",
    "# eqtl_results_df[\"pv_reject\"] = eqtl_results_df[\"min_pv\"] < THRESHOLD\n",
    "# eqtl_results_df[\"bf_pv\"] = np.clip(eqtl_results_df[\"min_pv\"]*eqtl_results_df[\"no_tested_variants\"], 0, 1)\n",
    "# eqtl_results_df[\"bf_pv_reject\"] = eqtl_results_df[\"bf_pv\"] < THRESHOLD\n",
    "# eqtl_results_df[\"q_val\"] = fdrcorrection(eqtl_results_df[\"bf_pv\"].values)[1]\n",
    "# eqtl_results_df[\"q_val_reject\"] = eqtl_results_df[\"q_val\"] < THRESHOLD\n",
    "# eqtl_results_df.to_csv(f\"/home/lollo/Work/hackathon/dump/eqtl_results.csv\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODOs\n",
    "\n",
    "- [x] Run on all genes on Chromosome 22\n",
    "- [x] For each gene store: minimum p value, number of variants tested, id of minimum pv variant, gene name\n",
    "- [x] Bonferroni correction per hit: pv_gene = pv * num_cis_variants, np clip to (0,1)\n",
    "- [x] subset to bonferroni sginifcant hits (pv_gene < 0.05)\n",
    "- [x] benjamini hochberg across tests -> qv\n",
    "- [x] report # of qv < 0.05\n",
    "- [ ] check how many hits you have compared to OneK1K\n",
    "- [x] add gwas to tools\n",
    "- [x] Figure out how to render several notebooks\n",
    "- [ ] Stretch goal: all cell types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "theislab-combosciplex",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
