{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "from pathlib import Path\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pandas_plink as pp\n",
    "import scanpy as sc\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from omegaconf import OmegaConf\n",
    "from tqdm.auto import trange\n",
    "from time import time\n",
    "from anndata.utils import asarray\n",
    "from sklearn.preprocessing import quantile_transform as sk_quantile_transform\n",
    "\n",
    "import numpy as np\n",
    "import scipy.linalg as la\n",
    "import scipy.stats as st\n",
    "\n",
    "\n",
    "log = logging.getLogger(__name__)\n",
    "\n",
    "DATA = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DATA\n",
    "def column_normalize(X):\n",
    "    assert X.ndim == 2  # noqa: PLR2004\n",
    "    return (X - X.mean(0)) / (X.std(0) * np.sqrt(X.shape[1]))\n",
    "\n",
    "\n",
    "def load_gene_annotations():\n",
    "    log.warning(\"Loading gene annotations\")\n",
    "\n",
    "    ann_path = DATA / \"onek1k_gene_annotations.pk\"\n",
    "    if ann_path.exists():\n",
    "        return pd.read_pickle(ann_path)\n",
    "    else:\n",
    "        gene_anns = sc.queries.biomart_annotations(\n",
    "            \"hsapiens\",\n",
    "            [\"ensembl_gene_id\", \"start_position\", \"end_position\", \"chromosome_name\", \"strand\"],\n",
    "            host=\"grch37.ensembl.org\",\n",
    "        ).set_index(\"ensembl_gene_id\")\n",
    "        gene_anns = gene_anns.rename(\n",
    "            columns={\n",
    "                \"start_position\": \"egene_start\",\n",
    "                \"end_position\": \"egene_end\",\n",
    "                \"chromosome_name\": \"chrom\",\n",
    "                \"strand\": \"egene_strand\",\n",
    "            }\n",
    "        )\n",
    "        gene_anns[\"tss\"] = np.where(\n",
    "            gene_anns[\"egene_strand\"] == 1, gene_anns[\"egene_start\"], gene_anns[\"egene_end\"]\n",
    "        )\n",
    "        gene_anns.to_pickle(ann_path)\n",
    "    return gene_anns\n",
    "\n",
    "\n",
    "def process_sdata(cfg, sdata):\n",
    "    log.warning(\"Filtering single cell AnnData\")\n",
    "\n",
    "    # filter cell type\n",
    "    sdata = sdata[sdata.obs.cell_label.isin(cfg.celltype_filter)].copy()\n",
    "\n",
    "    # require min cells per individual\n",
    "    cells_per_ind = sdata.obs.groupby(\"individual\", observed=True).size()\n",
    "    keep_indis = cells_per_ind[cells_per_ind >= cfg.min_cells_per_ind].index\n",
    "    sdata = sdata[sdata.obs.individual.isin(keep_indis)]\n",
    "\n",
    "    # filter non-standard genes\n",
    "    gene_anns = load_gene_annotations()\n",
    "    keep_genes = gene_anns.loc[sdata.var.index, \"chrom\"].isin(map(str, range(1, 23)))\n",
    "    sdata = sdata[:, keep_genes]\n",
    "\n",
    "    # collect all info in var\n",
    "    sdata.var = sdata.var.merge(\n",
    "        gene_anns[gene_anns.columns.difference(sdata.var.columns)],\n",
    "        how=\"left\",\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "    )\n",
    "    sdata.var = sdata.var.rename(columns={\"GeneSymbol\": \"egene_symbol\"})\n",
    "    sdata.var.index.name = \"egene_id\"\n",
    "\n",
    "    # drop unnecessary HVG columns\n",
    "    columns_to_drop = [\"features\", \"highly_variable\", \"means\", \"dispersions\", \"dispersions_norm\"]\n",
    "    sdata.var = sdata.var.drop(columns=columns_to_drop, errors=\"ignore\")\n",
    "\n",
    "    # make sure genes are ordered by chrom\n",
    "    sdata = sdata[:, sdata.var.chrom.sort_values().index]\n",
    "    \n",
    "    # lorenzo begin: computing pca\n",
    "    if \"X_pca\" not in sdata.obsm.keys():\n",
    "        sc.pp.pca(sdata)\n",
    "    # lorenzo end\n",
    "    sdata.obsm[\"X_pca\"] = column_normalize(sdata.obsm[\"X_pca\"][:, : cfg.n_cellstate_pcs])\n",
    "\n",
    "    return sdata\n",
    "\n",
    "\n",
    "def compute_mdatas(cfg, sdata):\n",
    "    log.warning(\"Creating pseudobulk mdatas\")\n",
    "\n",
    "    obs = sdata.obs.groupby(\"individual\", observed=True)[[\"sex\", \"age\"]].first()\n",
    "    obs.index = obs.index.astype(str)\n",
    "    obs[\"age_std\"] = StandardScaler().fit_transform(obs.age.values.reshape(-1, 1))\n",
    "    obs[[\"sex1\", \"sex2\"]] = np.eye(2)[(obs.sex.values - 1)]\n",
    "\n",
    "    mdatas = {}\n",
    "    keep_genes = np.full(sdata.shape[1], True)\n",
    "    for ct in [\"all\"] + list(sdata.obs.cell_label.unique()):\n",
    "        _sdata = sdata if ct == \"all\" else sdata[sdata.obs.cell_label == ct]\n",
    "\n",
    "        mdata = sc.get.aggregate(_sdata, by=\"individual\", func=\"mean\")\n",
    "        mdata.X = mdata.layers[\"mean\"]\n",
    "        keep_genes &= (mdata.X > 0).mean(axis=0) > cfg.min_frac_expressed\n",
    "        mdata.obs = obs.loc[mdata.obs.index]\n",
    "\n",
    "        sc.pp.highly_variable_genes(mdata, n_top_genes=5000)\n",
    "        sc.tl.pca(mdata, n_comps=cfg.n_expr_batch_pcs)\n",
    "        mdata.obsm[\"X_pca\"] = column_normalize(mdata.obsm[\"X_pca\"])\n",
    "\n",
    "        mdatas[ct] = mdata\n",
    "    for ct in list(mdatas.keys()):\n",
    "        mdatas[ct] = mdatas[ct][:, keep_genes]\n",
    "    return mdatas\n",
    "\n",
    "\n",
    "def load_F_and_gene_cis_snps(cfg, egene_ann, mdata, gdata=None):\n",
    "    if gdata is None or gdata.uns[\"chrom\"] != egene_ann.chrom:\n",
    "        # UNCOMMENT ME\n",
    "        # cfg[\"bed_file_path\"] = cfg[\"bed_file_path_template\"].format(chrom=egene_ann.chrom)\n",
    "        # bedfile = str(DATA / cfg.bed_file_path)\n",
    "        # pcfile = str(DATA / cfg.genetic_pc_path)\n",
    "        # gdata = read_plink(bedfile, pcfile, num_pcs=cfg.n_genetic_pcs)[mdata.obs.index]\n",
    "        # gdata.uns[\"chrom\"] = egene_ann.chrom\n",
    "        # lorenzo begin: \n",
    "        bedfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/plink/\")\n",
    "        bedfile = str(bedfile / cfg[\"bed_file_path_template\"].format(chrom=egene_ann.chrom))\n",
    "        pcfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/prunedir\")\n",
    "        pcfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/filter_vcf_r08\")\n",
    "        pcfile = str(pcfile / \"chr{chrom:}.dose.filtered.R2_0.8.filtered.pruned\".format(chrom=egene_ann.chrom))\n",
    "        print(f\"{pcfile=}, {bedfile=}\")\n",
    "        gdata = read_plink(bedfile)[mdata.obs.index]\n",
    "        gdata.uns[\"chrom\"] = egene_ann.chrom\n",
    "        # lorenzo end\n",
    "\n",
    "    F = np.concatenate(\n",
    "        [\n",
    "            mdata.obs[[\"sex1\", \"sex2\", \"age_std\"]],\n",
    "            gdata.obs,\n",
    "            mdata.obsm[\"X_pca\"][:, : cfg.n_expr_batch_pcs],\n",
    "        ],\n",
    "        axis=1,\n",
    "    )\n",
    "    F[:, 2:] = column_normalize(F[:, 2:])\n",
    "\n",
    "    cis_start = max(egene_ann.egene_start - cfg.cis_margin, 0)\n",
    "    cis_end = egene_ann.egene_end + cfg.cis_margin\n",
    "    keep_snps = (gdata.var.pos > cis_start) & (gdata.var.pos < cis_end)\n",
    "\n",
    "    if keep_snps.sum() == 0:\n",
    "        log.warning(\"Skipping %s: no cis SNPs\", egene_ann.name)\n",
    "        return gdata, None, None\n",
    "\n",
    "    cis_gdata = gdata[:, keep_snps]\n",
    "    G = pd.DataFrame(cis_gdata.X.compute(), columns=cis_gdata.var.index, index=cis_gdata.obs.index)\n",
    "    return gdata, F, G\n",
    "\n",
    "\n",
    "def read_plink(bfile, pcfile=None, num_pcs=None):\n",
    "    # load geno\n",
    "    bim, fam, bed = pp.read_plink(bfile)\n",
    "    del bim[\"cm\"]\n",
    "    del bim[\"i\"]\n",
    "    bim = bim.set_index(\"snp\")\n",
    "    fam = fam.set_index(\"iid\")[[]]\n",
    "\n",
    "    if pcfile is not None:\n",
    "        # read pcs\n",
    "        usecols = None if num_pcs is None else np.arange(num_pcs + 2)\n",
    "        dfpc = pd.read_csv(pcfile, sep=\" \", header=None, usecols=usecols)\n",
    "        dfpc.columns = [\"fid\", \"iid\"] + [f\"PC{i+1}\" for i in range(dfpc.shape[1] - 2)]\n",
    "        del dfpc[\"fid\"]\n",
    "        dfpc = dfpc.set_index(\"iid\")\n",
    "        assert (dfpc.index == fam.index).all(), \"Individuals in PC files not matching\"\n",
    "        fam = pd.concat([fam, dfpc], axis=1)\n",
    "\n",
    "    return ad.AnnData(bed.T, obs=fam, var=bim)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MODELS\n",
    "class GWAS:\n",
    "    r\"\"\"\n",
    "    Linear model for univariate association testing\n",
    "    between `P` phenotypes and `S` inputs (`P`x`S` tests)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Y : (`N`, `P`) ndarray\n",
    "        outputs\n",
    "    F : (`N`, `K`) ndarray\n",
    "        covariates. If not specified, an intercept is assumed.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, Y, F=None):\n",
    "        if F is None:\n",
    "            F = np.ones((Y.shape[0], 1))\n",
    "        self.Y = Y\n",
    "        self.F = F\n",
    "        self.df = Y.shape[0] - F.shape[1]\n",
    "        self._fit_null()\n",
    "\n",
    "    def _fit_null(self):\n",
    "        \"\"\"Internal functon. Fits the null model\"\"\"\n",
    "        self.FY = np.dot(self.F.T, self.Y)\n",
    "        self.FF = np.dot(self.F.T, self.F)\n",
    "        self.YY = np.einsum(\"ip,ip->p\", self.Y, self.Y)\n",
    "        # calc beta_F0 and s20\n",
    "        self.A0i = la.inv(self.FF)\n",
    "        self.beta_F0 = np.dot(self.A0i, self.FY)\n",
    "        self.s20 = (self.YY - np.einsum(\"kp,kp->p\", self.FY, self.beta_F0)) / self.df\n",
    "\n",
    "    def process(self, G, verbose=False):\n",
    "        r\"\"\"\n",
    "        Fit genotypes one-by-one.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        G : (`N`, `S`) ndarray\n",
    "            inputs\n",
    "        verbose : bool\n",
    "            verbose flag.\n",
    "        \"\"\"\n",
    "        t0 = time()\n",
    "        # precompute some stuff\n",
    "        GY = np.dot(G.T, self.Y)\n",
    "        GG = np.einsum(\"ij,ij->j\", G, G)\n",
    "        FG = np.dot(self.F.T, G)\n",
    "\n",
    "        # Let us denote the inverse of Areml as\n",
    "        # Ainv = [[A0i + m mt / n, m], [mT, n]]\n",
    "        A0iFG = np.dot(self.A0i, FG)\n",
    "        n = 1.0 / (GG - np.einsum(\"ij,ij->j\", FG, A0iFG))\n",
    "        M = -n * A0iFG\n",
    "        self.beta_F = (\n",
    "            self.beta_F0[:, None, :]\n",
    "            + np.einsum(\"ks,sp->ksp\", M, np.dot(M.T, self.FY)) / n[None, :, None]\n",
    "        )\n",
    "        self.beta_F += np.einsum(\"ks,sp->ksp\", M, GY)\n",
    "        self.beta_g = np.einsum(\"ks,kp->sp\", M, self.FY)\n",
    "        self.beta_g += n[:, None] * GY\n",
    "\n",
    "        # sigma\n",
    "        self.s2 = self.YY - np.einsum(\"kp,ksp->sp\", self.FY, self.beta_F)\n",
    "        self.s2 -= GY * self.beta_g\n",
    "        self.s2 /= self.df\n",
    "\n",
    "        # dlml and pvs\n",
    "        self.lrt = -self.df * np.log(self.s2 / self.s20)\n",
    "        self.pv = st.chi2(1).sf(self.lrt)\n",
    "\n",
    "        t1 = time()\n",
    "        if verbose:\n",
    "            print(\"Tested for %d variants in %.2f s\" % (G.shape[1], t1 - t0))\n",
    "\n",
    "    def getPv(self):\n",
    "        \"\"\"\n",
    "        Get pvalues\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pv : ndarray\n",
    "        \"\"\"\n",
    "        return self.pv\n",
    "\n",
    "    def getBetaSNP(self):\n",
    "        \"\"\"\n",
    "        get effect size SNPs\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        beta : ndarray\n",
    "        \"\"\"\n",
    "        return self.beta_g\n",
    "\n",
    "    def getLRT(self):\n",
    "        \"\"\"\n",
    "        get lik ratio test statistics\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        lrt : ndarray\n",
    "        \"\"\"\n",
    "        return self.lrt\n",
    "\n",
    "    def getBetaSNPste(self):\n",
    "        \"\"\"\n",
    "        get standard errors on betas\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        beta_ste : ndarray\n",
    "        \"\"\"\n",
    "        beta = self.getBetaSNP()\n",
    "        pv = self.getPv()\n",
    "        z = np.sign(beta) * np.sqrt(st.chi2(1).isf(pv))\n",
    "        ste = beta / z\n",
    "        return ste\n",
    "\n",
    "\n",
    "def quantile_transform(\n",
    "    x: np.ndarray,\n",
    "    seed: int = 1,\n",
    ") -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Gaussian quantile transform for values in a pandas Series.    :param x: Input pandas Series.\n",
    "    :type x: pd.Series\n",
    "    :param seed: Random seed.\n",
    "    :type seed: int\n",
    "    :return: Transformed Series.\n",
    "    :rtype: pd.Series    .. note::\n",
    "        “nan” values are kept\n",
    "    \"\"\"\n",
    "    np.random.seed(seed)\n",
    "    x_transform = x.copy()\n",
    "    if isinstance(x_transform, pd.Series):\n",
    "        x_transform = x_transform.to_numpy()\n",
    "    is_nan = np.isnan(x_transform)\n",
    "    n_quantiles = np.sum(~is_nan)\n",
    "    x_transform[~is_nan] = sk_quantile_transform(\n",
    "        x_transform[~is_nan].reshape([-1, 1]),\n",
    "        n_quantiles=n_quantiles,\n",
    "        subsample=n_quantiles,\n",
    "        output_distribution=\"normal\",\n",
    "        copy=True,\n",
    "    )[:, 0]\n",
    "    return x_transform\n",
    "\n",
    "def run_gwas(*, egene, mdata, F, G, suffix=None):\n",
    "    # E_pb = gaussianize(asarray(mdata[:, [egene]].X) + 1e-4 * np.random.randn(mdata.shape[0], 1))\n",
    "    E_pb = quantile_transform(asarray(mdata[:, [egene]].X) + 1e-4 * np.random.randn(mdata.shape[0], 1))\n",
    "    gwas = GWAS(Y=E_pb, F=F)\n",
    "    gwas.process(G=G)\n",
    "\n",
    "    snp_idx = gwas.getPv().argmin()\n",
    "\n",
    "    def get_top_snp(arr):\n",
    "        return arr.ravel()[snp_idx].item()\n",
    "\n",
    "    rdict = {\n",
    "        \"snp\": G.columns[snp_idx],\n",
    "        \"pv\": get_top_snp(gwas.getPv()),\n",
    "        \"beta\": get_top_snp(gwas.getBetaSNP()),\n",
    "        \"beta_ste\": get_top_snp(gwas.getBetaSNPste()),\n",
    "        \"stat\": get_top_snp(gwas.getLRT()),\n",
    "    }\n",
    "\n",
    "    if suffix is not None:\n",
    "        rdict = {f\"{k}_{suffix}\": v for k, v in rdict.items()}\n",
    "    return rdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUNS\n",
    "def run_eqtl_mapping(cfg, mdatas, sdata):\n",
    "    log.warning(\"Running GWAS\")\n",
    "    result_list = []\n",
    "\n",
    "    offset = int(cfg.offset) * cfg.n_genes\n",
    "    gdata = None\n",
    "    for gene_idx in trange(offset, offset + cfg.n_genes, desc=\"Genes\"):\n",
    "        if gene_idx >= sdata.var.shape[0]:\n",
    "            break\n",
    "        egene_ann = sdata.var.iloc[gene_idx]\n",
    "        egene = egene_ann.name\n",
    "\n",
    "        gdata, F, G = load_F_and_gene_cis_snps(cfg, egene_ann, mdatas[\"all\"], gdata)\n",
    "\n",
    "        test_results = (\n",
    "            {\"egene\": egene, \"n_cis_snps\": 0}\n",
    "            if G is None\n",
    "            else run_tests_on_gene(cfg, egene, sdata, mdatas, F, G)\n",
    "        )\n",
    "        result_list.append({**test_results, \"gene_idx\": gene_idx})\n",
    "\n",
    "    rdf = pd.DataFrame(result_list)\n",
    "\n",
    "\n",
    "# @catch_exceptions_as_dict\n",
    "def run_tests_on_gene(cfg, egene, sdata, mdatas, F, G):\n",
    "    res = {\"egene\": egene, \"n_cis_snps\": G.shape[1]}\n",
    "    # run_gwas(egene=egene, mdata=mdatas[\"all\"], F=F, G=G, suffix=\"all\", out=res)\n",
    "    run_gwas(egene=egene, mdata=mdatas[\"all\"], F=F, G=G, suffix=\"all\")\n",
    "\n",
    "    if cfg.run_coloc:\n",
    "        run_coloc(\n",
    "            egene=egene,\n",
    "            mdata1=mdatas[\"CD8 ET\"],\n",
    "            mdata2=mdatas[\"NK\"],\n",
    "            F=F,\n",
    "            G=G,\n",
    "            suffix=\"betw\",\n",
    "            out=res,\n",
    "        )\n",
    "    Gt = G[[res[\"snp_all\"]]]\n",
    "    for ct, mdata in mdatas.items():\n",
    "        ppct = ct.replace(\" \", \"_\").lower()\n",
    "\n",
    "        if ct != \"all\":\n",
    "            if cfg.run_locus:\n",
    "                run_gwas(egene=egene, mdata=mdata, F=F, G=G, suffix=f\"{ppct}_locus\", out=res)\n",
    "\n",
    "            if cfg.run_coloc:\n",
    "                run_coloc(\n",
    "                    egene=egene, mdata1=mdatas[\"all\"], mdata2=mdata, F=F, G=G, suffix=ppct, out=res\n",
    "                )\n",
    "\n",
    "            run_gwas(egene=egene, mdata=mdata, F=F, G=Gt, suffix=ppct, out=res)\n",
    "\n",
    "            if cfg.run_ivpblrt:\n",
    "                run_pb_lrt(cfg=cfg, egene=egene, mdata=mdata, F=F, G=Gt, suffix=ppct, out=res)\n",
    "\n",
    "        if cfg.run_ivgwas:\n",
    "            run_gwas_iv(egene=egene, mdata=mdata, F=F, G=Gt, suffix=ppct, out=res)\n",
    "\n",
    "    if cfg.run_mil:\n",
    "        fit_cis_mixmil(cfg=cfg, egene=egene, sdata=sdata, F=F, G=Gt, out=res)\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scdata_path = \"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad.gz\"\n",
    "scdata_path = \"/home/lollo/Work/hackathon/data/Yazar_OneK1K/debug_OneK1K_cohort_gene_expression_matrix_14_celltypes.h5ad\"\n",
    "\n",
    "conf_yaml = \"\"\"\n",
    "# process_sdata config\n",
    "celltype_filter: [\"CD4 ET\", ]\n",
    "min_cells_per_ind: 10\n",
    "n_cellstate_pcs: 50\n",
    "\n",
    "# compute_mdatas config\n",
    "min_frac_expressed: 0.1\n",
    "\n",
    "# run_eqtl mapping config\n",
    "offset: 100\n",
    "n_genes: 100\n",
    "\n",
    "# load_F_and_gene_cis_snps config\n",
    "bed_file_path_template: \"chr{chrom:}.dose.filtered.R2_0.8\"\n",
    "# genetic_pc_path: \"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/plink\"\n",
    "genetic_pc_path: \"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/pcdir/wgs.dose.filtered.R2_0.8\"\n",
    "n_genetic_pcs: 20\n",
    "n_expr_batch_pcs: 15\n",
    "cis_margin: 1_000_000\n",
    "\n",
    "# run_test_on_gene config\n",
    "run_coloc: False\n",
    "run_locus: False\n",
    "run_iivpblrt: False\n",
    "run_iv_gwas: False\n",
    "run_mil: False\n",
    "\"\"\"\n",
    "\n",
    "conf = OmegaConf.create(conf_yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Starting cis pipeline\n",
      "Loaded single cell AnnData with shape (25908, 32738)\n",
      "Filtering single cell AnnData\n",
      "Loading gene annotations\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/preprocessing/_pca.py:317: ImplicitModificationWarning: Setting element `.obsm['X_pca']` of view, initializing view as actual.\n",
      "  adata.obsm[key_obsm] = X_pca\n",
      "Creating pseudobulk mdatas\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/_utils/__init__.py:555: RuntimeWarning: overflow encountered in multiply\n",
      "  return x * y\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/preprocessing/_utils.py:41: RuntimeWarning: overflow encountered in square\n",
      "  var = mean_sq - mean**2\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/preprocessing/_utils.py:41: RuntimeWarning: invalid value encountered in subtract\n",
      "  var = mean_sq - mean**2\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/_utils/__init__.py:555: RuntimeWarning: overflow encountered in multiply\n",
      "  return x * y\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/preprocessing/_utils.py:41: RuntimeWarning: overflow encountered in square\n",
      "  var = mean_sq - mean**2\n",
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/scanpy/preprocessing/_utils.py:41: RuntimeWarning: invalid value encountered in subtract\n",
      "  var = mean_sq - mean**2\n"
     ]
    }
   ],
   "source": [
    "\n",
    "log.warning(\"Starting cis pipeline\")\n",
    "# sdata = sc.read_h5ad(RUNS / cfg.sc_adata_path)\n",
    "sdata = sc.read_h5ad(scdata_path)\n",
    "\n",
    "log.warning(\"Loaded single cell AnnData with shape %s\", sdata.shape)\n",
    "\n",
    "sdata = process_sdata(conf, sdata)\n",
    "\n",
    "mdatas = compute_mdatas(conf, sdata)\n",
    "\n",
    "sdata = sdata[:, sdata.var_names.isin(mdatas[\"all\"].var_names)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# offset = 100\n",
    "# n_genes = 100\n",
    "# n_genetic_pcs = 15\n",
    "\n",
    "# for gene_idx in trange(offset, offset + n_genes, desc=\"Genes\"):\n",
    "#     if gene_idx >= sdata.var.shape[0]:\n",
    "#         break\n",
    "#     egene_ann = sdata.var.iloc[gene_idx]\n",
    "#     egene = egene_ann.name\n",
    "#     # lorenzo begin: \n",
    "#     bedfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/plink/\")\n",
    "#     bedfile = str(bedfile / \"chr{chrom:}.dose.filtered.R2_0.8\".format(chrom=egene_ann.chrom))\n",
    "#     # pcfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/prunedir\")\n",
    "#     # pcfile = Path(\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/filter_vcf_r08\")\n",
    "#     # pcfile = str(pcfile / \"chr{chrom:}.dose.filtered.R2_0.8.filtered.pruned\".format(chrom=egene_ann.chrom))\n",
    "#     # pcfile = (\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned.fam\")\n",
    "#     pcfile = (\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned\")\n",
    "#     # pcfile = (\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned.bnosex\")\n",
    "#     # pcfile = (\"/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/pcdir/wgs.dose.filtered.R2_0.8.filtered.pruned.bim\")\n",
    "#     print(f\"{pcfile=}, {bedfile=}\")\n",
    "#     # lorenzo end\n",
    "#     # gdata = read_plink(bedfile, pcfile, num_pcs=n_genetic_pcs)[mdata[\"all\"].obs.index]\n",
    "#     gdata = read_plink(bedfile, num_pcs=n_genetic_pcs)[mdatas[\"all\"].obs.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Run eqtl mapping on adata\n",
      " View of AnnData object with n_obs × n_vars = 1332 × 11910\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'pool', 'individual', 'percent.mt', 'latent', 'nCount_SCT', 'nFeature_SCT', 'cell_type', 'cell_label', 'sex', 'age'\n",
      "    var: 'egene_symbol', 'chrom', 'egene_end', 'egene_start', 'egene_strand', 'tss'\n",
      "    uns: 'pca'\n",
      "    obsm: 'X_pca'\n",
      "    varm: 'PCs'\n",
      "Running GWAS\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0547b9c9cb8e498e877b95c823800a13",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Genes:   0%|          | 0/100 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lollo/miniconda3/envs/theislab-combosciplex/lib/python3.12/site-packages/pandera/engines/pandas_engine.py:67: UserWarning: Using typeguard < 3. Generic types like List[TYPE], Dict[TYPE, TYPE] will only validate the first element in the collection.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pcfile='/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/filter_vcf_r08/chr6.dose.filtered.R2_0.8.filtered.pruned', bedfile='/home/lollo/Work/hackathon/data/Yazar_OneK1K/OneK1K_imputation_post_qc_r2_08/plink/chr6.dose.filtered.R2_0.8'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [01:12<00:00, 24.02s/it]\n",
      "/tmp/ipykernel_167963/2380243166.py:120: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.\n",
      "  gdata.uns[\"chrom\"] = egene_ann.chrom\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'snp_all'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m log\u001b[38;5;241m.\u001b[39mwarning(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRun eqtl mapping on adata\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mstr\u001b[39m(sdata))\n\u001b[0;32m----> 2\u001b[0m \u001b[43mrun_eqtl_mapping\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmdatas\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m logging\u001b[38;5;241m.\u001b[39mwarning(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFinished cis pipeline\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[4], line 19\u001b[0m, in \u001b[0;36mrun_eqtl_mapping\u001b[0;34m(cfg, mdatas, sdata)\u001b[0m\n\u001b[1;32m     12\u001b[0m     egene \u001b[38;5;241m=\u001b[39m egene_ann\u001b[38;5;241m.\u001b[39mname\n\u001b[1;32m     14\u001b[0m     gdata, F, G \u001b[38;5;241m=\u001b[39m load_F_and_gene_cis_snps(cfg, egene_ann, mdatas[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m], gdata)\n\u001b[1;32m     16\u001b[0m     test_results \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     17\u001b[0m         {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124megene\u001b[39m\u001b[38;5;124m\"\u001b[39m: egene, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mn_cis_snps\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;241m0\u001b[39m}\n\u001b[1;32m     18\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m G \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m---> 19\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[43mrun_tests_on_gene\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcfg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43megene\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmdatas\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mF\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mG\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     20\u001b[0m     )\n\u001b[1;32m     21\u001b[0m     result_list\u001b[38;5;241m.\u001b[39mappend({\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mtest_results, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene_idx\u001b[39m\u001b[38;5;124m\"\u001b[39m: gene_idx})\n\u001b[1;32m     23\u001b[0m rdf \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(result_list)\n",
      "Cell \u001b[0;32mIn[4], line 42\u001b[0m, in \u001b[0;36mrun_tests_on_gene\u001b[0;34m(cfg, egene, sdata, mdatas, F, G)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m cfg\u001b[38;5;241m.\u001b[39mrun_coloc:\n\u001b[1;32m     33\u001b[0m     run_coloc(\n\u001b[1;32m     34\u001b[0m         egene\u001b[38;5;241m=\u001b[39megene,\n\u001b[1;32m     35\u001b[0m         mdata1\u001b[38;5;241m=\u001b[39mmdatas[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCD8 ET\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     40\u001b[0m         out\u001b[38;5;241m=\u001b[39mres,\n\u001b[1;32m     41\u001b[0m     )\n\u001b[0;32m---> 42\u001b[0m Gt \u001b[38;5;241m=\u001b[39m G[[\u001b[43mres\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msnp_all\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m]]\n\u001b[1;32m     43\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ct, mdata \u001b[38;5;129;01min\u001b[39;00m mdatas\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m     44\u001b[0m     ppct \u001b[38;5;241m=\u001b[39m ct\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39mlower()\n",
      "\u001b[0;31mKeyError\u001b[0m: 'snp_all'"
     ]
    }
   ],
   "source": [
    "\n",
    "log.warning(\"Run eqtl mapping on adata\\n %s\", str(sdata))\n",
    "run_eqtl_mapping(conf, mdatas, sdata)\n",
    "\n",
    "logging.warning(\"Finished cis pipeline\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "theislab-combosciplex",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
