{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import sgkit as sg\n",
    "import cellink  as cl\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cellink.io import read_sgkit_zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "zarr_file_path = \"/data/ouga/home/ag_gagneur/hoev/s_deeprvat/eva/theislab/hackathon24/chr22.dose.filtered.R2_0.8.vcz\"\n",
    "#genotype zarr file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data as zarr file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1034 × 143083\n",
       "    obs: 'id'\n",
       "    var: 'chrom', 'pos', 'a0', 'a1', 'AF', 'ER2', 'maf', 'R2', 'contig', 'id', 'id_mask', 'quality'\n",
       "    varm: 'filter'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdata = read_sgkit_zarr(zarr_file_path)\n",
    "gdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1034, 143083)\n",
      "variant_AF\n",
      "variant_ER2\n",
      "variant_MAF\n",
      "variant_R2\n",
      "variant_contig\n",
      "variant_id\n",
      "variant_id_mask\n",
      "variant_position\n",
      "variant_quality\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1034 × 143083\n",
       "    var: 'chrom', 'position', 'a0', 'a1', 'AF', 'ER2', 'MAF', 'R2', 'contig', 'id', 'id_mask', 'quality'\n",
       "    varm: 'filter'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# function for this to be merged soon \n",
    "X = _gdata.call_genotype.data.sum(-1).T\n",
    "print(X.shape)\n",
    "obs =  pd.DataFrame(index=_gdata.sample_id.data.compute())\n",
    "var = {}\n",
    "for k, v in _gdata.variables.items():\n",
    "    if len(v.shape) == 1 and v.shape[0] == _gdata.variants.shape[0]:\n",
    "        print(k)\n",
    "        var[k.replace(\"variant_\", \"\")] = v.data.compute()\n",
    "var = pd.DataFrame(var)\n",
    "var[\"chrom\"] = _gdata.variant_contig.to_series().map({i: v for i, v in enumerate(_gdata.contig_id.data.compute())}).values\n",
    "var[\"a0\"] = _gdata.variant_allele[:, 0].data.compute().ravel()\n",
    "var[\"a1\"] = _gdata.variant_allele[:, 1].data.compute().ravel()\n",
    "var.index = var[\"chrom\"].astype(str) + \"_\" + var[\"position\"].astype(str) + \"_\" + var[\"a0\"].astype(str) + \"_\" + var[\"a1\"].astype(str)\n",
    "var.index.name = \"variant_id\"\n",
    "first_cols = ['chrom', 'position', 'a0', 'a1']\n",
    "var = var[first_cols + [c for c in var.columns if c not in first_cols]]\n",
    "varm = {\"filter\": pd.DataFrame(_gdata.variant_filter.data.compute(), index=var.index, columns=_gdata.filter_id.data.compute())}\n",
    "gdata = ad.AnnData(X=X, obs=obs, var=var, varm=varm)\n",
    "gdata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write variants to VCF for annotation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2024-10-09 10:13:45,906] INFO:cellink.tl._annotate_snps_genotype_data: number of variants to annotate: 143083\n",
      "[2024-10-09 10:13:45,907] INFO:cellink.tl._annotate_snps_genotype_data: Writing variants to variants.vcf\n"
     ]
    }
   ],
   "source": [
    "cl.tl.write_variants_to_vcf(gdata, out_file=\"variants.vcf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using VEP "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### run VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_file = \"variant_vep_annotated.txt\"\n",
    "config_file  = \"../docs/example/config.yaml\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# cl.tl.run_vep(config_file,\n",
    "#              input_vcf=\"variants.vcf\",\n",
    "#              output=anno_file) #writes \"variant_vep_annotated.txt\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### process VEP annos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vep_annos(vep_anno_file, cols_to_explode=[\"Consequence\"],\n",
    "                  cols_to_dummy=[\"Consequence\"]):\n",
    "    \n",
    "    annos = pd.read_csv(vep_anno_file,\n",
    "                        sep = '\\t', \n",
    "                skiprows = _get_vep_start_row(anno_file))\n",
    "    logger.info(f\"{annos.columns}\")\n",
    "    annos.replace('-', np.nan, inplace=True)\n",
    "    for col in cols_to_explode:\n",
    "        annos = _explode_columns(annos, col)\n",
    "    for col in cols_to_dummy:\n",
    "        annos = _add_dummy_cols(annos, col)\n",
    "    #TODO: make function to collapse such that only one row per variant\n",
    "    return(annos)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2024-10-09 10:14:07,682] INFO:cellink.tl._annotate_snps_genotype_data: Index(['#Uploaded_variation', 'Location', 'Allele', 'Gene', 'Feature',\n",
      "       'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position',\n",
      "       'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation',\n",
      "       'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP',\n",
      "       'SIFT', 'PolyPhen', 'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF',\n",
      "       'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF',\n",
      "       'gnomADe_OTH_AF', 'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO',\n",
      "       'CADD_PHRED', 'CADD_RAW', 'TSSDistance'],\n",
      "      dtype='object')\n",
      "[2024-10-09 10:14:08,004] INFO:cellink.tl.utils: Exploding column Consequence\n",
      "[2024-10-09 10:14:08,540] INFO:cellink.tl.utils: Number of original rows: 182552.\n",
      " Number of exploded rows: 200116\n",
      "[2024-10-09 10:14:08,553] INFO:cellink.tl.utils: Making dummies from column Consequence\n",
      "[2024-10-09 10:14:08,748] INFO:cellink.tl._annotate_snps_genotype_data: Aggregating annotations from multiple contexts to get one row per variant\n",
      "[2024-10-09 10:14:08,880] INFO:cellink.tl._annotate_snps_genotype_data: Number of variant ids with >1 annotation 39047\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/nasif12/home_if12/hoev/git/sc-genetics/src/cellink/tl/_annotate_snps_genotype_data.py:186: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  annos_cond.index.to_frame()[annos_cond.index.value_counts() > 1].index.unique()\n",
      "/data/nasif12/home_if12/hoev/git/sc-genetics/src/cellink/tl/_annotate_snps_genotype_data.py:194: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  annos_cond_sub = annos_cond_sub.applymap(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>...</th>\n",
       "      <th>Consequence_splice_donor_5th_base_variant</th>\n",
       "      <th>Consequence_splice_donor_region_variant</th>\n",
       "      <th>Consequence_splice_donor_variant</th>\n",
       "      <th>Consequence_splice_polypyrimidine_tract_variant</th>\n",
       "      <th>Consequence_splice_region_variant</th>\n",
       "      <th>Consequence_start_lost</th>\n",
       "      <th>Consequence_stop_gained</th>\n",
       "      <th>Consequence_stop_retained_variant</th>\n",
       "      <th>Consequence_synonymous_variant</th>\n",
       "      <th>Consequence_upstream_gene_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22_47192595_G/A</th>\n",
       "      <td>22:47192595</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000054611</td>\n",
       "      <td>ENST00000337137</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_27004964_G/A</th>\n",
       "      <td>22:27004964</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000100122</td>\n",
       "      <td>ENST00000215939</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_34890077_C/G</th>\n",
       "      <td>22:34890077</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_46931077_G/C</th>\n",
       "      <td>22:46931077</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000075275</td>\n",
       "      <td>ENST00000262738</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>1991/11389</td>\n",
       "      <td>1991/9045</td>\n",
       "      <td>664/3014</td>\n",
       "      <td>S/W</td>\n",
       "      <td>tCg/tGg</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_35976572_A/G</th>\n",
       "      <td>22:35976572</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51202748_A/G</th>\n",
       "      <td>22:51202748</td>\n",
       "      <td>G</td>\n",
       "      <td>[ENSG00000079974, ENSG00000184319, ENSG0000018...</td>\n",
       "      <td>[ENST00000395593, ENST00000496652, ENST0000049...</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51208568_G/T</th>\n",
       "      <td>22:51208568</td>\n",
       "      <td>T</td>\n",
       "      <td>[ENSG00000184319, ENSG00000184319, ENSG0000007...</td>\n",
       "      <td>[ENST00000496652, ENST00000496652, ENST0000039...</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51211031_A/G</th>\n",
       "      <td>22:51211031</td>\n",
       "      <td>G</td>\n",
       "      <td>[ENSG00000184319, ENSG00000184319, ENSG0000007...</td>\n",
       "      <td>[ENST00000496652, ENST00000496652, ENST0000039...</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51213613_C/T</th>\n",
       "      <td>22:51213613</td>\n",
       "      <td>T</td>\n",
       "      <td>[ENSG00000079974, ENSG00000184319, ENSG0000018...</td>\n",
       "      <td>[ENST00000395593, ENST00000496652, ENST0000049...</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51216564_T/C</th>\n",
       "      <td>22:51216564</td>\n",
       "      <td>C</td>\n",
       "      <td>[ENSG00000079974, ENSG00000184319, ENSG0000018...</td>\n",
       "      <td>[ENST00000395593, ENST00000496652, ENST0000049...</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143083 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Location Allele  \\\n",
       "#Uploaded_variation                       \n",
       "22_47192595_G/A      22:47192595      A   \n",
       "22_27004964_G/A      22:27004964      A   \n",
       "22_34890077_C/G      22:34890077      G   \n",
       "22_46931077_G/C      22:46931077      C   \n",
       "22_35976572_A/G      22:35976572      G   \n",
       "...                          ...    ...   \n",
       "22_51202748_A/G      22:51202748      G   \n",
       "22_51208568_G/T      22:51208568      T   \n",
       "22_51211031_A/G      22:51211031      G   \n",
       "22_51213613_C/T      22:51213613      T   \n",
       "22_51216564_T/C      22:51216564      C   \n",
       "\n",
       "                                                                  Gene  \\\n",
       "#Uploaded_variation                                                      \n",
       "22_47192595_G/A                                        ENSG00000054611   \n",
       "22_27004964_G/A                                        ENSG00000100122   \n",
       "22_34890077_C/G                                                    NaN   \n",
       "22_46931077_G/C                                        ENSG00000075275   \n",
       "22_35976572_A/G                                                    NaN   \n",
       "...                                                                ...   \n",
       "22_51202748_A/G      [ENSG00000079974, ENSG00000184319, ENSG0000018...   \n",
       "22_51208568_G/T      [ENSG00000184319, ENSG00000184319, ENSG0000007...   \n",
       "22_51211031_A/G      [ENSG00000184319, ENSG00000184319, ENSG0000007...   \n",
       "22_51213613_C/T      [ENSG00000079974, ENSG00000184319, ENSG0000018...   \n",
       "22_51216564_T/C      [ENSG00000079974, ENSG00000184319, ENSG0000018...   \n",
       "\n",
       "                                                               Feature  \\\n",
       "#Uploaded_variation                                                      \n",
       "22_47192595_G/A                                        ENST00000337137   \n",
       "22_27004964_G/A                                        ENST00000215939   \n",
       "22_34890077_C/G                                                    NaN   \n",
       "22_46931077_G/C                                        ENST00000262738   \n",
       "22_35976572_A/G                                                    NaN   \n",
       "...                                                                ...   \n",
       "22_51202748_A/G      [ENST00000395593, ENST00000496652, ENST0000049...   \n",
       "22_51208568_G/T      [ENST00000496652, ENST00000496652, ENST0000039...   \n",
       "22_51211031_A/G      [ENST00000496652, ENST00000496652, ENST0000039...   \n",
       "22_51213613_C/T      [ENST00000395593, ENST00000496652, ENST0000049...   \n",
       "22_51216564_T/C      [ENST00000395593, ENST00000496652, ENST0000049...   \n",
       "\n",
       "                    Feature_type cDNA_position CDS_position Protein_position  \\\n",
       "#Uploaded_variation                                                            \n",
       "22_47192595_G/A       Transcript           NaN          NaN              NaN   \n",
       "22_27004964_G/A       Transcript           NaN          NaN              NaN   \n",
       "22_34890077_C/G              NaN           NaN          NaN              NaN   \n",
       "22_46931077_G/C       Transcript    1991/11389    1991/9045         664/3014   \n",
       "22_35976572_A/G              NaN           NaN          NaN              NaN   \n",
       "...                          ...           ...          ...              ...   \n",
       "22_51202748_A/G       Transcript           NaN          NaN              NaN   \n",
       "22_51208568_G/T       Transcript           NaN          NaN              NaN   \n",
       "22_51211031_A/G       Transcript           NaN          NaN              NaN   \n",
       "22_51213613_C/T       Transcript           NaN          NaN              NaN   \n",
       "22_51216564_T/C       Transcript           NaN          NaN              NaN   \n",
       "\n",
       "                    Amino_acids   Codons  ...  \\\n",
       "#Uploaded_variation                       ...   \n",
       "22_47192595_G/A             NaN      NaN  ...   \n",
       "22_27004964_G/A             NaN      NaN  ...   \n",
       "22_34890077_C/G             NaN      NaN  ...   \n",
       "22_46931077_G/C             S/W  tCg/tGg  ...   \n",
       "22_35976572_A/G             NaN      NaN  ...   \n",
       "...                         ...      ...  ...   \n",
       "22_51202748_A/G             NaN      NaN  ...   \n",
       "22_51208568_G/T             NaN      NaN  ...   \n",
       "22_51211031_A/G             NaN      NaN  ...   \n",
       "22_51213613_C/T             NaN      NaN  ...   \n",
       "22_51216564_T/C             NaN      NaN  ...   \n",
       "\n",
       "                    Consequence_splice_donor_5th_base_variant  \\\n",
       "#Uploaded_variation                                             \n",
       "22_47192595_G/A                                             0   \n",
       "22_27004964_G/A                                             0   \n",
       "22_34890077_C/G                                             0   \n",
       "22_46931077_G/C                                             0   \n",
       "22_35976572_A/G                                             0   \n",
       "...                                                       ...   \n",
       "22_51202748_A/G                                             0   \n",
       "22_51208568_G/T                                             0   \n",
       "22_51211031_A/G                                             0   \n",
       "22_51213613_C/T                                             0   \n",
       "22_51216564_T/C                                             0   \n",
       "\n",
       "                    Consequence_splice_donor_region_variant  \\\n",
       "#Uploaded_variation                                           \n",
       "22_47192595_G/A                                           0   \n",
       "22_27004964_G/A                                           0   \n",
       "22_34890077_C/G                                           0   \n",
       "22_46931077_G/C                                           0   \n",
       "22_35976572_A/G                                           0   \n",
       "...                                                     ...   \n",
       "22_51202748_A/G                                           0   \n",
       "22_51208568_G/T                                           0   \n",
       "22_51211031_A/G                                           0   \n",
       "22_51213613_C/T                                           0   \n",
       "22_51216564_T/C                                           0   \n",
       "\n",
       "                    Consequence_splice_donor_variant  \\\n",
       "#Uploaded_variation                                    \n",
       "22_47192595_G/A                                    0   \n",
       "22_27004964_G/A                                    0   \n",
       "22_34890077_C/G                                    0   \n",
       "22_46931077_G/C                                    0   \n",
       "22_35976572_A/G                                    0   \n",
       "...                                              ...   \n",
       "22_51202748_A/G                                    0   \n",
       "22_51208568_G/T                                    0   \n",
       "22_51211031_A/G                                    0   \n",
       "22_51213613_C/T                                    0   \n",
       "22_51216564_T/C                                    0   \n",
       "\n",
       "                    Consequence_splice_polypyrimidine_tract_variant  \\\n",
       "#Uploaded_variation                                                   \n",
       "22_47192595_G/A                                                   0   \n",
       "22_27004964_G/A                                                   0   \n",
       "22_34890077_C/G                                                   0   \n",
       "22_46931077_G/C                                                   0   \n",
       "22_35976572_A/G                                                   0   \n",
       "...                                                             ...   \n",
       "22_51202748_A/G                                                   0   \n",
       "22_51208568_G/T                                                   0   \n",
       "22_51211031_A/G                                                   0   \n",
       "22_51213613_C/T                                                   0   \n",
       "22_51216564_T/C                                                   0   \n",
       "\n",
       "                    Consequence_splice_region_variant Consequence_start_lost  \\\n",
       "#Uploaded_variation                                                            \n",
       "22_47192595_G/A                                     0                      0   \n",
       "22_27004964_G/A                                     0                      0   \n",
       "22_34890077_C/G                                     0                      0   \n",
       "22_46931077_G/C                                     0                      0   \n",
       "22_35976572_A/G                                     0                      0   \n",
       "...                                               ...                    ...   \n",
       "22_51202748_A/G                                     0                      0   \n",
       "22_51208568_G/T                                     0                      0   \n",
       "22_51211031_A/G                                     0                      0   \n",
       "22_51213613_C/T                                     0                      0   \n",
       "22_51216564_T/C                                     0                      0   \n",
       "\n",
       "                    Consequence_stop_gained Consequence_stop_retained_variant  \\\n",
       "#Uploaded_variation                                                             \n",
       "22_47192595_G/A                           0                                 0   \n",
       "22_27004964_G/A                           0                                 0   \n",
       "22_34890077_C/G                           0                                 0   \n",
       "22_46931077_G/C                           0                                 0   \n",
       "22_35976572_A/G                           0                                 0   \n",
       "...                                     ...                               ...   \n",
       "22_51202748_A/G                           0                                 0   \n",
       "22_51208568_G/T                           0                                 0   \n",
       "22_51211031_A/G                           0                                 0   \n",
       "22_51213613_C/T                           0                                 0   \n",
       "22_51216564_T/C                           0                                 0   \n",
       "\n",
       "                    Consequence_synonymous_variant  \\\n",
       "#Uploaded_variation                                  \n",
       "22_47192595_G/A                                  0   \n",
       "22_27004964_G/A                                  0   \n",
       "22_34890077_C/G                                  0   \n",
       "22_46931077_G/C                                  0   \n",
       "22_35976572_A/G                                  0   \n",
       "...                                            ...   \n",
       "22_51202748_A/G                                  0   \n",
       "22_51208568_G/T                                  0   \n",
       "22_51211031_A/G                                  0   \n",
       "22_51213613_C/T                                  0   \n",
       "22_51216564_T/C                                  0   \n",
       "\n",
       "                    Consequence_upstream_gene_variant  \n",
       "#Uploaded_variation                                    \n",
       "22_47192595_G/A                                     0  \n",
       "22_27004964_G/A                                     0  \n",
       "22_34890077_C/G                                     0  \n",
       "22_46931077_G/C                                     0  \n",
       "22_35976572_A/G                                     0  \n",
       "...                                               ...  \n",
       "22_51202748_A/G                                     0  \n",
       "22_51208568_G/T                                     0  \n",
       "22_51211031_A/G                                     0  \n",
       "22_51213613_C/T                                     0  \n",
       "22_51216564_T/C                                     0  \n",
       "\n",
       "[143083 rows x 58 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annos = cl.tl.read_vep_annos(anno_file, \n",
    "                             cols_to_explode=[\"Consequence\"],\n",
    "                             cols_to_dummy=[\"Consequence\"])\n",
    "annos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2024-10-09 10:14:37,060] INFO:cellink.tl._annotate_snps_genotype_data: Joining gdata.var with annos on index\n",
      "Index(['chrom', 'pos', 'a0', 'a1', 'AF', 'ER2', 'maf', 'R2', 'contig', 'id',\n",
      "       'id_mask', 'quality', 'Location', 'Allele', 'Gene', 'Feature',\n",
      "       'Feature_type', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE',\n",
      "       'STRAND', 'FLAGS', 'BIOTYPE', 'CANONICAL', 'ENSP', 'SIFT', 'PolyPhen',\n",
      "       'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF',\n",
      "       'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF',\n",
      "       'gnomADe_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'CADD_PHRED',\n",
      "       'CADD_RAW', 'TSSDistance', 'Consequence_3_prime_UTR_variant',\n",
      "       'Consequence_5_prime_UTR_variant', 'Consequence_NMD_transcript_variant',\n",
      "       'Consequence_coding_sequence_variant',\n",
      "       'Consequence_downstream_gene_variant',\n",
      "       'Consequence_incomplete_terminal_codon_variant',\n",
      "       'Consequence_intergenic_variant', 'Consequence_intron_variant',\n",
      "       'Consequence_mature_miRNA_variant', 'Consequence_missense_variant',\n",
      "       'Consequence_non_coding_transcript_exon_variant',\n",
      "       'Consequence_non_coding_transcript_variant',\n",
      "       'Consequence_splice_acceptor_variant',\n",
      "       'Consequence_splice_donor_5th_base_variant',\n",
      "       'Consequence_splice_donor_region_variant',\n",
      "       'Consequence_splice_donor_variant',\n",
      "       'Consequence_splice_polypyrimidine_tract_variant',\n",
      "       'Consequence_splice_region_variant', 'Consequence_start_lost',\n",
      "       'Consequence_stop_gained', 'Consequence_stop_retained_variant',\n",
      "       'Consequence_synonymous_variant', 'Consequence_upstream_gene_variant'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "gdata = cl.tl.merge_annos_into_gdata(annos, gdata)\n",
    "print(gdata.var.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "eh_scgenetics_v1",
   "language": "python",
   "name": "eh_scgenetics_v1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
